// @arcgis/charts-components/dist/components/arcgis-charts-config-guides/customElement.js
import { c as p } from "../../chunks/runtime.js";
import { cloneDeep as r } from "lodash-es";
import { d as g } from "../../chunks/default.js";
import { g as u, H as c } from "../../chunks/resource.js";
import { u as f } from "../../chunks/useT9n.js";
import { css as A, html as a } from "lit";
import { createRef as x, ref as C } from "lit/directives/ref.js";
import { LitElement as b, renderElement as v, safeClassMap as m } from "@arcgis/lumina";
import { G as o } from "../../chunks/common.js";
import { n as l, o as h } from "../../chunks/type-guards.js";
import "../../chunks/series-types.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const P = A`:host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin)}.add-guide{display:flex;justify-content:center;padding:.5rem}.add-guide-btn{width:100%}.no-guides{border-radius:2rem}.no-messages{text-align:center}.guide-icon{padding-left:.5rem;padding-top:.5rem;width:1.5rem;height:1.5rem}`;
class $ extends b {
  constructor() {
    super(...arguments), this._messages = f(), this.addGuideButton = x(), this.guideTypePopover = null, this.addXGuide = (e) => {
      if (l(this.model)) {
        const t = this.getUniqueName(this._messages.xAxisGuide ?? ""), i = this.model.getXGuidesLength() ?? 0;
        this.model.addXAxisGuide(t, i), this.model.setGuideStyle(r(g.style), i, o.X), this.latestGuidePosition = u(i, o.X, this.model), this.togglePopOver(e), (this.el.shadowRoot?.querySelectorAll("calcite-block") ?? []).forEach((d, n) => {
          n !== this.latestGuidePosition && (d.open = !1);
        });
      }
    }, this.addYGuide = (e) => {
      if (h(this.model)) {
        const t = this.getUniqueName(this._messages.yAxisGuide ?? ""), i = this.model.getYGuidesLength() ?? 0;
        this.model.addYAxisGuide(t, i), this.model.setGuideStyle(r(g.style), i, o.Y), this.latestGuidePosition = u(i, o.Y, this.model), this.togglePopOver(e);
        for (const s of this.el.shadowRoot?.querySelectorAll("calcite-block") ?? [])
          s.open = !1;
      }
    }, this.togglePopOver = (e) => {
      e.stopImmediatePropagation(), this.enableXAxisGuide && this.enableYAxisGuide && (this.showAddButton = !this.showAddButton);
    }, this.onModelConfigChange = () => {
      this.requestUpdate();
    }, this.closeAddPopOver = (e) => {
      e.stopImmediatePropagation(), this.showAddButton = !0;
    }, this.showAddButton = !0, this.latestGuidePosition = 0, this.open = !1;
  }
  static {
    this.properties = { showAddButton: 16, latestGuidePosition: 16, messageOverrides: 0, model: 0, timeZone: 3, calculatedMinMaxBounds: 2, enableXAxisGuide: 5, enableYAxisGuide: 5, open: 7 };
  }
  static {
    this.styles = P;
  }
  load() {
    this.modelChange(this.model);
  }
  willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model"));
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, t) {
    t?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  getAllGuidesLength() {
    const e = l(this.model) ? this.model.getXGuidesLength() ?? 0 : 0, t = h(this.model) ? this.model.getYGuidesLength() ?? 0 : 0;
    return e + t;
  }
  getUniqueName(e, t = 1) {
    const i = `${e} ${t}`;
    let s = !1;
    if (l(this.model)) {
      const d = this.model.getXGuidesLength() ?? 0;
      for (let n = 0; n < d && !s; n += 1)
        this.model.getGuideName(n, o.X) === i && (s = !0);
    }
    if (h(this.model)) {
      const d = this.model.getYGuidesLength() ?? 0;
      for (let n = 0; n < d && !s; n += 1)
        this.model.getGuideName(n, o.Y) === i && (s = !0);
    }
    return s ? this.getUniqueName(e, t + 1) : i;
  }
  setChangedGuide(e) {
    const { index: t, orientation: i, isDelete: s } = e.detail, d = u(t, i, this.model);
    s && d !== void 0 && d <= this.latestGuidePosition && this.latestGuidePosition > 0 && (this.latestGuidePosition -= 1);
  }
  setupGuidePopover(e) {
    e.stopImmediatePropagation(), this.togglePopOver(e);
  }
  renderGuideTypePopover() {
    const e = this.getAllGuidesLength();
    this.guideTypePopover = v(a`<arcgis-charts-config-guides-type-popover .enableXAxisGuide=${this.enableXAxisGuide} .enableYAxisGuide=${this.enableYAxisGuide} .referenceElement=${this.addGuideButton.value} .placement=${e === 0 ? "bottom" : "top"} .open=${!this.showAddButton} @arcgisChartsAddXAxisGuide=${this.addXGuide} @arcgisChartsAddYAxisGuide=${this.addYGuide} @arcgisChartsConfigPopoverClose=${this.closeAddPopOver}></arcgis-charts-config-guides-type-popover>`), !this.showAddButton && !document.body.contains(this.guideTypePopover) && this.guideTypePopover !== null && document.body.appendChild(this.guideTypePopover);
  }
  renderGuide(e, t) {
    let i;
    return (this.enableXAxisGuide && t === o.X || this.enableYAxisGuide && t === o.Y) && (i = a`<arcgis-charts-config-guide .model=${this.model} .timeZone=${this.timeZone} .calculatedMinMaxBounds=${this.calculatedMinMaxBounds} .index=${e} .orientation=${t} .latestGuidePosition=${this.latestGuidePosition} @arcgisChartsConfigGuideChange=${this.setChangedGuide}></arcgis-charts-config-guide>`), i;
  }
  renderGuides() {
    const e = [];
    if (l(this.model)) {
      const t = this.model.getXGuidesLength() ?? 0;
      for (let i = 0; i < t; i += 1)
        e.push(this.renderGuide(i, o.X));
    }
    if (h(this.model)) {
      const t = this.model.getYGuidesLength() ?? 0;
      for (let i = 0; i < t; i += 1)
        e.push(this.renderGuide(i, o.Y));
    }
    return e;
  }
  renderAddGuideButton(e) {
    return a`<calcite-fab slot=fab appearance=outline-fill kind=neutral text-enabled .text=${this._messages.addGuide} .label=${this._messages.addGuide} @click=${e} scale=s></calcite-fab>`;
  }
  render() {
    const t = this.getAllGuidesLength() > 0, i = this.showAddButton && this.enableXAxisGuide && this.enableYAxisGuide, s = this.enableXAxisGuide && !this.enableYAxisGuide, d = !this.enableXAxisGuide && this.enableYAxisGuide;
    return a`${this.renderGuideTypePopover()}${t ? this.renderGuides() : a`<div class=${m(c.noMessages)}>${this._messages.noGuidesMsg}</div>`}<div class=${m(c.addGuide)} ${C(this.addGuideButton)}>${i && this.renderAddGuideButton(this.setupGuidePopover) || ""}${s && this.renderAddGuideButton(this.addXGuide) || ""}${d && this.renderAddGuideButton(this.addYGuide) || ""}</div>`;
  }
}
p("arcgis-charts-config-guides", $);
export {
  $ as ArcgisChartsConfigGuides
};
