// @arcgis/charts-components/dist/components/arcgis-charts-config-line-chart-series/customElement.js
import { c as f } from "../../chunks/runtime.js";
import { keyed as u } from "lit/directives/keyed.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "../../chunks/series-types.js";
import "d3-array";
import { cC as I } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "@arcgis/toolkit/dom";
import { r as d } from "../../chunks/popover-ui-utils.js";
import { LitElement as $, renderElement as L, safeStyleMap as x } from "@arcgis/lumina";
import { css as P, html as o } from "lit";
import { repeat as y } from "lit/directives/repeat.js";
import { u as b } from "../../chunks/useT9n.js";
import { g as M } from "../../chunks/chart-ui-utils2.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const O = P`:host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin)}`;
class k extends $ {
  constructor() {
    super(...arguments), this._messages = b(), this.seriesPopover = null, this.open = !1, this.seriesNameChange = (e) => {
      if (e.detail !== null || e.detail !== void 0) {
        const s = e.detail?.name;
        this.model.setSeriesName(s, this.selectedSeriesIndex ?? -1);
      }
    }, this.seriesColorChange = (e) => {
      if (e.detail !== null && e.detail !== void 0) {
        const { color: s } = e.detail;
        this.model.colorMatch = !1, s !== void 0 && (this.model.setSeriesColor(s, this.selectedSeriesIndex ?? -1), this.model.setSeriesMarkerColor(s, this.selectedSeriesIndex ?? -1));
      }
    }, this.seriesLineStyleChange = (e) => {
      if (e.detail !== null && e.detail !== void 0) {
        const { lineStyle: s } = e.detail;
        s !== void 0 && this.model.setSeriesLineStyle(s, this.selectedSeriesIndex ?? -1);
      }
    }, this.seriesLineWidthChange = (e) => {
      if (e.detail !== null && e.detail !== void 0) {
        const { lineWidth: s } = e.detail;
        s !== void 0 && (this.model.setSeriesLineWidth(s, this.selectedSeriesIndex ?? -1), this.model.setSeriesMarkerSize(s * 2, this.selectedSeriesIndex ?? -1));
      }
    }, this.seriesPopoverClose = (e) => {
      const { selectedSeriesIndex: s } = e.detail;
      this.selectedSeriesIndex === s && (this.seriesPopover = d(this.seriesPopover), this.selectedSeriesIndex = void 0);
    }, this.onModelConfigChange = () => {
      this.requestUpdate();
    }, this.seriesOrderResetCounter = 0, this.activeColorMatch = !1, this.popoverPlacement = "leading";
  }
  static {
    this.properties = { selectedSeriesIndex: 16, seriesOrderResetCounter: 16, messageOverrides: 0, model: 0, activeColorMatch: 5, popoverPlacement: 1 };
  }
  static {
    this.styles = O;
  }
  load() {
    this.modelChange(this.model), this.onModelConfigChange();
  }
  willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model")), e.has("selectedSeriesIndex") && this.selectedSeriesIndexChange();
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.removeSeriesPopover(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, s) {
    s?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  selectedSeriesIndexChange() {
    this.selectedSeriesIndex !== void 0 && (this.seriesPopover = L(o`<arcgis-charts-config-line-chart-series-popover .referenceElement=${this.el} .selectedName=${this.model.getSeriesName(this.selectedSeriesIndex)} .selectedColor=${this.model.getSeriesColor(this.selectedSeriesIndex)} .selectedLineStyle=${this.model.getSeriesLineStyle(this.selectedSeriesIndex)} .selectedLineWidth=${this.model.getSeriesLineWidth(this.selectedSeriesIndex)} .placement=${this.popoverPlacement} .open=${this.open} .selectedSeriesIndex=${this.selectedSeriesIndex} @arcgisChartsConfigSeriesPopoverNameChange=${this.seriesNameChange} @arcgisChartsConfigSeriesPopoverColorChange=${this.seriesColorChange} @arcgisChartsConfigSeriesPopoverLineStyleChange=${this.seriesLineStyleChange} @arcgisChartsConfigSeriesPopoverLineWidthChange=${this.seriesLineWidthChange} @arcgisChartsConfigPopoverClose=${this.seriesPopoverClose}></arcgis-charts-config-line-chart-series-popover>`), this.seriesPopover && !document.body.contains(this.seriesPopover) && document.body.appendChild(this.seriesPopover));
  }
  removeSeriesPopover() {
    this.seriesPopover !== void 0 && this.seriesPopover !== null && (this.seriesPopover.open = !1, this.seriesPopover = d(this.seriesPopover));
  }
  buildSeriesListItems() {
    const e = this.model.seriesLength, s = Array.from({ length: e }, (i, t) => t);
    return y(s, (i) => this.model.getSeriesId(i), (i) => {
      const t = this.model.getSeriesId(i), l = this.model.getSeriesColor(i), h = this.model.getSeriesName(i), c = this.model.chartSubType === I.BarAndLineFromFields ? this._messages.series : t, m = this.model.getSeriesLineStyle(i), g = M(m);
      let n = "#CCCCCC", a = "minus-circle";
      if ((e > 1 || !this.activeColorMatch) && l !== void 0) {
        a = `line-${g}`;
        const [p, C, S, v] = [...l];
        n = `rgba(${p},${C},${S},${v / 255})`;
      }
      return o`<calcite-list-item .label=${h ?? t ?? ""} .description=${c} .value=${t} .selected=${this.selectedSeriesIndex === i} @click=${this.openSeriesPopOver}><calcite-icon slot=content-end .icon=${a} style=${x({ color: n })}></calcite-icon></calcite-list-item>`;
    });
  }
  buildSeriesList() {
    return u(this.seriesOrderResetCounter, o`<calcite-list label drag-enabled selection-mode=single selection-appearance=border @calciteListOrderChange=${this.onCalciteListOrderChange}>${this.buildSeriesListItems()}</calcite-list>`);
  }
  async resetSeriesOriginal() {
    this.selectedSeriesIndex = void 0, this.model.resetSeriesStyling(), this.seriesOrderResetCounter += 1;
  }
  onCalciteListOrderChange(e) {
    if (e.detail !== null && e.detail !== void 0) {
      const { newIndex: s, oldIndex: r } = e.detail;
      this.model.moveSeries(r, s), this.seriesOrderResetCounter += 1;
    }
  }
  openSeriesPopOver(e) {
    e.stopImmediatePropagation(), this.removeSeriesPopover();
    const s = e.target, r = s.localName === "calcite-list-item" ? s : s.parentNode, i = this.model.getSeriesIndex(r.value);
    this.open = this.selectedSeriesIndex !== i, this.selectedSeriesIndex = this.selectedSeriesIndex !== i ? i : void 0;
  }
  render() {
    return o`<calcite-panel .heading=${this._messages.series}><calcite-button slot=header-actions-end @click=${this.resetSeriesOriginal} .disabled=${!this.model.resetAvailable()} appearance=transparent scale=s icon-end=reset>${this._messages.reset}</calcite-button>${this.buildSeriesList()}</calcite-panel>`;
  }
}
f("arcgis-charts-config-line-chart-series", k);
export {
  k as ArcgisChartsConfigLineChartSeries
};
