// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart/customElement.js
import { c as l } from "../../chunks/runtime.js";
import "@arcgis/toolkit/intl";
import { isEqual as r } from "lodash-es";
import "../../chunks/series-types.js";
import "d3-array";
import { g as n, ba as h, be as c } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { b as d, h as g } from "../../chunks/index4.js";
import "@arcgis/toolkit/dom";
import { LitElement as m, createEvent as f } from "@arcgis/lumina";
import { css as p, html as t } from "lit";
import { C as s } from "../../chunks/interfaces3.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const u = p`:host{margin-bottom:var(--arcgis-charts-intra-section-margin);margin-top:var(--arcgis-charts-intra-section-margin);height:100%}`;
class C extends m {
  constructor() {
    super(...arguments), this.onModelConfigChange = async (e) => {
      const i = e?.detail.config ?? this.model.config;
      this.layer === void 0 && await this.updateFetchLayerFieldsInfo(i);
    }, this.activeColorMatch = !1, this.popoverPlacement = "leading", this.arcgisChartsConfigNoSeriesUpdate = f();
  }
  static {
    this.properties = { layerFieldsInfo: 16, activeColorMatch: 16, model: 0, timeZone: 3, popoverPlacement: 3, seriesColor: 2, calculatedMinMaxBounds: 2, page: 3, layer: 0, usePopupTemplateFieldsInfo: 5 };
  }
  static {
    this.styles = u;
  }
  async load() {
    this.layer !== void 0 && await this.layerChange(this.layer, void 0), this.seriesColor !== void 0 && this.seriesColorChange(this.seriesColor), this.usePopupTemplateFieldsInfo !== void 0 && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, void 0), this.modelChange(this.model), this.onModelConfigChange(), this.updateDataFilters();
  }
  async willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model")), e.has("layer") && await this.layerChange(this.layer, e.get("layer")), e.has("usePopupTemplateFieldsInfo") && this.usePopupTemplateFieldsInfo !== void 0 && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, e.get("usePopupTemplateFieldsInfo")), e.has("seriesColor") && this.seriesColor !== void 0 && this.seriesColorChange(this.seriesColor), this.layerFieldsInfo === void 0 && await this.updateFetchLayerFieldsInfo();
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, a) {
    a?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  async layerChange(e, a) {
    e !== void 0 && !r(a, e) && await this.updateFetchLayerFieldsInfo();
  }
  async usePopupTemplateFieldsInfoChange(e, a) {
    e !== void 0 && e !== a && await this.updateFetchLayerFieldsInfo();
  }
  seriesColorChange(e) {
    this.activeColorMatch = e?.get(0) === void 0, this.model.colorMatch && e?.forEach((a, i) => {
      d(a) && e?.get(0) !== void 0 && a?.color !== void 0 && !r(this.model.getSeriesColor(i), a?.color) && this.model.setSeriesColor(a?.color, i);
    });
  }
  async updateDataFilters() {
    const e = this.model.getDataFilters(), a = e?.where, i = this.layer?.definitionExpression ?? void 0;
    r(a, i) || await this.model.setDataFilters({
      ...e,
      where: i
    });
  }
  async updateFetchLayerFieldsInfo(e) {
    const a = e ?? this.model.config;
    let i = this.layer;
    if (!i) {
      const o = g(a) ? a.iLayer : void 0;
      o && (i = await n(o));
    }
    i && (this.layerFieldsInfo = await h(i, this.usePopupTemplateFieldsInfo));
  }
  render() {
    const e = this.layerFieldsInfo?.find((o) => o.name === this.model.xAxisField), a = e !== void 0 && c(e);
    let i;
    switch (this.page) {
      case s.Data:
        i = t`<arcgis-charts-config-bar-chart-data .model=${this.model} .layerFieldsInfo=${this.layerFieldsInfo}></arcgis-charts-config-bar-chart-data>`;
        break;
      case s.Series:
        i = t`<arcgis-charts-config-bar-chart-series .model=${this.model} .activeColorMatch=${this.activeColorMatch} .popoverPlacement=${this.popoverPlacement}></arcgis-charts-config-bar-chart-series>`;
        break;
      case s.Axes:
        i = t`<arcgis-charts-config-axes .model=${this.model} show-max-y-bound-option .calculatedMinMaxBounds=${this.calculatedMinMaxBounds}></arcgis-charts-config-axes>`;
        break;
      case s.Guides:
        i = t`<arcgis-charts-config-guides .model=${this.model} .timeZone=${this.timeZone} .enableXAxisGuide=${a} enable-y-axis-guide .calculatedMinMaxBounds=${this.calculatedMinMaxBounds}></arcgis-charts-config-guides>`;
        break;
      case s.Format:
        i = t`<arcgis-charts-config-format .model=${this.model}></arcgis-charts-config-format>`;
        break;
      case s.General:
        i = t`<arcgis-charts-config-general .model=${this.model}></arcgis-charts-config-general>`;
        break;
      default:
        i = "";
        break;
    }
    return i;
  }
}
l("arcgis-charts-config-bar-chart", C);
export {
  C as ArcgisChartsConfigBarChart
};
