// @arcgis/charts-components/dist/components/arcgis-charts-config-guide/customElement.js
import { c as B } from "../../chunks/runtime.js";
import { keyed as X } from "lit/directives/keyed.js";
import "@arcgis/toolkit/intl";
import { cloneDeep as P } from "lodash-es";
import { W as z } from "../../chunks/series-types.js";
import { R as u, c as N } from "../../chunks/rest-js-object-literals.js";
import "d3-array";
import { j as p, ad as Z } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { c as g, b as j } from "../../chunks/index4.js";
import { c as O, a as E, b as F } from "../../chunks/index3.js";
import { r as V } from "../../chunks/popover-ui-utils.js";
import { m as $, d as W, g as q, j as L, c as G } from "../../chunks/chart-ui-utils2.js";
import { d as M } from "../../chunks/default.js";
import { g as U } from "../../chunks/resource.js";
import { u as Y } from "../../chunks/useT9n.js";
import { css as H, html as d } from "lit";
import { createRef as h, ref as m } from "lit/directives/ref.js";
import { LitElement as J, createEvent as K, renderElement as A, safeStyleMap as _, safeClassMap as o } from "@arcgis/lumina";
import { G as S } from "../../chunks/common.js";
import { j as Q, n as ee, o as te } from "../../chunks/type-guards.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const n = {
  row: "row",
  column: "column",
  colorPicker: "color-picker"
}, ie = H`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{display:flex;flex-direction:column}.row{display:flex;flex-direction:row;column-gap:var(--arcgis-charts-inter-section-margin)}.column{display:flex;flex-direction:column;flex-wrap:wrap;flex-grow:1}.column>calcite-input{margin-bottom:.5rem;flex-grow:2}.column>calcite-input[type=number]{margin-bottom:.5rem}input{padding:.5rem}.label{display:flex;flex-direction:row;justify-content:space-between;margin-bottom:var(--arcgis-charts-intra-section-margin);flex-grow:1}.color-picker{border:1px solid var(--calcite-color-border-2);box-shadow:none;margin-bottom:var(--arcgis-charts-inter-section-margin)}.delete-guide{display:flex;flex-direction:row-reverse;height:2rem;margin:0 1rem;font-weight:550}.delete-guide:dir(rtl){flex-direction:row}`;
class se extends J {
  constructor() {
    super(...arguments), this._messages = Y(), this.startCalciteInput = h(), this.endCalciteInput = h(), this.startDateCalciteInput = h(), this.endDateCalciteInput = h(), this.startTimeCalciteInput = h(), this.endTimeCalciteInput = h(), this.colorElement = h(), this.colorChange = (e) => {
      const t = e.detail, i = this.model.getGuideStart(this.index, this.orientation), a = this.model.getGuideEnd(this.index, this.orientation);
      if (p(i, a) === u.SLS) {
        const l = P(M.style), s = this.model.getGuideStyle(this.index, this.orientation), r = {
          ...l,
          ...g(s) ? s : {},
          color: t
        };
        this.lineStyleChange({ detail: r });
      } else
        this.setColor(t);
    }, this.symbolChange = (e) => {
      const t = e?.detail, i = this.model.getGuideStart(this.index, this.orientation), a = this.model.getGuideEnd(this.index, this.orientation);
      if (p(i, a) === u.SLS) {
        const l = P(M.style), s = this.model.getGuideStyle(this.index, this.orientation), r = {
          ...l,
          ...g(s) ? s : {},
          ...g(t) ? t : {}
        };
        this.lineStyleChange({ detail: r });
      } else t?.color !== void 0 && this.setColor(t.color);
    }, this.onModelConfigChange = () => {
      this.requestUpdate();
    }, this.guidePopover = null, this.latestGuidePosition = 0, this.arcgisChartsConfigGuideChange = K();
  }
  static {
    this.properties = { guidePopover: 16, messageOverrides: 0, model: 0, timeZone: 3, calculatedMinMaxBounds: 2, index: 11, orientation: 3, latestGuidePosition: 9 };
  }
  static {
    this.styles = ie;
  }
  get isXAxisDateField() {
    return this.orientation === S.X && Q(this.model) && this.model.getAxisValueFormat(0)?.type === z.DateAxisFormat;
  }
  load() {
    this.modelChange(this.model);
  }
  willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model"));
  }
  loaded() {
    this.isXAxisDateField && this.setGuideDateTimeRange();
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, t) {
    t?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  onColorElementSelect(e) {
    this.guidePopover?.open === !0 ? this.removeGuidePopover() : this.setupGuidePopover(e);
  }
  setGuideStart(e) {
    const { value: t } = e.target;
    this.startCalciteInput.value !== void 0 && (t === "" ? $({
      inputElement: this.startCalciteInput.value,
      status: G.INVALID
    }) : (this.model.setGuideStart(t, this.index, this.orientation), this.symbolChange(), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    }), $({
      inputElement: this.startCalciteInput.value,
      status: G.IDLE
    }))), this.guidePopover?.open === !0 && this.removeGuidePopover();
  }
  setGuideEnd(e) {
    const { value: t } = e.target;
    this.endCalciteInput.value !== void 0 && (this.model.setGuideEnd(t === "" ? null : t, this.index, this.orientation), this.symbolChange(), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    }), $({
      inputElement: this.endCalciteInput.value,
      status: G.IDLE
    })), this.guidePopover?.open === !0 && this.removeGuidePopover();
  }
  setGuideDateTimeRange() {
    const e = this.calculatedMinMaxBounds?.find((c) => c.isXAxis)?.minimum, t = this.calculatedMinMaxBounds?.find((c) => c.isXAxis)?.maximum, i = typeof e == "number" ? new Date(e) : /* @__PURE__ */ new Date(), a = typeof t == "number" ? new Date(t) : /* @__PURE__ */ new Date(), l = new window.Intl.DateTimeFormat().resolvedOptions().timeZone, s = O(this.timeZone) ?? l, r = E(i, s), f = F(i, s), v = E(a, s), R = F(a, s), T = this.startDateCalciteInput.value?.value ?? "", x = T === "" ? r : T, D = this.startTimeCalciteInput.value?.value ?? "", w = D === "" ? f : D, C = this.endDateCalciteInput.value?.value ?? "", y = C === "" ? v : C, I = this.endTimeCalciteInput.value?.value ?? "", k = I === "" ? R : I;
    let b = !1;
    if (!Array.isArray(x) && x !== void 0 && w !== void 0) {
      const c = `${x}T${w}`;
      this.model.setGuideStart(c, this.index, this.orientation), b = !0;
    }
    if (!Array.isArray(y) && y !== void 0 && k !== void 0) {
      const c = C === "" ? void 0 : `${y}T${k}`;
      this.model.setGuideEnd(c, this.index, this.orientation), b = !0;
    }
    b && (this.symbolChange(), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    }), this.guidePopover?.open === !0 && this.removeGuidePopover());
  }
  setGuideName(e) {
    const t = e.target.value;
    this.model.setGuideName(t, this.index, this.orientation), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    });
  }
  setGuideLabel(e) {
    const t = e.target.value;
    this.model.setGuideLabelText(t, this.index, this.orientation), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    });
  }
  setGuideRendered(e) {
    const t = e.target?.selectedItem?.value;
    if (t !== void 0) {
      const i = t === this._messages.inFront;
      this.model.setGuideAbove(i, this.index, this.orientation), this.arcgisChartsConfigGuideChange.emit({
        index: this.index,
        orientation: this.orientation,
        isDelete: !1
      });
    }
  }
  lineStyleChange(e) {
    const t = e.detail;
    this.model.setGuideStyle(t, this.index, this.orientation), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    });
  }
  setupGuidePopover(e) {
    e.stopImmediatePropagation();
    const t = this.model.getGuideStyle(this.index, this.orientation), i = this.model.getGuideStart(this.index, this.orientation), a = this.model.getGuideEnd(this.index, this.orientation);
    let l;
    p(i, a) === u.SLS ? l = g(t) ? t : void 0 : l = t?.color;
    const s = this.colorElement.value;
    Array.isArray(t?.color) ? (this.guidePopover = A(d`<arcgis-charts-config-guide-popover .referenceElement=${s} open .selectedElement=${l} @arcgisChartsConfigGuidePopoverColorChange=${this.colorChange} @arcgisChartsConfigGuidePopoverSymbolChange=${this.symbolChange} @arcgisChartsConfigPopoverClose=${this.removeGuidePopover}></arcgis-charts-config-guide-popover>`), this.guidePopover !== void 0 && !document.body.contains(this.guidePopover) && document.body.appendChild(this.guidePopover)) : this.removeGuidePopover();
  }
  removeGuidePopover() {
    this.guidePopover !== null && this.guidePopover !== void 0 && (this.guidePopover = V(this.guidePopover));
  }
  setColor(e) {
    const t = Z({ opacity: 0 }), i = this.model.getGuideStyle(this.index, this.orientation), a = {
      ...j(i) ? i : {},
      type: u.SFS,
      color: e,
      outline: t
    };
    this.model.setGuideStyle(a, this.index, this.orientation), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !1
    });
  }
  getDateStringFromBound(e) {
    return typeof e == "string" ? e.split("T")[0] : "";
  }
  getTimeStringFromBound(e) {
    return typeof e == "string" ? e.split("T")[1] : "";
  }
  createTooltip(e) {
    const t = e.target;
    let i = "";
    switch (t.icon) {
      case "view-visible":
        i = this._messages.hideGuide ?? "";
        break;
      case "view-hide":
        i = this._messages.showGuide ?? "";
        break;
      case "trash":
      default:
        i = this._messages.deleteGuide ?? "";
        break;
    }
    this.tooltip = A(d`<calcite-tooltip style=${_({ opacity: 1, zIndex: "var(--calcite-floating-ui-z-index)" })} .referenceElement=${t} open placement=top>${i}</calcite-tooltip>`), document.body.appendChild(this.tooltip);
  }
  toggleGuide() {
    const e = this.model.getGuideVisibility(this.index, this.orientation);
    this.model.setGuideVisibility(!e, this.index, this.orientation);
  }
  destroyTooltip() {
    W(this.tooltip);
  }
  deleteGuide() {
    this.destroyTooltip(), this.orientation === S.X && ee(this.model) ? this.model.deleteXAxisGuide(this.index) : this.orientation === S.Y && te(this.model) && this.model.deleteYAxisGuide(this.index), this.arcgisChartsConfigGuideChange.emit({
      index: this.index,
      orientation: this.orientation,
      isDelete: !0
    });
  }
  renderColorElement() {
    const e = this.model.getGuideStyle(this.index, this.orientation), [t, i, a, l] = [...e?.color ?? [1, 1, 1, 255]], s = this.model.getGuideStart(this.index, this.orientation), r = this.model.getGuideEnd(this.index, this.orientation), f = g(e) ? q(e.style) : N.Solid, v = p(s, r) === u.SLS ? `line-${f}` : "square-f";
    return d`<calcite-list label selection-mode=single selection-appearance=border><calcite-list-item .value=${this._messages.style} class=${o(n.colorPicker)} .label=${this._messages.style} .selected=${this.guidePopover?.open} @click=${this.onColorElementSelect} ${m(this.colorElement)}><calcite-icon slot=content-end .icon=${v} style=${_({ color: `rgba(${t},${i},${a},${l / 255})` })}></calcite-icon></calcite-list-item></calcite-list>`;
  }
  render() {
    const e = this.model.getGuideStart(this.index, this.orientation), t = this.model.getGuideEnd(this.index, this.orientation), i = this.model.getGuideName(this.index, this.orientation), a = this.model.getGuideLabelText(this.index, this.orientation), l = this.model.getGuideAbove(this.index, this.orientation), s = this.model.getGuideVisibility(this.index, this.orientation), r = this.isXAxisDateField;
    return d`<calcite-block .heading=${i ?? ""} .open=${this.latestGuidePosition === U(this.index, this.orientation, this.model)} collapsible .messageOverrides=${{
      expand: this._messages.expand,
      collapse: this._messages.collapse
    }}><calcite-action slot=control text .icon=${s ? "view-visible" : "view-hide"} @click=${this.toggleGuide} @mouseover=${this.createTooltip} @mouseout=${this.destroyTooltip}></calcite-action><calcite-action slot=control text icon=trash @click=${this.deleteGuide} @mouseover=${this.createTooltip} @mouseout=${this.destroyTooltip}></calcite-action><div class=${o(n.row)}><calcite-label class=${o(n.column)}>${this._messages.start}${r ? d`<calcite-input-date-picker .value=${this.getDateStringFromBound(e)} required scale=s placement=bottom-end @calciteInputDatePickerChange=${this.setGuideDateTimeRange} ${m(this.startDateCalciteInput)}></calcite-input-date-picker>` : d`<calcite-input type=number required @calciteInputInput=${this.setGuideStart} @keydown=${L} number-button-type=none .value=${e?.toString()} ${m(this.startCalciteInput)}></calcite-input>`}</calcite-label><calcite-label class=${o(n.column)}>${this._messages.end}${r ? d`<calcite-input-date-picker .value=${this.getDateStringFromBound(t)} scale=s placement=bottom-end @calciteInputDatePickerChange=${this.setGuideDateTimeRange} ${m(this.endDateCalciteInput)}></calcite-input-date-picker>` : d`<calcite-input type=number @calciteInputInput=${this.setGuideEnd} @keydown=${L} number-button-type=none .value=${t?.toString()} ${m(this.endCalciteInput)}></calcite-input>`}</calcite-label></div>${r ? X("time-picker-row", d`<div class=${o(n.row)}>${e != null ? d`<calcite-label class=${o(n.column)}><calcite-input-time-picker .value=${this.getTimeStringFromBound(e)} scale=s @calciteInputTimePickerChange=${this.setGuideDateTimeRange} ${m(this.startTimeCalciteInput)}></calcite-input-time-picker></calcite-label>` : null}${t != null ? d`<calcite-label class=${o(n.column)}><calcite-input-time-picker .value=${this.getTimeStringFromBound(t)} scale=s @calciteInputTimePickerChange=${this.setGuideDateTimeRange} ${m(this.endTimeCalciteInput)}></calcite-input-time-picker></calcite-label>` : null}</div>`) : null}<div class=${o(n.row)}><div class=${o(n.column)}>${this.renderColorElement()}</div></div><div class=${o(n.row)}><calcite-label class=${o(n.column)}>${this._messages.guideName}<calcite-input type=text .value=${i} @calciteInputInput=${this.setGuideName}></calcite-input></calcite-label></div><div class=${o(n.row)}><calcite-label class=${o(n.column)}>${this._messages.guideLabel}<calcite-input type=text .value=${a} @calciteInputInput=${this.setGuideLabel}></calcite-input></calcite-label></div><div class=${o(n.row)}><calcite-label class=${o(n.column)}>${this._messages.display}<calcite-segmented-control @calciteSegmentedControlChange=${this.setGuideRendered} scale=s><calcite-segmented-control-item .checked=${l} .value=${this._messages.inFront}>${this._messages.inFront}</calcite-segmented-control-item><calcite-segmented-control-item .checked=${!l} .value=${this._messages.inBack}>${this._messages.inBack}</calcite-segmented-control-item></calcite-segmented-control></calcite-label></div></calcite-block>`;
  }
}
B("arcgis-charts-config-guide", se);
export {
  se as ArcgisChartsConfigGuide
};
