// @arcgis/charts-components/dist/model.js
import { M as i } from "./chunks/common.js";
import { B as m, C as y, F as g, G as w, L as C, a as f, P as u, b, S as p, T as L, c as T } from "./chunks/common.js";
import { s as r, t as s, u as n } from "./chunks/type-guards.js";
import { a as B, e as P, i as S, g as x, b as W, r as v, h as G, l as K, m as D, p as F, j as H, k as O, q as R, n as A, o as E, d as X, f as Y, c as $ } from "./chunks/type-guards.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
async function c(t) {
  const o = "config" in t ? r(t.config) : t.chartType;
  let a;
  if (!("layer" in t) && !("iLayer" in t) && "config" in t && t.config.iLayer === void 0)
    throw new Error(
      "The source is missing. You can provide it to createModel through either `iLayer`, `layer`, or by setting the `config.iLayer` property."
    );
  switch (o) {
    case i.LineChart: {
      const { LineChartModel: e } = await import("./chunks/line-chart-model.js");
      a = new e();
      break;
    }
    case i.ComboBarLineChart: {
      const { ComboBarLineChartModel: e } = await import("./chunks/combo-bar-line-chart-model.js");
      a = new e();
      break;
    }
    case i.PieChart: {
      const { PieChartModel: e } = await import("./chunks/pie-chart-model.js");
      a = new e();
      break;
    }
    case i.Histogram: {
      const { HistogramModel: e } = await import("./chunks/histogram-model.js");
      a = new e();
      break;
    }
    case i.Scatterplot: {
      const { ScatterplotModel: e } = await import("./chunks/scatterplot-model.js");
      a = new e();
      break;
    }
    case i.BoxPlot: {
      const { BoxPlotModel: e } = await import("./chunks/box-plot-model.js");
      a = new e();
      break;
    }
    case i.Gauge: {
      const { GaugeModel: e } = await import("./chunks/gauge-model.js");
      a = new e();
      break;
    }
    case i.RadarChart: {
      const { RadarChartModel: e } = await import("./chunks/radar-chart-model.js");
      a = new e();
      break;
    }
    case i.HeatChart: {
      const { HeatChartModel: e } = await import("./chunks/heat-chart-model.js");
      a = new e();
      break;
    }
    case i.BarChart:
    default: {
      const { BarChartModel: e } = await import("./chunks/bar-chart-model.js");
      a = new e();
      break;
    }
  }
  if (await a?.setup({
    layer: "layer" in t ? t.layer : void 0,
    iLayer: "iLayer" in t ? t.iLayer : void 0,
    config: "config" in t ? t.config : void 0
  }), o !== void 0 && s(a, o))
    return a;
  {
    const e = n(a.chartType);
    throw new Error(
      `Please check the arguments provided to \`createModel\`. The generated model's type (\`${e}\`) does not match the type specified by either the \`chartType\` or the \`config\` argument (\`${o}\`).`
    );
  }
}
export {
  m as BoxPlotDataSortingKinds,
  y as ColorFormatKeys,
  g as FormatKeys,
  w as GuideOrientation,
  C as LineFormatKeys,
  i as ModelTypes,
  f as PieChartDataModes,
  u as PieChartDataSortingKinds,
  b as PieChartSlicesDisplay,
  p as SerialChartDataSortingKinds,
  L as TextFormatKeys,
  c as createModel,
  T as customSortString,
  B as isBarChartModel,
  P as isBoxPlotModel,
  S as isChartModel,
  x as isHistogramModel,
  W as isLineChartModel,
  v as isModelOfType,
  G as isModelWithBasicXAxis,
  K as isModelWithDataLabels,
  D as isModelWithGuides,
  F as isModelWithLegend,
  H as isModelWithMultiAxes,
  O as isModelWithRotation,
  R as isModelWithTemporalBinning,
  A as isModelWithXGuides,
  E as isModelWithYGuides,
  X as isPieChartModel,
  Y as isScatterplotModel,
  $ as isSerialChartModel
};
