// @arcgis/map-components/dist/chunks/measurementUtils.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const t = ["metric", "imperial"], r = [
  "inches",
  "feet",
  "yards",
  "miles",
  "nautical-miles",
  "us-feet",
  "millimeters",
  "centimeters",
  "decimeters",
  "meters",
  "kilometers"
], a = [
  "square-inches",
  "square-feet",
  "square-yards",
  "square-miles",
  "square-nautical-miles",
  "square-us-feet",
  "square-millimeters",
  "square-centimeters",
  "square-decimeters",
  "square-meters",
  "square-kilometers",
  "acres",
  "ares",
  "hectares"
], i = [...t, ...a], n = [...t, ...r];
function m(e, s = !0) {
  return s ? e == null ? { state: "unavailable" } : { state: "available", value: e } : { state: "invalid" };
}
function u(e, s) {
  switch (s.type) {
    case "length":
      return e.formatLength(s);
    case "area":
      return e.formatArea(s);
    default:
      return "";
  }
}
function c(e) {
  return e === "imperial" || e === "metric";
}
export {
  i as a,
  u as f,
  m as g,
  c as i,
  n as m
};
