// @arcgis/map-components/dist/chunks/utils.js
import { formatNumber as r } from "@arcgis/core/intl.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const p = /* @__PURE__ */ new Map([
  ["a3-landscape", "a3-landscape"],
  ["a3-portrait", "a3-portrait"],
  ["a4-landscape", "a4-landscape"],
  ["a4-portrait", "a4-portrait"],
  ["letter-ansi-a-landscape", "ansi-a-landscape"],
  ["letter-ansi-a-portrait", "ansi-a-portrait"],
  ["tabloid-ansi-b-landscape", "ansi-b-landscape"],
  ["tabloid-ansi-b-portrait", "ansi-b-portrait"]
]), u = /^https:\/\/([a-z\d-]+)(\.maps([^.]*))?\.arcgis\.com/iu;
function f(a) {
  if (!a)
    return;
  const t = a.layoutTemplateInfo;
  if (!t)
    return;
  const [e, n] = t.pageSize;
  return `${r(e)} × ${r(n)} ${a.pageUnits}`;
}
function d(a, t) {
  const { label: e, layout: n, layoutItem: i, type: o } = a;
  return o === "print-service-template" && t[n] ? t[n] : (e ?? n ?? i?.id)?.replaceAll("_", " ") ?? t.untitled;
}
function g(a) {
  return p.get(a);
}
const b = /* @__PURE__ */ new Map([
  ["AIX", "aix"],
  ["EPS", "eps"],
  ["GIF", "gif"],
  ["JPG", "jpg"],
  ["PDF", "pdf"],
  ["PNG32", "png32"],
  ["PNG8", "png8"],
  ["SVG", "svg"],
  ["SVGZ", "svgz"],
  ["TIFF", "tiff"]
]), s = "map-only";
function m(a) {
  return a?.toLowerCase() === s;
}
export {
  d as a,
  g as b,
  b as f,
  f as g,
  m as i,
  u as r
};
