// @arcgis/charts-components/dist/chunks/action-bar-utils.js
import { W as a, u as E } from "./series-types.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { cz as m, ck as h, p as A, cC as B, y } from "./interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import { A as u } from "./interfaces2.js";
import "@arcgis/toolkit/dom";
import { D as e, a as o, A as n } from "./index5.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const x = [
  e.EditChart,
  e.Legend,
  e.RotateChart,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsCSV,
  e.ExportAsImage
], f = [
  e.EditChart,
  e.Legend,
  e.RotateChart,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], b = [
  e.EditChart,
  e.Legend,
  e.RotateChart,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], F = [
  e.EditChart,
  e.Legend,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Selection,
  e.ClearSelection,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], D = [
  e.EditChart,
  e.Legend,
  e.FilterBySelection,
  e.Selection,
  e.ClearSelection,
  e.SwitchSelection
], I = [
  e.EditChart,
  e.Legend,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], T = [
  e.EditChart,
  e.Legend,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], L = [
  e.EditChart,
  e.Legend,
  e.RotateChart,
  e.FilterBySelection,
  e.FilterByExtent,
  e.Zoom,
  e.Selection,
  e.ClearSelection,
  e.FullExtent,
  e.SwitchSelection,
  e.ExportAsImage,
  e.ExportAsCSV
], v = [
  e.EditChart,
  e.FilterByExtent,
  e.ExportAsImage,
  e.ExportAsCSV
];
function Q(c) {
  switch (c) {
    case a.BarSeries:
      return x;
    case a.LineSeries:
      return f;
    case a.ComboLineAndBarSeries:
      return b;
    case a.PieSeries:
      return F;
    case a.HistogramSeries:
      return I;
    case a.ScatterSeries:
      return T;
    case a.BoxPlotSeries:
      return L;
    case a.GaugeSeries:
      return v;
    case a.HeatSeries:
      return D;
    default:
      return [];
  }
}
const U = (c, t, i) => {
  const l = c.target.assignedNodes().find((S) => "legendToggle" in S), s = E(i?.series);
  return l && (l.chartType = s, w({
    actionBar: l,
    chartComponent: t,
    config: i
  })), l;
};
function w(c) {
  const { actionBar: t, chartComponent: i, config: r, isWarningDisplayed: l } = c, s = m(i.selectionData), S = r?.legend ? r.legend.visible ?? !0 : !1;
  if (t.chartType = E(r?.series), i.enableConfiguration) {
    const C = t.hiddenActions ?? [];
    t.hiddenActions = C.filter((p) => p !== e.EditChart);
  }
  t.editChartState = i.enableConfiguration ? o.Enabled : o.Disabled;
  const d = r && h(r.series);
  t.legendState = d ? o.Enabled : o.Disabled, t.legendToggle = d && (i.legendVisibility ?? S) ? n.Active : n.Inactive, t.rotateChartToggle = (i.rotation ?? r?.rotated) === !0 ? n.Active : n.Inactive;
  const g = A(r?.series ?? []) === B.PieFromFields;
  t.filterBySelectionState = !g && (i.filterBySelection === !0 || !s && i.filterByExtent !== !0) ? o.Enabled : o.Disabled, t.filterBySelectionToggle = i.filterBySelection === !0 ? n.Active : n.Inactive, t.filterByExtentState = i.filterByExtent === !0 || i.filterBySelection !== !0 ? o.Enabled : o.Disabled, t.filterByExtentToggle = i.filterByExtent === !0 ? n.Active : n.Inactive, t.zoomState = i.filterBySelection ? o.Disabled : o.Enabled, t.zoomToggle = i.actionMode === u.Zoom ? n.Active : n.Inactive, t.selectionState = i.filterBySelection ? o.Disabled : o.Enabled, t.selectionToggle = (i.actionMode ?? t.actionBarSelectionMode) === t.actionBarSelectionMode ? n.Active : n.Inactive, t.clearSelectionState = s ? o.Disabled : o.Enabled, t.fullExtentState = i.filterBySelection ? o.Disabled : o.Enabled, t.switchSelectionState = i.filterBySelection ? o.Disabled : o.Enabled, t.exportActionState = l || y(r) ? o.Disabled : o.Enabled;
}
export {
  Q as g,
  U as o,
  w as u
};
