// @arcgis/charts-components/dist/chunks/axis-model-wrapper.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
class e {
  /**
   * `AxisModelWrapper` constructor.
   *
   * @param model The wrapped chart model.
   * @param isXAxis Whether the wrapper is for X-axis or Y-axis.
   */
  constructor(s, i = !1) {
    this.model = s, this.isXAxis = i;
  }
  getMaxBound() {
    return this.model?.getMaxBound(this.isXAxis ? 0 : 1) ?? Number.NaN;
  }
  setMaxBound(s) {
    this.model?.setMaxBound(s, this.isXAxis ? 0 : 1);
  }
  getMinBound() {
    return this.model?.getMinBound(this.isXAxis ? 0 : 1) ?? Number.NaN;
  }
  setMinBound(s) {
    this.model?.setMinBound(s, this.isXAxis ? 0 : 1);
  }
  getValueFormat() {
    return this.model?.getAxisValueFormat(this.isXAxis ? 0 : 1);
  }
  setValueFormat(s) {
    this.model?.setAxisValueFormat(this.isXAxis ? 0 : 1, s);
  }
  getLogarithmic() {
    return this.model?.getLogarithmic(this.isXAxis ? 0 : 1) ?? !1;
  }
  setLogarithmic(s) {
    this.model?.setLogarithmic(s, this.isXAxis ? 0 : 1);
  }
}
export {
  e as A
};
