// @arcgis/charts-components/dist/chunks/chart-ui-utils2.js
import { a as r } from "./series-types.js";
import { c as l } from "./rest-js-object-literals.js";
import { ct as u, b as d } from "./interfaces.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import { d as f } from "./index3.js";
import { F as o } from "./interfaces3.js";
import { F as a } from "./common.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const x = {
  DebounceTimer: 400,
  SliderTimer: 100,
  PopoverTimer: 60,
  customSortLimit: 100
}, C = {
  style: "decimal",
  minimumFractionDigits: 0,
  maximumFractionDigits: 3
}, w = [
  r.Count,
  r.Sum,
  r.Average,
  r.ContinuousPercentile,
  r.Minimum,
  r.Maximum,
  r.NoAggregation
], P = [r.Count, r.Sum];
function U(e) {
  let t;
  switch (e) {
    case a.Title:
    case a.XAxisTitle:
    case a.YAxisTitle:
    case a.LegendTitle: {
      t = o.Title;
      break;
    }
    case a.Description:
    case a.LegendText: {
      t = o.Text;
      break;
    }
    case a.AxisLabels:
    case a.GuideLabels:
    case a.DataLabels: {
      t = o.Label;
      break;
    }
    case a.Background:
    case a.LeaderLines:
    case a.AxisLines:
    case a.GridLines: {
      t = o.Palette;
      break;
    }
  }
  return t;
}
function K(e, t, i) {
  let s = !1;
  const n = typeof e == "string" ? Number.parseFloat(e) : e, m = typeof t == "string" ? Number.parseFloat(t) : t, c = typeof i == "string" ? Number.parseFloat(i) : i;
  return Number.isNaN(n) || (s = (m === void 0 || Number.isNaN(m) || n >= m) && (c === void 0 || Number.isNaN(c) || n <= c)), s;
}
function M(e, t) {
  let i;
  if (e !== null) {
    const s = typeof e == "string" ? Number.parseFloat(e) : e;
    u(s) ? i = s : u(t) && (i = t);
  }
  return i === void 0 ? "" : `${i}`;
}
function z(e) {
  const { inputElement: t, status: i } = e;
  t.status = i;
}
const B = {
  IDLE: "idle",
  INVALID: "invalid"
};
function G(e, t) {
  let i = "";
  if (e !== void 0 && !Number.isNaN(e)) {
    const s = f(document.documentElement);
    i = Intl.NumberFormat(s, t).format(e);
  }
  return i;
}
function O(e, t) {
  let i = !1;
  if (t !== void 0) {
    const s = typeof t == "string" ? t.trim() : t, n = e.trim();
    typeof t == "string" ? i = s === n || s === d && n === "" : i = s.includes(n);
  }
  return i;
}
function R(e) {
  let t;
  switch (e) {
    case "trailing":
      t = -200;
      break;
    case "leading":
      t = 20;
      break;
    default:
      t = 0;
      break;
  }
  return t;
}
function j(e) {
  return e === void 0 || Number.isNaN(e.r) || Number.isNaN(e.g) || Number.isNaN(e.b) ? null : e;
}
const W = (e) => {
  ["e", "E"].includes(e.key) && e.preventDefault();
}, X = (e) => {
  if (e.pickListElement !== void 0) {
    const t = e.pickListElement.selectedItems;
    e.eventEmitter.emit({
      contentKind: e.contentKind,
      value: t.map((i) => i.value)
    });
  }
};
function Y(e) {
  let t = "solid";
  switch (e) {
    case l.Dash:
      t = "dashed";
      break;
    case l.Dot:
      t = "dotted";
      break;
    case l.Solid:
    default:
      t = "solid";
      break;
  }
  return t;
}
function $(e, t) {
  const i = document.createElement("a");
  i.style.display = "none", i.download = t, i.setAttribute("href", e), i.click(), i.remove(), URL.revokeObjectURL(e);
}
function q(e) {
  return e != null && (e.style.removeProperty("opacity"), e.remove()), null;
}
export {
  C as D,
  x as U,
  R as a,
  K as b,
  B as c,
  q as d,
  X as e,
  G as f,
  Y as g,
  w as h,
  O as i,
  W as j,
  j as k,
  U as l,
  z as m,
  $ as n,
  P as p,
  M as s
};
