// @arcgis/charts-components/dist/chunks/common.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const e = "customSort", s = {
  xAxisAsc: "xAxisAsc",
  xAxisDesc: "xAxisDesc",
  yAxisAsc: "yAxisAsc",
  yAxisDesc: "yAxisDesc",
  customSort: e
}, a = {
  xAxisAsc: "xAxisAsc",
  xAxisDesc: "xAxisDesc",
  meanAsc: "meanAsc",
  meanDesc: "meanDesc",
  medianAsc: "medianAsc",
  medianDesc: "medianDesc",
  customSort: e
}, t = {
  labelAsc: "labelAsc",
  labelDesc: "labelDesc",
  valueAsc: "valueAsc",
  valueDesc: "valueDesc",
  customSort: e
}, i = {
  Category: "category",
  Fields: "fields"
}, n = {
  Category: "category",
  Value: "value",
  Percentage: "percentage",
  Both: "both"
}, r = {
  TextElements: "textElements",
  XAxisTitle: "xAxisTitle",
  YAxisTitle: "yAxisTitle",
  AxisLabels: "axisLabels",
  GuideLabels: "guideLabels",
  Title: "title",
  LegendText: "legendText",
  LegendTitle: "legendTitle",
  Description: "description",
  DataLabels: "dataLabels",
  Background: "background",
  LeaderLines: "leaderLines",
  AxisLines: "axisLines",
  GridLines: "gridLines"
}, l = {
  AxisLines: "axisLines",
  GridLines: "gridLines",
  LeaderLines: "leaderLines"
}, o = {
  Title: "title",
  XAxisTitle: "xAxisTitle",
  YAxisTitle: "yAxisTitle",
  AxisLabels: "axisLabels",
  GuideLabels: "guideLabels",
  LegendText: "legendText",
  LegendTitle: "legendTitle",
  Description: "description",
  DataLabels: "dataLabels"
}, c = {
  Background: "background"
}, x = {
  X: "x",
  Y: "y"
}, d = {
  BarChart: "barChart",
  LineChart: "lineChart",
  ComboBarLineChart: "comboBarLineChart",
  BoxPlot: "boxPlot",
  PieChart: "pieChart",
  Scatterplot: "scatterplot",
  Histogram: "histogram",
  Gauge: "gauge",
  RadarChart: "radarChart",
  HeatChart: "heatChart"
};
export {
  a as B,
  c as C,
  r as F,
  x as G,
  l as L,
  d as M,
  t as P,
  s as S,
  o as T,
  i as a,
  n as b,
  e as c
};
