// @arcgis/charts-components/dist/chunks/histogram-model.js
import "@arcgis/toolkit/intl";
import "lodash-es";
import { W as e } from "./series-types.js";
import "d3-array";
import { ci as i, aU as s } from "./interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import "@arcgis/toolkit/dom";
import { d as r } from "./data2.js";
import { g as a, s as o, a as n, b as m, c as h, d as g, e as l, f as d, h as y, i as c, j as f, k as S, l as v, m as u, n as p, o as w, p as O, q as b, r as C, t as M, u as D, v as _, w as F, x as A, y as T, z as U, A as N } from "./histogram-data.js";
import { C as W, M as x, a as E } from "./model-with-multi-axes.js";
import { M as I, a as B, b as H, c as L } from "./model-with-data-labels.js";
import "@arcgis/core/intl.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
class j extends W {
  constructor() {
    super(...arguments), this._chartType = e.HistogramSeries;
  }
  async setup(t) {
    await this.loadConfigFromLayer(t), typeof this.resolveSetup == "function" && this.resolveSetup();
  }
  // ---------------------------------------------------------------------------------------------
  // #region Data Props
  // ---------------------------------------------------------------------------------------------
  /**
   * Numeric field used to create the histogram to visualize the frequency distribution of its values.
   */
  get numericField() {
    return a(this._config);
  }
  set numericField(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      o({
        numericField: t,
        config: this._config,
        layerInfo: this.layerInfo,
        commonStrings: this.commonStrings
      }), this.dispatchEvent(new CustomEvent("arcgisChartsModelDataChange")), this.emitConfigUpdate("numericField");
    });
  }
  /**
   * Number of bins used to create the histogram.
   */
  get binCount() {
    return n(this._config);
  }
  set binCount(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      m(t, this._config), this.dispatchEvent(new CustomEvent("arcgisChartsModelDataChange")), this.emitConfigUpdate("binCount");
    });
  }
  /**
   * Fill symbol used for the bins in the histogram.
   */
  get binSymbol() {
    return h(this._config);
  }
  set binSymbol(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      g(t, this._config), this.emitConfigUpdate("binSymbol");
    });
  }
  /**
   * Data transformation type used in the histogram.
   */
  get dataTransformationType() {
    return l(this._config);
  }
  set dataTransformationType(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      d(t, this._config), this.dispatchEvent(new CustomEvent("arcgisChartsModelDataChange")), this.emitConfigUpdate("dataTransformationType");
    });
  }
  /**
   * Line symbol used for the mean overlay in the histogram.
   */
  get meanSymbol() {
    return y(this._config);
  }
  set meanSymbol(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      c(t, this._config), this.emitConfigUpdate("meanSymbol");
    });
  }
  /**
   * Line symbol used for the median overlay in the histogram.
   */
  get medianSymbol() {
    return f(this._config);
  }
  set medianSymbol(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      S(t, this._config), this.emitConfigUpdate("medianSymbol");
    });
  }
  /**
   * Line symbol used for the normal distribution overlay in the histogram.
   */
  get normalDistSymbol() {
    return v(this._config);
  }
  set normalDistSymbol(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      u(t, this._config), this.emitConfigUpdate("normalDistSymbol");
    });
  }
  /**
   * Line symbol used for the standard deviation overlay in the histogram.
   */
  get standardDevSymbol() {
    return p(this._config);
  }
  set standardDevSymbol(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      w(t, this._config), this.emitConfigUpdate("standardDevSymbol");
    });
  }
  /**
   * Whether the normal distribution overlay is shown in the histogram.
   */
  get showNormalDistOverlay() {
    return O(this._config);
  }
  set showNormalDistOverlay(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      b(t, this._config), this.emitConfigUpdate("showNormalDistOverlay");
    });
  }
  /**
   * Whether the mean overlay is shown in the histogram.
   */
  get showMeanOverlay() {
    return C(this._config);
  }
  set showMeanOverlay(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      M(t, this._config), this.emitConfigUpdate("showMeanOverlay");
    });
  }
  /**
   * Whether the median overlay is shown in the histogram.
   */
  get showMedianOverlay() {
    return D(this._config);
  }
  set showMedianOverlay(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      _(t, this._config), this.emitConfigUpdate("showMedianOverlay");
    });
  }
  /**
   * Whether the standard deviation overlay is shown in the histogram.
   */
  get showStandardDevOverlay() {
    return F(this._config);
  }
  set showStandardDevOverlay(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      A(t, this._config), this.emitConfigUpdate("showStandardDevOverlay");
    });
  }
  /**
   * Chart overlays for the series. Histogram supports 4 overlays: **mean**, **median**, **standardDeviation** and **comparisonDistribution**.
   */
  get seriesOverlays() {
    return T(this._config);
  }
  set seriesOverlays(t) {
    this.guardSetterFromAsyncOverwrite(() => {
      U(t, this._config), this.emitConfigUpdate("seriesOverlays");
    });
  }
  // ---------------------------------------------------------------------------------------------
  // #region General methods
  // ---------------------------------------------------------------------------------------------
  /**
   * Generate the proper chart title given the current config.
   *
   * @returns {string}
   */
  generateChartTitle() {
    return N(this._config, this.layerInfo, this.commonStrings);
  }
  /**
   * Generate the proper x-axis title given the current config.
   *
   * @returns {string}
   */
  generateXAxisTitle() {
    return i(this.layerInfo.layerFieldsInfo, this.numericField);
  }
  // ---------------------------------------------------------------------------------------------
  // #region Protected methods
  // ---------------------------------------------------------------------------------------------
  generateDefaultConfig() {
    return s(this.commonStrings);
  }
}
r(j, [
  x,
  I,
  B,
  H,
  L,
  E
]);
export {
  j as HistogramModel
};
