// @arcgis/charts-components/dist/chunks/histogram-data.js
import { b as f } from "./series-types.js";
import { substitute as O } from "@arcgis/core/intl.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { aT as i, ad as n, ci as m, y as M, aR as y, al as r } from "./interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import "@arcgis/toolkit/dom";
import { g as C, a as A, u as h, s as l } from "./data2.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const v = (s) => {
  l({ config: s, value: null, axisIndex: 0, isMax: !1 }), l({ config: s, value: null, axisIndex: 0, isMax: !0 }), l({ config: s, value: null, axisIndex: 1, isMax: !1 }), l({ config: s, value: null, axisIndex: 1, isMax: !0 });
};
function u(s, e, a) {
  const { layerFieldsInfo: t } = e, o = m(t, p(s));
  return O(a.titleLabels.distributionOfField, { valueField: o });
}
const p = (s) => s.series[0]?.x ?? "", U = (s) => {
  const { numericField: e, config: a, layerInfo: t, commonStrings: o } = s, { layerFieldsInfo: d } = t, c = u(a, t, o), D = C(a), S = A(0, a), x = m(d, p(a)), b = M(a);
  if (v(a), a?.series?.[0] && (a.series[0].x = e), x === S || b) {
    const T = m(d, e);
    h(a, 0, T);
  }
  a.title && (c === D || b) && (a.title.content.text = u(a, t, o));
}, V = (s) => s.series[0]?.dataTransformationType ?? f.None, Y = (s, e) => {
  v(e), e?.series?.[0] && (e.series[0].dataTransformationType = s);
}, Z = (s) => s.series[0]?.binCount ?? 32, _ = (s, e) => {
  v(e), e?.series?.[0] && (e.series[0].binCount = s);
}, $ = (s) => s.series[0]?.fillSymbol ?? i().fillSymbol, g = (s, e) => {
  e?.series?.[0] && (e.series[0].fillSymbol = {
    ...e.series[0].fillSymbol,
    ...s
  });
}, F = (s) => s.series[0]?.overlays?.comparisonDistribution?.symbol ?? i().overlays.comparisonDistribution?.symbol ?? n(), ss = (s, e) => {
  e.series[0].overlays ??= y(), e.series[0].overlays.comparisonDistribution ??= r(4), e.series[0].overlays.comparisonDistribution.symbol = {
    ...e.series[0].overlays.comparisonDistribution.symbol,
    ...s
  };
}, I = (s) => s.series[0]?.overlays?.mean?.symbol ?? i().overlays.mean?.symbol ?? n(), es = (s, e) => {
  e.series[0].overlays ??= y(), e.series[0].overlays.mean ??= r(1), e.series[0].overlays.mean.symbol = {
    ...e.series[0].overlays.mean.symbol,
    ...s
  };
}, N = (s) => s.series[0]?.overlays?.median?.symbol ?? i().overlays.median?.symbol ?? n(), as = (s, e) => {
  e.series[0].overlays ??= y(), e.series[0].overlays.median ??= r(2), e.series[0].overlays.median.symbol = {
    ...e.series[0].overlays.median.symbol,
    ...s
  };
}, B = (s) => s.series[0]?.overlays?.standardDeviation?.symbol ?? i().overlays.standardDeviation?.symbol ?? n(), rs = (s, e) => {
  e.series[0].overlays ??= y(), e.series[0].overlays.standardDeviation ??= r(3), e.series[0].overlays.standardDeviation.symbol = {
    ...e.series[0].overlays.standardDeviation.symbol,
    ...s
  };
}, ts = (s) => (s.series[0]?.overlays?.comparisonDistribution?.created ?? !1) && (s.series[0]?.overlays?.comparisonDistribution?.visible ?? !1), is = (s, e) => {
  const a = r(0);
  a.created = s, a.visible = s, a.symbol = F(e), e?.series?.[0]?.overlays && (e.series[0].overlays.comparisonDistribution = a);
}, os = (s) => (s.series[0]?.overlays?.mean?.created ?? !1) && (s.series[0]?.overlays?.mean?.visible ?? !1), ls = (s, e) => {
  const a = r(1);
  a.created = s, a.visible = s, a.symbol = I(e), e?.series?.[0]?.overlays && (e.series[0].overlays.mean = a);
}, ns = (s) => (s.series[0]?.overlays?.median?.created ?? !1) && (s.series[0]?.overlays?.median?.visible ?? !1), ys = (s, e) => {
  const a = r(2);
  a.created = s, a.visible = s, a.symbol = N(e), e?.series?.[0]?.overlays && (e.series[0].overlays.median = a);
}, ms = (s) => (s.series[0]?.overlays?.standardDeviation?.created ?? !1) && (s.series[0]?.overlays?.standardDeviation?.visible ?? !1), vs = (s, e) => {
  const a = r(3);
  a.created = s, a.visible = s, a.symbol = B(e), e?.series?.[0]?.overlays && (e.series[0].overlays.standardDeviation = a);
}, ds = (s) => s?.series?.[0]?.overlays, bs = (s, e) => {
  e?.series?.[0] !== void 0 && (e.series[0].overlays = s);
};
export {
  u as A,
  Z as a,
  _ as b,
  $ as c,
  g as d,
  V as e,
  Y as f,
  p as g,
  I as h,
  es as i,
  N as j,
  as as k,
  F as l,
  ss as m,
  B as n,
  rs as o,
  ts as p,
  is as q,
  os as r,
  U as s,
  ls as t,
  ns as u,
  ys as v,
  ms as w,
  vs as x,
  ds as y,
  bs as z
};
