// @arcgis/charts-components/dist/chunks/index5.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const t = {
  DataProcessComplete: "arcgisDataProcessComplete",
  DataFetchComplete: "arcgisDataFetchComplete",
  DataProcessError: "arcgisDataProcessError",
  LegendItemVisibilityChange: "arcgisLegendItemVisibilityChange",
  UpdateComplete: "arcgisUpdateComplete",
  SeriesColorChange: "arcgisSeriesColorChange",
  AxesMinMaxChange: "arcgisAxesMinMaxChange",
  DefaultActionSelect: "arcgisDefaultActionSelect",
  SelectionComplete: "arcgisSelectionComplete",
  ActionBarToggle: "arcgisActionBarToggle",
  NoRenderPropChange: "arcgisNoRenderPropChange",
  BadDataWarningRaise: "arcgisBadDataWarningRaise",
  SeriesOrder: "arcgisSeriesOrder",
  RuntimeError: "arcgisRuntimeError",
  ConfigChange: "arcgisConfigChange"
}, a = {
  FilterBySelection: "filterBySelection",
  FilterByExtent: "filterByExtent",
  Legend: "legend",
  RotateChart: "rotateChart",
  Selection: "selection",
  ClearSelection: "clearSelection",
  SwitchSelection: "switchSelection",
  Zoom: "zoom",
  FullExtent: "fullExtent",
  EditChart: "editChart",
  ExportAsImage: "exportAsImage",
  ExportAsCSV: "exportAsCSV"
}, r = {
  Active: "active",
  Inactive: "inactive"
}, i = {
  Enabled: "enabled",
  Disabled: "disabled"
}, o = {
  Ignore: "ignore",
  Throw: "throw"
}, e = {
  Ignore: "ignore",
  Refresh: "refresh"
}, n = e, s = e, c = e, l = e, g = {
  /**
   * Enable syncing selections between chart and layer view
   */
  Enabled: "enabled",
  /**
   * Disable syncing selections between chart and layer view
   */
  Disabled: "disabled"
};
export {
  r as A,
  t as C,
  a as D,
  o as E,
  n as L,
  g as S,
  l as V,
  i as a,
  s as b,
  c
};
