// @arcgis/charts-components/dist/chunks/index3.js
import { unsafeGetElementDirection as c } from "@arcgis/toolkit/dom";
import { getElementLocale as r, supportedLocales as i } from "@arcgis/toolkit/intl";
import { g as m } from "./runtime.js";
import { ca as l } from "./interfaces.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
async function a(t) {
  const e = m(`assets/common/t9n/messages.${t}.json`), n = await fetch(e);
  if (!n.ok)
    throw new Error(`Failed to load t9n file: ${e}`);
  return await n.json();
}
function F(t) {
  const { t9nLocale: e } = r(t ?? document.documentElement), n = f(t ?? document.documentElement), o = c(t ?? document.documentElement) === "rtl";
  return { locale: e, formatLocale: n, rtl: o };
}
function L(t) {
  let e;
  return t === "unknown" ? e = "UTC" : t !== void 0 && t !== "system" && (e = t), e;
}
function h(t, e) {
  const n = {
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    timeZone: e
  };
  return new Intl.DateTimeFormat("en-CA", n).format(t);
}
function w(t, e) {
  const n = {
    hour: "2-digit",
    minute: "2-digit",
    hour12: !1,
    timeZone: e
  };
  return new Intl.DateTimeFormat("en-US", n).format(t);
}
function y(t, e) {
  const [, n, , , , o] = 1111.1.toLocaleString(e), s = t.replace(n, "").replace(o, ".");
  return Number.parseFloat(s);
}
function u(t) {
  const e = Array.from(i);
  return l.includes(t) && !e.includes(t);
}
function f(t) {
  const { lang: e, t9nLocale: n } = r(t);
  return u(e) ? e : n;
}
async function E() {
  return (await a("en")).errorStrings;
}
async function T() {
  return (await a("en")).utilsStrings;
}
export {
  h as a,
  w as b,
  L as c,
  f as d,
  E as e,
  T as f,
  F as g,
  a as h,
  y as l
};
