// @arcgis/charts-components/dist/chunks/interfaces4.js
import { g as n, e as l, h as o, d } from "./series-types.js";
import { c as r, R as a, d as s, f as t, e } from "./rest-js-object-literals.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { a2 as g, a$ as c, Y as u } from "./interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import { A as h } from "./interfaces2.js";
import "@arcgis/toolkit/dom";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const i = [150, 150, 150, 255], P = {
  aria: {
    defaultChartItemRole: "figure"
  },
  title: {
    marginBottom: 8,
    wrap: !0
  },
  subtitle: {
    marginBottom: 10,
    wrap: !0
  },
  rotated: !1,
  legendProperties: {
    legendLabels: {
      marginLeft: 5,
      marginRight: 5,
      truncate: !1,
      applyOnClones: !0,
      wrap: !0,
      maxWidth: 200
    },
    valueLabels: {
      wrap: !0,
      maxWidth: 200
    },
    legendTitle: {
      wrap: !0,
      maxWidth: 220,
      marginLeft: 10,
      marginRight: 10,
      marginTop: 10
    },
    legendMarkersSize: 14,
    legendLabelReducedOpacity: 0.5,
    // 50% of original opacity
    rectangleMarkerRadius: 3,
    // for rounded rectangles
    roundedMarkerRadius: 12,
    // for circles
    legendPaddingRightWithScrollbarVisible: 10,
    maxHeight: 100,
    marginTop: 2,
    marginBottom: 2,
    marginLeft: 0,
    marginRight: 0,
    paddingTop: 0,
    paddingBottom: 0,
    paddingLeft: 0,
    paddingRight: 0,
    position: d.Right
  },
  footer: {
    wrap: !0,
    horizontalCenter: e.Center,
    verticalCenter: t.Middle,
    align: e.Center
  },
  axisLabel: {
    rotation: 0,
    align: e.Center,
    horizontalCenter: e.Center,
    verticalCenter: t.Top,
    dy: 5,
    paddingTop: 5,
    paddingRight: 5,
    paddingLeft: 5,
    maxWidth: Number.NaN
  },
  axisTickLength: 5,
  rotatedAxisLabel: {
    rotation: -45,
    horizontalCenter: e.Left,
    verticalCenter: t.Bottom,
    dy: 6,
    paddingTop: 5,
    paddingLeft: 0,
    paddingRight: 5
  },
  histogramAxisLabel: {
    rotation: 0,
    horizontalCenter: e.Center,
    verticalCenter: t.Middle,
    dy: 0,
    dx: 0,
    paddingLeft: 3,
    paddingRight: 3
  },
  xAxisTitleLabel: {
    align: e.Center,
    valign: t.Middle,
    rotation: 0,
    fontSize: 10,
    wrap: !0
  },
  yAxisTitleLabel: {
    align: e.Center,
    valign: t.Middle,
    rotation: 270,
    wrap: !0
  },
  titleLabel: {
    align: e.Center,
    wrap: !0
  },
  subtitleLabel: {
    align: e.Center,
    wrap: !0
  },
  horizontalSurfaceGuideLabels: {
    horizontalCenter: e.Right,
    align: e.Right,
    verticalCenter: t.Middle,
    wrap: !0
  },
  horizontalLineGuideLabels: {
    horizontalCenter: e.Right,
    align: e.Right,
    verticalCenter: t.Top,
    wrap: !0
  },
  verticalSurfaceGuideLabels: {
    horizontalCenter: e.Center,
    verticalCenter: t.Top,
    valign: t.Top
  },
  verticalLineGuideLabels: {
    horizontalCenter: e.Right,
    verticalCenter: t.Bottom,
    valign: t.Top,
    wrap: !0,
    rotation: 90
  },
  guideAbove: !1,
  cursorCrosshair: {
    style: {
      type: a.SLS,
      style: r.Dash,
      color: [0, 0, 0, 102],
      width: 1
    },
    horizontalLineVisible: !1,
    verticalLineVisible: !1
  },
  gauge: {
    innerRadius: 80,
    radius: 80,
    startAngle: -180,
    endAngle: 180,
    innerLabelToInnerChartRatio: 0.5,
    // 50% of the chart's size, considering the inner radius
    axisLabelToInnerChartRatio: 0.13,
    // ~= 40 / 300: font size 40 for a 300px gauge
    innerLabelAutoMinFontSize: 14,
    innerLabelAutoMaxFontSize: 100,
    axisLabelAutoMinFontSize: 14,
    axisLabelAutoMaxFontSize: 40,
    gaugeMinGridDistance: 1,
    axisMinimum: 0,
    axisMaximum: 100,
    numberOfLabelIntervals: 5,
    numberOfTicksBetweenTwoLabels: 4,
    maximumNumberOfLabelIntervals: 24,
    // result in 25 maximum number of axis labels
    paddingTop: 10,
    paddingBottom: 10,
    paddingRight: 5,
    paddingLeft: 5
  },
  gaugeTicksLength: 7,
  gaugeInnerLabel: {
    horizontalCenter: e.Center,
    verticalCenter: t.Middle
  },
  gaugeNeedle: {
    startWidth: 10,
    endWidth: 1,
    innerRadius: 25,
    displayPin: !1
  },
  gaugeFirstLabelId: "arcgis_charts_gauge_first_label",
  gaugeLastLabelId: "arcgis_charts_gauge_last_label",
  // containers
  headerContainer: {
    layout: "vertical",
    width: "100%",
    paddingBottom: 5
  },
  footerContainer: {
    layout: "vertical",
    width: "100%",
    paddingTop: 5,
    paddingBottom: 15
  },
  paddingMinSize: 0,
  marginMinSize: 0,
  chartPaddingMinSize: 7,
  chartPaddingRightWhenLegendNotOnRight: 15,
  chartAndLegendContainer: {
    paddingTop: 0
  },
  chartContainer: {
    paddingBottom: 0,
    paddingTop: 15
  },
  chartContainerPaddingTopWhenRotated: 5,
  dataLabelInitialize: {
    strokeWidth: 0,
    applyOnClones: !0
  },
  dataLabel: {
    align: e.Center,
    horizontalCenter: e.Center,
    valign: t.Bottom,
    verticalCenter: t.Top,
    dy: -3
  },
  dataLabelRotated: {
    align: e.Right,
    horizontalCenter: e.Right,
    valign: t.Middle,
    verticalCenter: t.Middle,
    dx: 2
  },
  dataLabelStackedRotated: {
    align: e.Center,
    horizontalCenter: e.Center,
    valign: t.Middle,
    verticalCenter: t.Middle
  },
  dataLabelStackedNonRotated: {
    align: e.Center,
    horizontalCenter: e.Center,
    valign: t.Middle,
    verticalCenter: t.Middle
  },
  dataLabelNegative: {
    paddingTop: 5,
    paddingRight: 5
  },
  rotatedChartPadding: {
    paddingTop: 2,
    paddingRight: 2
  },
  scatterplotSeries: {
    strokeOpacity: 0,
    r2: Number.NaN,
    trendLinePoints: [{}, {}]
  },
  histogramSeries: {
    normalDistribNumberOfPoints: 100
  },
  lineMarker: {
    markerSize: 7,
    markerStyle: s.Circle
  },
  lineSeries: {
    strokeWidth: 2,
    lineSmoothed: !1,
    lineSmoothedKind: o.Bezier,
    showArea: !1,
    areaOpacity: 0.25,
    markerVisible: !0,
    rotated: !1,
    stackedType: l.Side,
    stackNegativeValuesToBaseline: !0,
    marker: {
      markerSize: 4,
      markerStrokeWidth: 2
    }
  },
  buffer: !1,
  elementLimitForSmoothRendering: 1e3,
  // for line and scatter-plot
  defaultStatParameterValue: 0.5,
  // outStatistics
  pieChart: {
    dataLabels: { maxWidth: 100, wrap: !1 },
    dataLabelsOffset: 10,
    dataLabelsInside: !1,
    chartRadius: 80,
    chartOptimizedRadius: 70,
    innerRadius: c,
    min: -90,
    max: 270,
    displayCategoryOnLegend: !0,
    displayCategoryOnTooltip: !0,
    displayCategoryOnDataLabel: !0,
    displayNumericValueOnLegend: !1,
    displayNumericValueOnDataLabel: !0,
    displayNumericValueOnTooltip: !0,
    displayPercentageOnLegend: !1,
    displayPercentageOnDataLabel: !1,
    displayPercentageOnTooltip: !0,
    sliceGrouping: {
      sliceId: g.defaultGroupOfSlicesId,
      percentageThreshold: 0,
      // 0 means no grouping
      fillSymbol: {
        type: a.SFS,
        color: [210, 210, 210, 255 * u]
        // light grey, 70% opacity
      }
    },
    defaultPadding: 0,
    alignDataLabels: !1,
    ticksOpacity: 1,
    minimumSize: 20,
    technicalFieldSliceId: "arcgis_charts_slice_id",
    technicalFieldSlicePercent: "arcgis_charts_slice_percent",
    technicalFieldGroupedSlices: "arcgis_charts_grouped_slices"
  },
  pieSliceCustomLabelName: "customLabel",
  yAxisLabelMaxWidth: 100,
  xAxisLabelMaxWidth: 75,
  axisLabelPadding: 1.5,
  categoryAxisLabelPadding: 3,
  seriesSort: n.Ascending,
  technicalSortingFieldBySum: "arcgis_charts_category_sum",
  technicalSortingFieldByPercent: "arcgis_charts_category_first_series_percent",
  technicalFieldSplitByField: "arcgis_charts_split_by_field",
  color: [0, 0, 0, 255],
  chartEllipsis: "...",
  backgroundColor: [255, 255, 255, 255],
  lightTextColor: [255, 255, 255, 255],
  // white
  darkTextColor: [0, 0, 0, 255],
  // black
  grayColor: i,
  // medium gray
  xAxisMinGridDistance: 120,
  xCategoryAxisMinGridDistance: 40,
  yAxisMinGridDistance: 40,
  yCategoryAxisMinGridDistance: 20,
  secondRotatedValueAxisMinGridDistance: 80,
  staggerMinGridDistance: 40,
  shortestMinGridDistance: 20,
  maxNumberOfXAxisLabelsInWrapModeForDateAxis: 15,
  maxNumberOfXAxisLabelsInWrapMode: 20,
  lineWidth: 1,
  loaderRotationSpeed: 400,
  loaderIntervalTimeout: 400,
  loaderBackground: [255, 255, 255, 230],
  loaderSpinnerColor: [0, 122, 194, 255],
  loaderLabelFontSize: 20,
  selectionTheme: {
    selectedElementsTheme: {
      elementOutlineColor: [0, 255, 255, 255],
      elementOutlineWidth: 5
    }
  },
  columnTooltipPosition: "pointer",
  markerTooltipPosition: "fixed",
  patternLineProperties: {
    width: 5,
    height: 5,
    strokeWidth: 1
  },
  patternRectProperties: {
    rectWidth: 9,
    rectHeight: 9,
    width: 10,
    height: 10,
    strokeWidth: 0.3,
    fillOpacity: 0
  },
  originalCategoryValue: "arcgis_charts_original_category_value",
  originalDateValue: "arcgis_charts_original_date_value",
  typeOrDomainFieldName: "arcgis_charts_type_domain_field_name",
  typeOrDomainIdValue: "arcgis_charts_type_domain_id_value",
  dateBinningRange: "arcgis_charts_bin_range",
  originalServerBinningRange: "arcgis_charts_server_bin_range",
  dateBinModel: "arcgis_charts_bin_model",
  xAxisField: "arcgis_charts_axis_field",
  nullValue: "arcgis_charts_null",
  queryMaxFloatPrecision: 10,
  dateFormatUTC: {
    timeZone: "UTC",
    year: "numeric",
    month: "numeric",
    day: "numeric",
    hour: "numeric",
    minute: "numeric",
    second: "numeric",
    hour12: !0
  },
  dayDefaultFormat: {
    day: "numeric",
    month: "short"
  },
  weekDefaultFormat: {
    day: "numeric",
    month: "short"
  },
  rendererOtherValue: "arcgisChartsrendererOtherValue",
  rendererSymbolField: "arcgis_charts_renderer_symbol",
  scatterplotAggregationFields: {
    dataContextXMinField: "arcgis_charts_xmin",
    dataContextXMaxField: "arcgis_charts_xmax",
    dataContextYMinField: "arcgis_charts_ymin",
    dataContextYMaxField: "arcgis_charts_ymax",
    dataContextCountDistinctX: "distinctX",
    dataContextCountDistinctY: "distinctY",
    queryNumberOfFeatures: "numberOfFeatures",
    queryNumberOfAggregatedFeatures: "numberOfAggregatedFeatures",
    xStep: "xStep",
    yStep: "yStep",
    rowCount: "rowCount",
    columnCount: "columnCount",
    realGridSize: "realGridSize"
  },
  dataTooltipVisible: !0,
  dataTooltipStyle: {
    outline: {
      type: a.SLS,
      color: [255, 255, 255, 255],
      width: 1
    }
  },
  dataTooltipStyleReversed: {
    color: [255, 255, 255, 255],
    outline: {
      type: a.SLS,
      width: 3
    }
  },
  tooltipWidth: 300,
  axisLabelOveredOpacity: 0.5,
  // 50% of original opacity
  generalFontSize: 10,
  actionMode: h.MultiSelectionWithCtrlKey,
  debounceTime: 100,
  // 100 ms
  chartDefaultPaddingRight: 0,
  // 0 px default chart's plotting area right padding
  htmlEmptyContent: "<p></p>",
  boxPlot: {
    boxPlotCategory: "arcgis_charts_box_plot_category",
    boxPlotOutliersSuffix: "_arcgis_charts_outliers",
    boxPlotOutliersVisibility: !1,
    boxPlotStandardizeValuesVisibility: !1,
    boxPlotOutliersSize: 4,
    boxPlotLineWidth: 2,
    boxPlotOutlierId: "is_arcgis_charts_outlier",
    boxPlotMeanLineMarkerId: "is_arcgis_charts_mean_marker",
    boxPlotWhiskersColor: [153, 153, 153, 178.5],
    boxPlotTooltipLabelWidth: 250
  },
  binTemporalData: !0,
  lastTemporalBin: "is_arcgis_charts_last_bin",
  options: {
    hideEmptySeries: !0,
    replaceNoValueCategoryWithZero: !0,
    autoInverseDataLabelTextColor: !1
  },
  nullCategory: {
    text: "null",
    symbol: {
      type: a.SFS,
      color: i,
      outline: {
        type: a.SLS,
        style: r.Solid,
        color: i,
        width: 1
      }
    }
  }
};
export {
  P as C
};
