// @arcgis/charts-components/dist/chunks/line-chart-model.js
import "@arcgis/toolkit/intl";
import "lodash-es";
import { W as l } from "./series-types.js";
import "d3-array";
import { aL as n, az as r, a6 as m, Y as h } from "./interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "./runtime.js";
import "@arcgis/toolkit/dom";
import "@arcgis/core/intl.js";
import "./model-with-multi-axes.js";
import { M as f, a as S } from "./model-with-baseline-stacking.js";
import { s as d, a as c, b as g, c as y, d as u } from "./baseline-stacking.js";
import { S as p, s as b, a as I } from "./serial-chart-model.js";
import { c as o, d as L } from "./data2.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
class k extends p {
  constructor() {
    super(...arguments), this._chartType = l.LineSeries;
  }
  async setup(e) {
    await this.loadConfigFromLayer(e), this.setupLocalProps(), e.config !== void 0 && await this.updateSeriesFromLayerFilter(), this.initDefaultSeries(), typeof this.resolveSetup == "function" && this.resolveSetup();
  }
  // #end region
  // ---------------------------------------------------------------------------------------------
  // #region Protected methods
  // ---------------------------------------------------------------------------------------------
  generateDefaultConfig() {
    const e = n(this.commonStrings);
    return e.series = [
      r({
        commonStrings: this.commonStrings,
        objectIDField: this.layerInfo?.objectIdField
      })
    ], e;
  }
  /**
   * Initializes the default series for the line chart. Used in resetting the series styling.
   * @category Private
   */
  initDefaultSeries() {
    this.defaultSeries = r({
      commonStrings: this.commonStrings,
      objectIDField: this.layerInfo?.objectIdField
    });
  }
  // #region
  /**
   * Resets the series styling to the default styling.
   */
  resetSeriesStyling() {
    this.guardSetterFromAsyncOverwrite(() => {
      for (let e = 0; e < this.seriesLength; e += 1) {
        const t = m({
          colorIndex: e,
          opacity: h
        });
        if (b(t, e, this._config), d(t, e, this._config), this.layerInfo !== void 0) {
          const i = this.getSeriesId(e), s = o({
            config: this._config,
            seriesIndex: e,
            layerInfo: this.layerInfo,
            commonStrings: this.commonStrings
          });
          s !== this.layerInfo.objectIdField && I(s ?? i, e, this._config);
        }
        c(this.defaultSeries?.lineSymbol?.style, e, this._config), g(this.defaultSeries?.lineSymbol?.width, e, this._config), y(this.defaultSeries?.markerSymbol?.style, e, this._config), u(this.defaultSeries?.markerSymbol?.size, e, this._config);
      }
      this._config.colorMatch = !0, this.emitConfigUpdate("resetSeriesStyling");
    });
  }
  /**
   * Returns whether the series styling is able to be reset.
   * @returns {boolean}
   */
  resetAvailable() {
    const e = !this.colorMatch;
    let t = !1;
    if (!e)
      for (let i = 0; i < this.seriesLength; i += 1) {
        if (this.layerInfo !== void 0) {
          const s = o({
            config: this._config,
            seriesIndex: i,
            layerInfo: this.layerInfo,
            commonStrings: this.commonStrings
          }), a = this.getSeriesName(i);
          s !== void 0 && s !== this.layerInfo.objectIdField && s !== a && (t = !0);
        }
        if (this.defaultSeries?.lineSymbol?.style !== void 0 && this.defaultSeries?.lineSymbol.style !== this.getSeriesLineStyle(i) || this.defaultSeries?.lineSymbol?.width !== void 0 && this.defaultSeries?.lineSymbol.width !== this.getSeriesLineWidth(i) || this.defaultSeries?.markerSymbol?.style !== void 0 && this.defaultSeries?.markerSymbol?.style !== this.getSeriesMarkerStyle(i) || this.defaultSeries?.markerSymbol?.size !== void 0 && this.defaultSeries?.markerSymbol?.size !== this.getSeriesMarkerSize(i)) {
          t = !0;
          break;
        }
      }
    return e || t;
  }
}
L(k, [f, S]);
export {
  k as LineChartModel
};
