// @arcgis/charts-components/dist/chunks/rest-js-object-literals.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const e = {
  Blob: "esriFieldTypeBlob",
  Date: "esriFieldTypeDate",
  DateOnly: "esriFieldTypeDateOnly",
  TimestampOffset: "esriFieldTypeTimestampOffset",
  Double: "esriFieldTypeDouble",
  Geometry: "esriFieldTypeGeometry",
  GlobalID: "esriFieldTypeGlobalID",
  GUID: "esriFieldTypeGUID",
  Integer: "esriFieldTypeInteger",
  SmallInteger: "esriFieldTypeSmallInteger",
  BigInteger: "esriFieldTypeBigInteger",
  OID: "esriFieldTypeOID",
  Raster: "esriFieldTypeRaster",
  Single: "esriFieldTypeSingle",
  String: "esriFieldTypeString",
  XML: "esriFieldTypeXML"
}, i = {
  Dash: "esriSLSDash",
  DashDot: "esriSLSDashDot",
  DashDotDot: "esriSLSDashDotDot",
  Dot: "esriSLSDot",
  Null: "esriSLSNull",
  Solid: "esriSLSSolid"
}, l = {
  Circle: "esriSMSCircle",
  Cross: "esriSMSCross",
  Diamond: "esriSMSDiamond",
  Square: "esriSMSSquare",
  X: "esriSMSX",
  Triangle: "esriSMSTriangle"
}, s = {
  BackwardDiagonal: "esriSFSBackwardDiagonal",
  Cross: "esriSFSCross",
  DiagonalCross: "esriSFSDiagonalCross",
  ForwardDiagonal: "esriSFSForwardDiagonal",
  Horizontal: "esriSFSHorizontal",
  Null: "esriSFSNull",
  Solid: "esriSFSSolid",
  Vertical: "esriSFSVertical"
}, r = {
  Left: "left",
  Right: "right",
  Center: "center",
  Justify: "justify"
}, t = {
  Baseline: "baseline",
  Top: "top",
  Middle: "middle",
  Bottom: "bottom"
}, S = {
  SLS: "esriSLS",
  SMS: "esriSMS",
  SFS: "esriSFS",
  PMS: "esriPMS",
  PFS: "esriPFS",
  TS: "esriTS"
}, a = {
  Italic: "italic",
  Normal: "normal",
  Oblique: "oblique"
}, o = {
  Bold: "bold",
  Bolder: "bolder",
  Lighter: "lighter",
  Normal: "normal"
}, n = {
  LineThrough: "line-through",
  Underline: "underline",
  None: "none"
}, d = {
  UniqueValue: "unique-value",
  Simple: "simple",
  ClassBreaks: "class-breaks",
  Heatmap: "heatmap",
  DotDensity: "dot-density",
  Dictionary: "dictionary",
  PieChart: "pie-chart"
}, T = {
  Feet: "feet",
  Miles: "miles",
  NauticalMiles: "nautical-miles",
  USNauticalMiles: "us-nautical-miles",
  Meters: "meters",
  Kilometers: "kilometers"
};
export {
  S as R,
  e as a,
  s as b,
  i as c,
  l as d,
  r as e,
  t as f,
  a as g,
  o as h,
  n as i,
  d as j,
  T as k
};
