// @arcgis/charts-components/dist/chunks/series-types.js
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const a = {
  Chart: "chart",
  Text: "chartText",
  Overlay: "chartOverlay",
  Overlays: "chartOverlays",
  Legend: "chartLegend",
  CursorCrosshair: "cursorCrosshair",
  Axis: "chartAxis",
  GaugeAxisTick: "gaugeAxisTick",
  GaugeNeedle: "gaugeNeedle",
  CategoryAxisFormat: "category",
  NumberAxisFormat: "number",
  DateAxisFormat: "date",
  Guide: "chartGuide",
  ScatterSeries: "scatterSeries",
  PieSeries: "pieSeries",
  PieTick: "pieTick",
  GaugeSeries: "gaugeSeries",
  LineSeries: "lineSeries",
  BarSeries: "barSeries",
  HistogramSeries: "histogramSeries",
  BoxPlotSeries: "boxPlotSeries",
  RadarSeries: "radarSeries",
  ComboLineAndBarSeries: "comboLineAndBarSeries",
  HeatSeries: "heatSeries",
  Buffer: "buffer",
  SizePolicy: "sizeScale",
  CalendarDateParts: "calendarDateParts"
}, l = {
  Count: "count",
  Sum: "sum",
  Minimum: "min",
  Maximum: "max",
  Average: "avg",
  StandardDeviation: "stddev",
  Variance: "var",
  ContinuousPercentile: "percentile-continuous",
  DiscretePercentile: "percentile-discrete",
  EnvelopeAggregate: "envelope-aggregate",
  CentroidAggregate: "centroid-aggregate",
  ConvexHullAggregate: "convex-hull-aggregate",
  NoAggregation: "no_aggregation"
}, u = {
  None: "none",
  Log: "logarithmic",
  Sqrt: "squareRoot"
}, h = {
  Single: "singleColor",
  // fixed color value as indicated in markerSymbol
  Match: "colorMatch"
  // apply the color from the layer's renderer symbology
}, S = {
  Left: "left",
  Right: "right",
  Top: "top",
  Bottom: "bottom"
}, C = {
  Side: "sideBySide",
  Stacked: "stacked",
  Stacked100: "stacked100"
}, m = {
  Side: "sideBySide",
  MeanLines: "meanLines"
}, d = {
  Ascending: "ASC",
  Descending: "DESC"
}, y = {
  statisticType: l.Count,
  sortingOrder: d.Ascending
}, f = {
  Bezier: "bezier",
  MonotoneX: "monotoneX",
  MonotoneY: "monotoneY"
}, b = {
  Second: "seconds",
  Minute: "minutes",
  Hour: "hours",
  Day: "days",
  Week: "weeks",
  Month: "months",
  Quarter: "quarters",
  Year: "years"
}, p = {
  DayOfMonth: "dayOfMonth",
  DayOfWeek: "dayOfWeek",
  DayOfYear: "dayOfYear",
  HourOfDay: "hourOfDay",
  MonthOfYear: "monthOfYear",
  MinuteOfDay: "minuteOfDay",
  WeekOfYear: "weekOfYear",
  QuarterOfYear: "quarterOfYear"
}, O = {
  Second: "seconds",
  Minute: "minutes",
  Hour: "hours",
  Day: "days"
}, B = {
  Gradient: "gradient"
}, W = {
  Null: "null",
  Zero: "zero",
  Interpolate: "interpolate"
}, v = {
  Start: "equalIntervalsFromStartTime",
  End: "equalIntervalsFromEndTime"
}, x = {
  Hide: "hide",
  Rotate: "rotate",
  Stagger: "stagger",
  Wrap: "wrap"
}, A = {
  Same: "same",
  Newer: "newer",
  Older: "older",
  Unrecognized: "unrecognized"
}, T = {
  Value: "value",
  Date: "date",
  Category: "category"
}, D = {
  FeatureGauge: "featureGauge",
  StatisticGauge: "statisticGauge"
}, P = {
  Linear: "linear",
  Logarithmic: "logarithmic"
}, L = {
  Radial: "radial",
  Circular: "circular",
  Horizontal: "horizontal"
}, M = {
  YValue: "arcgis-charts-y-value",
  Mean: "arcgis-charts-mean",
  Median: "arcgis-charts-median",
  Category: "arcgis-charts-category"
};
function i(e) {
  return e.size === 2 && e.has(a.LineSeries) && e.has(a.BarSeries);
}
function n(e) {
  return e.size === 2 && e.has(a.LineSeries) && e.has(a.BoxPlotSeries) || e.size === 1 && e.has(a.BoxPlotSeries);
}
function k(e, t) {
  let r;
  if (e !== void 0 && e.length > 0)
    if (t !== void 0 && t >= 0 && t < e.length)
      r = e[t].type;
    else {
      const s = new Set(e.map((g) => g.type)), o = i(s), c = n(s);
      o === !0 ? r = a.ComboLineAndBarSeries : c ? r = a.BoxPlotSeries : s.size >= 2 ? r = void 0 : r = e[0].type;
    }
  return r;
}
function z(e, t = !0) {
  return t ? e?.xTemporalBinning : e?.yTemporalBinning;
}
const G = {
  isComboLineAndBarChart: i,
  isBoxPlotChart: n
};
export {
  D as G,
  y as S,
  a as W,
  l as a,
  u as b,
  h as c,
  S as d,
  C as e,
  m as f,
  d as g,
  f as h,
  b as i,
  p as j,
  O as k,
  B as l,
  W as m,
  v as n,
  x as o,
  A as p,
  T as q,
  P as r,
  L as s,
  M as t,
  k as u,
  z as v,
  G as w
};
