// @arcgis/charts-components/dist/chunks/type-guards.js
import { u as a, W as i } from "./series-types.js";
import { M as t } from "./common.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
function s(e) {
  switch (e) {
    case i.BarSeries:
      return t.BarChart;
    case i.LineSeries:
      return t.LineChart;
    case i.ComboLineAndBarSeries:
      return t.ComboBarLineChart;
    case i.BoxPlotSeries:
      return t.BoxPlot;
    case i.PieSeries:
      return t.PieChart;
    case i.ScatterSeries:
      return t.Scatterplot;
    case i.HeatSeries:
      return t.HeatChart;
    case i.HistogramSeries:
      return t.Histogram;
    case i.GaugeSeries:
      return t.Gauge;
    case i.RadarSeries:
      return t.RadarChart;
    default:
      return;
  }
}
function c(e) {
  const r = a(e.series);
  return s(r);
}
function d(e, r) {
  const n = s(e?.chartType);
  return e !== void 0 && n === r;
}
function h(e) {
  return e !== void 0 && "config" in e;
}
function p(e) {
  return e !== void 0 && e.chartType === i.BarSeries;
}
function f(e) {
  return e !== void 0 && e.chartType === i.LineSeries;
}
function y(e) {
  return e !== void 0 && (e.chartType === i.LineSeries || e.chartType === i.BarSeries || e.chartType === i.ComboLineAndBarSeries || e.chartType === i.RadarSeries);
}
function S(e) {
  return e !== void 0 && e.chartType === i.PieSeries;
}
function T(e) {
  return e !== void 0 && e.chartType === i.BoxPlotSeries;
}
function l(e) {
  return e !== void 0 && e.chartType === i.ScatterSeries;
}
function M(e) {
  return e !== void 0 && e.chartType === i.HistogramSeries;
}
function g(e) {
  return e !== void 0 && "xAxisTitleText" in e;
}
function v(e) {
  const r = [
    i.BarSeries,
    i.BoxPlotSeries,
    i.HistogramSeries,
    i.ScatterSeries,
    i.LineSeries,
    i.ComboLineAndBarSeries
  ];
  return e !== void 0 && r.includes(e.chartType) && "getAxisTitleText" in e && typeof e.getAxisTitleText == "function";
}
function x(e) {
  return e !== void 0 && "rotatedState" in e;
}
function C(e) {
  return e !== void 0 && "dataLabelsVisibility" in e;
}
function B(e) {
  return e !== void 0 && "getGuideVisibility" in e && typeof e.getGuideVisibility == "function";
}
function A(e) {
  return e !== void 0 && "addXAxisGuide" in e && typeof e.addXAxisGuide == "function";
}
function L(e) {
  return e !== void 0 && "addYAxisGuide" in e && typeof e.addYAxisGuide == "function";
}
function b(e) {
  return e !== void 0 && "legendVisibility" in e;
}
function G(e) {
  return e !== void 0 && "temporalBinningSize" in e;
}
function W(e, r) {
  const n = s(e?.chartType);
  return e !== void 0 && n === r;
}
export {
  p as a,
  f as b,
  y as c,
  S as d,
  T as e,
  l as f,
  M as g,
  g as h,
  h as i,
  v as j,
  x as k,
  C as l,
  B as m,
  A as n,
  L as o,
  b as p,
  G as q,
  W as r,
  c as s,
  d as t,
  s as u
};
