// @arcgis/map-components/dist/components/arcgis-feature-content/customElement.js
import { c as l } from "../../chunks/runtime.js";
import { css as g, html as r } from "lit";
import { ref as f } from "lit/directives/ref.js";
import { LitElement as h, safeClassMap as i } from "@arcgis/lumina";
import { n, j as d } from "../../chunks/feature-utils.js";
import { watch as u } from "@arcgis/core/core/reactiveUtils.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const m = g`.arcgis-feature-content__loader-container{display:flex;align-items:center;justify-content:center;padding:0;height:150px}.arcgis-feature-content{font-size:var(--calcite-font-size);line-height:var(--calcite-font-line-height-relative-normal)}.arcgis-feature-content h1,.arcgis-feature-content h2,.arcgis-feature-content h3,.arcgis-feature-content h4,.arcgis-feature-content h5,.arcgis-feature-content h6{margin:var(--calcite-spacing-sm) 0;line-height:var(--calcite-font-line-height-relative-tight);letter-spacing:0;color:var(--calcite-color-text-1);font-weight:var(--calcite-font-weight-semibold)}.arcgis-feature-content h1{font-size:var(--calcite-font-size-xxl)}.arcgis-feature-content h2{font-size:var(--calcite-font-size-xl)}.arcgis-feature-content h3{font-size:var(--calcite-font-size-lg)}.arcgis-feature-content h4,.arcgis-feature-content h5,.arcgis-feature-content h6{font-size:var(--calcite-font-size)}.arcgis-feature-element-info__description{margin-bottom:var(--calcite-spacing-xs)}.arcgis-feature-content p{margin:0 0 var(--calcite-spacing-xs);font-size:var(--calcite-font-size)}.arcgis-feature-content p:last-child{margin-block-end:0}.arcgis-feature-content img{max-width:100%;max-height:100%;image-orientation:from-image}.arcgis-feature-content video{max-width:100%;max-height:100%}.arcgis-feature-content figure{margin:0;padding:0}.arcgis-feature-content figcaption{display:block;margin:var(--calcite-spacing-xs) 0 0;padding:0;font-size:var(--calcite-font-size-sm);font-style:italic}.arcgis-feature-content ul,.arcgis-feature-content ol{margin-block:var(--calcite-spacing-sm)}.arcgis-feature-content ul:first-child,.arcgis-feature-content ol:first-child{margin-block-start:0}.arcgis-feature-content a,.arcgis-feature-content a:hover,.arcgis-feature-content a:focus{color:var(--calcite-color-text-link)}`, c = "arcgis-feature-content", o = {
  base: c,
  loaderContainer: `${c}__loader-container`
};
class p extends h {
  static {
    this.properties = { _loadingPromise: 16, _created: 16, creator: 1, destroyer: 0, graphic: 0 };
  }
  static {
    this.styles = m;
  }
  loaded() {
    this.manager.onLifecycle(() => u(() => this.creator, (t) => {
      this._destroyContent(), this._createContent(t);
    }, { initial: !0 }));
  }
  async _createContent(t) {
    const { graphic: e } = this;
    if (!e || !t)
      return;
    const a = n({
      type: "content",
      value: t,
      event: { graphic: e }
    });
    this._loadingPromise = a;
    const s = await a;
    a === this._loadingPromise && (this._loadingPromise = void 0, this._created = s);
  }
  _destroyContent() {
    const { _created: t, graphic: e, destroyer: a } = this;
    t && e && (n({ type: "content", value: a, event: { graphic: e } }), this._created = null);
  }
  _addTargetToAnchors(t) {
    t && Array.from(t.querySelectorAll("a")).forEach((e) => {
      d(e.href) && !e.hasAttribute("target") && e.setAttribute("target", "_blank");
    });
  }
  render() {
    return r`<div class=${i(o.base)}>${this._loadingPromise ? this._renderLoading() : this._renderCreated()}</div>`;
  }
  _renderLoading() {
    return r`<div class=${i(o.loaderContainer)}><calcite-loader inline label></calcite-loader></div>`;
  }
  _renderCreated() {
    const { _created: t } = this;
    return t instanceof HTMLElement ? t : typeof t == "string" ? r`<div .innerHTML=${t ?? ""} ${f(this._addTargetToAnchors)}></div>` : null;
  }
}
l("arcgis-feature-content", p);
export {
  p as ArcgisFeatureContent
};
