// @arcgis/charts-components/dist/components/arcgis-charts-config-scatter-plot/customElement.js
import { c as r } from "../../chunks/runtime.js";
import "@arcgis/toolkit/intl";
import { isEqual as n } from "lodash-es";
import "../../chunks/series-types.js";
import "d3-array";
import { g as l, ba as h, c0 as d } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { h as c } from "../../chunks/index4.js";
import "@arcgis/toolkit/dom";
import { LitElement as g, createEvent as m } from "@arcgis/lumina";
import { css as p, html as t } from "lit";
import { C as o } from "../../chunks/interfaces3.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const f = p`:host{overflow-y:auto;overflow-x:hidden;margin-bottom:var(--arcgis-charts-intra-section-margin);margin-top:var(--arcgis-charts-intra-section-margin);height:100%}`;
class u extends g {
  constructor() {
    super(...arguments), this.onModelConfigChange = async (e) => {
      const i = e?.detail.config ?? this.model.config;
      this.layer === void 0 && await this.updateFetchLayerFieldsInfo(i);
    }, this.popoverPlacement = "leading", this.arcgisChartsConfigWebChartChange = m();
  }
  static {
    this.properties = { layerFieldsInfo: 16, model: 0, popoverPlacement: 3, r2: 11, calculatedDataMinMaxBounds: 0, calculatedMinMaxBounds: 2, page: 3, layer: 0, usePopupTemplateFieldsInfo: 5 };
  }
  static {
    this.styles = f;
  }
  async load() {
    this.layer !== void 0 && await this.layerChange(this.layer, void 0), this.usePopupTemplateFieldsInfo !== void 0 && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, void 0), this.modelChange(this.model), this.onModelConfigChange();
  }
  async willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model")), e.has("r2") && this.r2Change(this.r2), e.has("layer") && await this.layerChange(this.layer, e.get("layer")), e.has("usePopupTemplateFieldsInfo") && await this.usePopupTemplateFieldsInfoChange(this.usePopupTemplateFieldsInfo, e.get("usePopupTemplateFieldsInfo")), this.layerFieldsInfo === void 0 && await this.updateFetchLayerFieldsInfo();
  }
  modelChange(e, a) {
    a?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  r2Change(e) {
    Number.isFinite(e) || (this.r2 = void 0);
  }
  async layerChange(e, a) {
    e !== void 0 && !n(a, e) && await this.updateFetchLayerFieldsInfo();
  }
  async usePopupTemplateFieldsInfoChange(e, a) {
    e !== void 0 && e !== a && await this.updateFetchLayerFieldsInfo();
  }
  async updateFetchLayerFieldsInfo(e) {
    const a = e ?? this.model.config;
    let i = this.layer;
    if (!i) {
      const s = c(a) ? a.iLayer : void 0;
      s && (i = await l(s));
    }
    i && (this.layerFieldsInfo = await h(i, this.usePopupTemplateFieldsInfo));
  }
  render() {
    let e;
    switch (this.page) {
      case o.Data:
        e = t`<arcgis-charts-config-scatter-plot-data .model=${this.model} .r2=${this.r2} .layerFieldsInfo=${this.layerFieldsInfo} .isTableLayer=${d(this.layer)?.isTable}></arcgis-charts-config-scatter-plot-data>`;
        break;
      case o.Axes:
        e = t`<arcgis-charts-config-axes .model=${this.model} show-min-x-bound-option show-max-x-bound-option show-min-y-bound-option show-max-y-bound-option show-enable-x-log-option show-enable-y-log-option .calculatedMinMaxBounds=${this.calculatedMinMaxBounds} .calculatedDataMinMaxBounds=${this.calculatedDataMinMaxBounds}></arcgis-charts-config-axes>`;
        break;
      case o.Guides:
        e = t`<arcgis-charts-config-guides .model=${this.model} enable-x-axis-guide enable-y-axis-guide></arcgis-charts-config-guides>`;
        break;
      case o.Format:
        e = t`<arcgis-charts-config-format .model=${this.model}></arcgis-charts-config-format>`;
        break;
      case o.General:
        e = t`<arcgis-charts-config-general .model=${this.model}></arcgis-charts-config-general>`;
        break;
      default:
        e = "";
        break;
    }
    return e;
  }
}
r("arcgis-charts-config-scatter-plot", u);
export {
  u as ArcgisChartsConfigScatterplot
};
