// @arcgis/charts-components/dist/components/arcgis-charts-config-box-plot-data/customElement.js
import { c as C } from "../../chunks/runtime.js";
import { keyed as f } from "lit/directives/keyed.js";
import { B as c } from "../../chunks/common.js";
import { W as y } from "../../chunks/series-types.js";
import "@arcgis/toolkit/intl";
import { isEqual as b } from "lodash-es";
import "d3-array";
import { a3 as d, cj as S, ci as u } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "@arcgis/toolkit/dom";
import { r as p } from "../../chunks/popover-ui-utils.js";
import { repeat as v } from "lit/directives/repeat.js";
import { a as l, e as m } from "../../chunks/interfaces3.js";
import { d as $, s as x } from "../../chunks/store.js";
import { U as F } from "../../chunks/chart-ui-utils2.js";
import { u as I } from "../../chunks/useT9n.js";
import { css as w, nothing as B, html as a } from "lit";
import { createRef as h, ref as g } from "lit/directives/ref.js";
import { LitElement as P, createEvent as O, safeClassMap as o, renderElement as L } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const r = {
  label: "label",
  fieldList: "field-list",
  fab: "fab",
  dimText: "dim-text",
  list: "list",
  customSortButtons: "custom-sort-buttons"
}, k = w`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin);margin-top:var(--arcgis-charts-top-section-margin)}.label,.disabled-label{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:var(--arcgis-charts-config-margin-minor);justify-content:space-between}.text{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);width:calc(100% - 1.5rem);height:2rem;margin-bottom:var(--arcgis-charts-config-margin-minor);justify-content:space-between}.field-list{max-height:50vh;overflow-y:auto}.disabled-label{pointer-events:none}.fab{display:flex;margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:1.5rem;justify-content:center}.list{margin-left:var(--arcgis-charts-config-margin-minor);margin-right:var(--arcgis-charts-config-margin-minor);margin-bottom:var(--arcgis-charts-config-margin-major)}.custom-sort-buttons{display:flex;justify-content:space-between;align-items:center;font-weight:var(--calcite-font-weight-medium);font-size:var(--calcite-font-size-0)}`;
class E extends P {
  constructor() {
    super(...arguments), this._messages = I(), this.boxPlotPopover = null, this.categoryElement = h(), this.numericFieldsElement = h(), this.splitByFieldElement = h(), this.sortOrderElement = h(), this.originalCustomSortLabels = [], this.isCustomSortDisabled = !1, this.onModelConfigChange = () => {
      this.requestUpdate();
    }, this.popoverClose = (t) => {
      this.boxPlotPopover === t?.target && (this.boxPlotPopover = p(this.boxPlotPopover));
    }, this.onBoxPlotDataChange = async (t) => {
      try {
        if (t.detail !== null && t.detail?.value !== void 0) {
          const { contentKind: e } = t.detail, s = [...t.detail.value];
          switch ($(x), e) {
            case l.category: {
              const i = s[0]?.trim() || d;
              if (this.model.category = i, this.model.getSortOrder() === c.customSort) {
                const n = await this.updateCustomSortProps();
                this.model.orderByList = n, this.sortOrderResetCounter += 1;
              }
              break;
            }
            case l.numericFields: {
              this.model.numericFields = s;
              const i = s.length > 1;
              if (this.model.standardizeValues = i, this.model.getSortOrder() === c.customSort) {
                const n = await this.updateCustomSortProps();
                this.model.orderByList = n, this.sortOrderResetCounter += 1;
              }
              break;
            }
            case l.splitByField: {
              const i = s[0]?.trim() || "";
              this.model.splitByField = i;
              break;
            }
            case l.sortOrder:
              {
                const i = s[0];
                if (i === c.customSort) {
                  const n = await this.updateCustomSortProps();
                  this.model.setSortOrder(i, n);
                } else
                  this.model.setSortOrder(i);
                this.sortOrderResetCounter += 1;
              }
              break;
            default:
              break;
          }
        }
      } catch (e) {
        throw this.arcgisChartsConfigDataChangeError.emit({ error: e, model: this.model }), e;
      }
      this.requestUpdate();
    }, this.selectedCustomSortItemId = "", this.sortOrderResetCounter = 0, this.layerFieldsInfo = [], this.arcgisChartsConfigDataChangeError = O();
  }
  static {
    this.properties = { selectedCustomSortItemId: 16, sortOrderResetCounter: 16, messageOverrides: 0, model: 0, layerFieldsInfo: 0 };
  }
  static {
    this.styles = k;
  }
  load() {
    this.modelChange(this.model), this.onModelConfigChange(), this.currentDataFilter = this.model.getDataFilters(), this.updateCustomSortProps();
  }
  async willUpdate(t) {
    t.has("model") && this.modelChange(this.model, t.get("model")), b(this.model.getDataFilters(), this.currentDataFilter) || (await this.updateCustomSortProps(), this.currentDataFilter = this.model.getDataFilters()), this.disableSplitBy = this.model.numericFields.length === 0 || this.model.numericFields.length > 1 && this.model.category !== d;
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.boxPlotPopover = p(this.boxPlotPopover), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(t, e) {
    e?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), t?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  onCalciteListOrderChange(t) {
    if (t.detail !== null && t.detail !== void 0) {
      const { newIndex: e, oldIndex: s } = t.detail;
      this.model.moveCustomSortValuesInOrderByList(s, e);
    }
  }
  onCustomSortItemSelected(t) {
    const { id: e } = t.target;
    this.selectedCustomSortItemId = this.selectedCustomSortItemId === e ? "" : e;
  }
  onCustomSortItemMoved(t) {
    const { id: e } = t.target;
    let s = this.model.orderByList.length;
    const i = this.model.orderByList.indexOf(this.selectedCustomSortItemId);
    e === m.Up ? s = i - 1 : e === m.Down && (s = i + 1), i >= 0 && s >= 0 && s < this.model.orderByList.length && (this.model.moveCustomSortValuesInOrderByList(i, s), this.sortOrderResetCounter += 1);
  }
  buildCustomSortListItems() {
    return v(this.model.orderByList, (e) => e, (e) => {
      const s = S(this.layerFieldsInfo, e);
      return a`<calcite-list-item id=${e ?? B} .label=${e ?? ""} .description=${s} .value=${e} .selected=${this.selectedCustomSortItemId === e} @calciteListItemSelect=${this.onCustomSortItemSelected}><calcite-icon slot=content-end></calcite-icon></calcite-list-item>`;
    });
  }
  buildCustomSortList() {
    return f(this.sortOrderResetCounter, a`<calcite-list label class=${o(r.list)} drag-enabled selection-mode=single selection-appearance=border @calciteListOrderChange=${this.onCalciteListOrderChange}>${this.buildCustomSortListItems()}</calcite-list>`);
  }
  buildNumericFieldItem(t) {
    return f(t, a`<calcite-list-item .label=${u(this.layerFieldsInfo, t)} .value=${t} closable @calciteListItemClose=${this.onRemoveNumericField}></calcite-list-item>`);
  }
  onCategorySelect(t) {
    this.categoryElement.value && !this.categoryElement.value.disabled && (this.contentKind = l.category, this.setupPopover(t));
  }
  async onRemoveNumericField(t) {
    const e = t.target.value;
    if (this.model.numericFields = this.model.numericFields.filter((s) => s !== e), this.model.numericFields.length === 0 && this.model.resetConfig(), this.model.getSortOrder() === c.customSort) {
      const s = await this.updateCustomSortProps();
      this.model.orderByList = s, this.sortOrderResetCounter += 1;
    }
  }
  async updateCustomSortProps() {
    const t = await this.model.getCustomSortValues();
    this.originalCustomSortLabels = [...t];
    const e = t.length > F.customSortLimit;
    return e && (this.sortOrderElement.value && (this.sortOrderElement.value.label = this._messages.xAxisAsc), this.model.setSortOrder(c.xAxisAsc)), this.isCustomSortDisabled = e, t;
  }
  onStandardizeValuesChange(t) {
    this.model.standardizeValues = t.target.checked;
  }
  onShowOutliersChange(t) {
    this.model.showOutliers = t.target.checked;
  }
  onSplitBySelect(t) {
    this.splitByFieldElement.value && !this.splitByFieldElement.value.disabled && (this.contentKind = l.splitByField, this.setupPopover(t));
  }
  onSortOrderSelect(t) {
    this.contentKind = l.sortOrder, this.setupPopover(t);
  }
  setupPopover(t) {
    t.stopImmediatePropagation(), this.boxPlotPopover = p(this.boxPlotPopover);
    let e, s, i, n = !1;
    switch (this.contentKind) {
      case l.category:
        e = this.categoryElement.value, s = this._messages.categoryHeader, i = this.model.category === d ? "" : this.model.category;
        break;
      case l.numericFields:
        e = this.numericFieldsElement.value, s = this._messages.numericFieldsHeader, i = this.model.numericFields;
        break;
      case l.splitByField:
        e = this.splitByFieldElement.value, s = this._messages.splitByFieldHeader, i = this.model.splitByField;
        break;
      case l.sortOrder:
        e = this.sortOrderElement.value, s = this._messages.selectSortOrder, i = this.model.getSortOrder(), n = this.isCustomSortDisabled;
        break;
    }
    this.boxPlotPopover = L(a`<arcgis-charts-config-box-plot-popover .headingTitle=${s} .referenceElement=${e} open .contentKind=${this.contentKind} .layerFieldsInfo=${this.layerFieldsInfo} .selectedContent=${i} .isCustomSortDisabled=${n} @arcgisChartsConfigBoxPlotDataPopoverChange=${this.onBoxPlotDataChange} @arcgisChartsConfigPopoverClose=${this.popoverClose}></arcgis-charts-config-box-plot-popover>`), !document.body.contains(this.boxPlotPopover) && this.boxPlotPopover !== null && document.body.appendChild(this.boxPlotPopover);
  }
  onAddNumericFields(t) {
    this.contentKind = l.numericFields, this.setupPopover(t);
  }
  getCorrectedSplitByValue() {
    return this.model.splitByField?.trim() === "" ? "" : u(this.layerFieldsInfo, this.model.splitByField);
  }
  getCorrectedCategoryValue() {
    return this.model.category === d ? "" : u(this.layerFieldsInfo, this.model.category);
  }
  resetToOriginalCustomSortLabels() {
    this.selectedCustomSortItemId = "", this.model.orderByList = [...this.originalCustomSortLabels], this.sortOrderResetCounter += 1;
  }
  renderCustomSortUI() {
    const t = this.model.getSortOrder(), e = this.model.getAxisValueFormat(0)?.type, s = this.model.orderByList.indexOf(this.selectedCustomSortItemId);
    return e === y.CategoryAxisFormat && t === c.customSort && !this.isCustomSortDisabled ? a`<calcite-label class=${o(r.label)}><div class=${o(r.customSortButtons)}><div><calcite-button appearance=transparent .disabled=${s === 0 || s === -1 || this.selectedCustomSortItemId.trim() === ""} icon-start=arrow-bold-up id=${m.Up} kind=neutral scale=s @click=${this.onCustomSortItemMoved}></calcite-button><calcite-button appearance=transparent .disabled=${s === -1 || s === this.model.orderByList.length - 1 || this.selectedCustomSortItemId.trim() === ""} icon-start=arrow-bold-down id=${m.Down} kind=neutral scale=s @click=${this.onCustomSortItemMoved}></calcite-button></div><calcite-button @click=${this.resetToOriginalCustomSortLabels} .disabled=${b(this.model.orderByList, this.originalCustomSortLabels)} appearance=transparent icon-end=reset>${this._messages.reset}</calcite-button></div>${this.model.category === d ? this._messages.values : this._messages.category}${this.buildCustomSortList()}</calcite-label>` : null;
  }
  renderNumericFields(t) {
    const e = [];
    for (const s of t)
      e.push(this.buildNumericFieldItem(s));
    return a`<calcite-list label class=${o(r.fieldList)} ${g(this.numericFieldsElement)}>${e}</calcite-list>`;
  }
  renderCategory() {
    const t = this.getCorrectedCategoryValue(), e = this.model.numericFields.length === 0 || (this.model.numericFields ?? []).length > 1 && this.model.splitByField !== "";
    return a`<arcgis-charts-config-field-select .label=${t} .disabled=${e} @click=${this.onCategorySelect} ${g(this.categoryElement)}></arcgis-charts-config-field-select>`;
  }
  renderSplitBy() {
    const t = this.getCorrectedSplitByValue();
    return a`<arcgis-charts-config-field-select .label=${t} .disabled=${this.disableSplitBy} @click=${this.onSplitBySelect} ${g(this.splitByFieldElement)}></arcgis-charts-config-field-select>`;
  }
  renderSorting() {
    return a`<arcgis-charts-config-field-select .label=${this._messages[this.model.getSortOrder()]} @click=${this.onSortOrderSelect} ${g(this.sortOrderElement)}></arcgis-charts-config-field-select>`;
  }
  render() {
    const t = this.renderCategory(), e = this.renderNumericFields([...this.model.numericFields ?? []]), s = this.renderSplitBy(), i = this.renderSorting(), n = this.renderCustomSortUI();
    return a`<div><calcite-label class=${o(r.label)}>${this._messages.numericFields}${e}</calcite-label><div class=${o(r.fab)}><calcite-fab appearance=outline-fill kind=neutral scale=m .label=${this._messages.selectNumericFields} text-enabled .text=${this._messages.selectNumericFields} icon=plus @click=${this.onAddNumericFields}></calcite-fab></div><calcite-label class=${o(r.label)}><div>${this._messages.category} <span class=${o(r.dimText)}>(${this._messages.optional})</span></div>${t}</calcite-label><calcite-label class=${o(r.label)}><div class=${o(this.disableSplitBy ? r.dimText : "")}>${this._messages.splitBy} <span class=${o(r.dimText)}>(${this._messages.optional})</span></div>${s}</calcite-label><div>${(this.model.numericFields ?? []).length > 1 ? a`<calcite-label class=${o(r.label)} layout=inline-space-between>${this._messages.standardizeValues}<calcite-switch .checked=${this.model.standardizeValues} @calciteSwitchChange=${this.onStandardizeValuesChange}></calcite-switch></calcite-label>` : null}</div><calcite-label class=${o(r.label)} layout=inline-space-between>${this._messages.showOutliers}<calcite-switch .checked=${this.model.showOutliers} @calciteSwitchChange=${this.onShowOutliersChange}></calcite-switch></calcite-label><calcite-label class=${o(r.label)}>${this._messages.sortOrder}${i}</calcite-label>${n}</div>`;
  }
}
C("arcgis-charts-config-box-plot-data", E);
export {
  E as ArcgisChartsConfigBoxPlotData
};
