// @arcgis/charts-components/dist/components/arcgis-charts-config-box-plot-popover/customElement.js
import { c as k } from "../../chunks/runtime.js";
import { keyed as p } from "lit/directives/keyed.js";
import { repeat as y } from "lit/directives/repeat.js";
import { B as r } from "../../chunks/common.js";
import "../../chunks/series-types.js";
import { a as u } from "../../chunks/rest-js-object-literals.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { ci as v, bf as g, bd as $ } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { d as F } from "../../chunks/index3.js";
import { e as L, p as x } from "../../chunks/popover-ui-utils.js";
import { substitute as P } from "@arcgis/core/intl.js";
import { a as l } from "../../chunks/interfaces3.js";
import { U as h, e as I, i as f, d as w } from "../../chunks/chart-ui-utils2.js";
import { u as D } from "../../chunks/useT9n.js";
import { css as S, html as n } from "lit";
import { createRef as b, ref as m } from "lit/directives/ref.js";
import { LitElement as T, createEvent as C, renderElement as E, safeStyleMap as B, safeClassMap as d } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const o = {
  jsAppFlyout: "js-app-flyout",
  name: "bar-chart-data-popover",
  numericFieldPickList: "bar-chart-data-numeric-fields-pick-list",
  header: "header",
  fab: "fab",
  pickListItemLabel: "pick-list-item-label",
  sortMean: "sort-mean",
  sortMedian: "sort-median",
  sortOrderPickList: "sort-order-pick-list",
  panel: "panel"
}, N = S`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.name{display:flex;flex-direction:column;margin:.5rem;padding:.5rem;background:#fff;min-width:250px}.header{margin:0;font-weight:bolder}.fab{display:flex;justify-content:stretch}.pick-list-item-label{font-size:small}.sort-icon-color{color:#000}.sort-mean,.sort-median{transform:rotate(90deg)}.sort-order-pick-list{overflow-y:hidden}.panel{max-height:60vh}`;
class z extends T {
  constructor() {
    super(...arguments), this._messages = D(), this.picklist = b(), this.placement = "leading", this.popoverElement = b(), this.offsetDistance = -200, this.pickListHasChanged = !1, this.layerFieldsInfo = [], this.isCustomSortDisabled = !1, this.arcgisChartsConfigBoxPlotDataPopoverChange = C(), this.arcgisChartsConfigPopoverClose = C();
  }
  static {
    this.properties = { messageOverrides: 0, headingTitle: 3, referenceElement: 2, open: 7, contentKind: 1, layerFieldsInfo: 0, isCustomSortDisabled: 5, selectedContent: 1 };
  }
  static {
    this.styles = N;
  }
  async reposition() {
    await this.popoverElement.value?.reposition();
  }
  willUpdate(e) {
    e.has("selectedContent") && this.selectedContentChange(this.selectedContent, e.get("selectedContent"));
  }
  async updated() {
    setTimeout(async () => {
      typeof this.picklist.value?.setFocus == "function" && await this.picklist.value?.setFocus();
    }, h.PopoverTimer), L(this.popoverElement.value, this.open), this.hasUpdated || setTimeout(() => {
      this.popoverElement.value?.reposition();
    }, h.PopoverTimer);
  }
  selectedContentChange(e, t) {
    this.pickListHasChanged = x(e, t);
  }
  closePopover() {
    if (this.contentKind === l.numericFields && this.pickListHasChanged) {
      const e = {
        pickListElement: this.picklist.value,
        eventEmitter: this.arcgisChartsConfigBoxPlotDataPopoverChange,
        contentKind: this.contentKind
      };
      I(e);
    }
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  async onDataContentTypeChange(e) {
    this.pickListHasChanged = !0;
    const t = e.target.selectedItems;
    this.arcgisChartsConfigBoxPlotDataPopoverChange.emit({
      contentKind: this.contentKind,
      value: t.map((i) => i.value)
    }), this.open = !1;
  }
  buildPickListItem(e) {
    return p(e.name, n`<calcite-list-item .label=${v(this.layerFieldsInfo, e.name)} .value=${e.name} .selected=${f(e.name, this.selectedContent)}></calcite-list-item>`);
  }
  buildPickListItems(e) {
    const t = this.layerFieldsInfo?.length ?? 0, i = [];
    switch (e) {
      case l.numericFields: {
        for (let s = 0; s < t; s += 1) {
          const a = this.layerFieldsInfo[s];
          $(a) && i.push(this.buildPickListItem(a));
        }
        break;
      }
      case l.category: {
        const s = p(" ", n`<calcite-list-item label=" " value=" " .selected=${f("", this.selectedContent)}></calcite-list-item>`);
        i.push(s);
        for (let a = 0; a < t; a += 1) {
          const c = this.layerFieldsInfo[a];
          (g(c) || c.type === u.String) && i.push(this.buildPickListItem(c));
        }
        break;
      }
      case l.splitByField: {
        const s = p(" ", n`<calcite-list-item label=" " value=" " .selected=${f("", this.selectedContent)}></calcite-list-item>`);
        i.push(s);
        for (let a = 0; a < t; a += 1) {
          const c = this.layerFieldsInfo[a];
          (g(c) || c.type === u.String) && i.push(this.buildPickListItem(c));
        }
        break;
      }
    }
    return i;
  }
  onNumericFieldsChange(e) {
    e.target !== null && (this.selectedContent = e.target.selectedItems.map((t) => t.value));
  }
  buildSortOrder() {
    const e = Object.keys(r);
    let t;
    return y(e, (i) => i, (i) => {
      let s = "";
      switch (i) {
        case r.xAxisAsc: {
          t = "a-z-down";
          break;
        }
        case r.xAxisDesc: {
          t = "a-z-up";
          break;
        }
        case r.meanAsc:
          t = "sort-descending", s = o.sortMean;
          break;
        case r.meanDesc: {
          t = "sort-ascending", s = o.sortMean;
          break;
        }
        case r.medianAsc: {
          t = "sort-descending", s = o.sortMedian;
          break;
        }
        case r.medianDesc: {
          t = "sort-ascending", s = o.sortMedian;
          break;
        }
        case r.customSort: {
          t = "arrow-up-down";
          break;
        }
      }
      return n`<calcite-list-item .disabled=${this.isCustomSortDisabled && i === r.customSort} .label=${this._messages[i]} .selected=${this.selectedContent === i} .value=${r[i]} @mouseover=${this.createTooltip} @mouseout=${this.destroyTooltip}><calcite-icon class=${`sort-icon-color ${s}`} slot=content-start .icon=${t}></calcite-icon></calcite-list-item>`;
    });
  }
  async onNumericFieldSelectionDone() {
    this.pickListHasChanged && this.closePopover();
  }
  createTooltip(e) {
    if (e.target?.label === this._messages.customSort && this.isCustomSortDisabled) {
      const t = e.target, i = F(t), a = new Intl.NumberFormat(i).format(h.customSortLimit), c = P(this._messages.customSortLimit ?? "", {
        totalLimit: a
      });
      this.tooltip = E(n`<calcite-tooltip style=${B({ opacity: 1, zIndex: "var(--calcite-floating-ui-z-index)" })} .referenceElement=${t} open placement=top>${c}</calcite-tooltip>`), document.body.appendChild(this.tooltip);
    }
  }
  destroyTooltip() {
    w(this.tooltip);
  }
  renderCategory() {
    const e = this.buildPickListItems(l.category);
    return n`<calcite-list label class=${d(o.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${m(this.picklist)}>${e}</calcite-list>`;
  }
  renderNumericFields() {
    const e = this.buildPickListItems(l.numericFields);
    return n`<calcite-list label class=${d(o.numericFieldPickList)} selection-mode=multiple filter-enabled @calciteListChange=${this.onNumericFieldsChange} ${m(this.picklist)}>${e}</calcite-list>`;
  }
  renderSplitByField() {
    const e = this.buildPickListItems(l.splitByField);
    return n`<calcite-list label class=${d(o.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${m(this.picklist)}>${e}</calcite-list>`;
  }
  renderSortOrder() {
    const e = this.buildSortOrder();
    return n`<calcite-list label class=${d(o.sortOrderPickList)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderPopoverInfo() {
    let e;
    switch (this.contentKind) {
      case l.category:
        e = this.renderCategory();
        break;
      case l.numericFields:
        e = this.renderNumericFields();
        break;
      case l.splitByField:
        e = this.renderSplitByField();
        break;
      case l.sortOrder:
        e = this.renderSortOrder();
        break;
    }
    return e;
  }
  render() {
    let e;
    return this.contentKind === l.numericFields && (e = n`<div slot=footer><calcite-fab class=${d(o.fab)} appearance=outline-fill kind=neutral icon .label=${this._messages.selectionDone} scale=s text-enabled .text=${this._messages.selectionDone} @click=${this.onNumericFieldSelectionDone}></calcite-fab></div>`), n`<div class=${d(o.jsAppFlyout)}><calcite-popover class=${d(o.name)} .referenceElement=${this.referenceElement ?? document.body} .placement=${this.placement} .offsetDistance=${this.offsetDistance} .open=${this.open} @calcitePopoverClose=${this.closePopover} label auto-close ${m(this.popoverElement)}><calcite-panel class=${d(o.panel)} closable .closed=${!this.open} @calcitePanelClose=${this.closePopover}><div slot=header-content class=${d(o.header)}>${this.headingTitle ?? ""}</div>${this.renderPopoverInfo()}${e}</calcite-panel></calcite-popover></div>`;
  }
}
k("arcgis-charts-config-box-plot-popover", z);
export {
  z as ArcgisChartsConfigBoxPlotPopover
};
