// @arcgis/portal-components/dist/components/arcgis-portal-classification-banner/customElement.js
import { c as e } from "../../chunks/runtime.js";
import { LitElement as r, safeStyleMap as l } from "@arcgis/lumina";
import { u as c, c as n } from "../../chunks/loadClassificationData.js";
import { u as f } from "../../chunks/useT9n.js";
import { css as d, html as t } from "lit";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const m = d`.banner{box-sizing:border-box;width:100%;left:0;text-align:center;font-family:var(--calcite-font-family);background-color:#fff;color:#000;padding:var(--calcite-spacing-sm);font-size:var(--calcite-font-size);text-transform:uppercase;overflow-wrap:break-word}.invalid-container{background-color:#fff;color:#000;display:flex;align-items:center;justify-content:center;gap:8px}.invalid-icon{--calcite-icon-color: var(--calcite-color-status-danger)}.invalid-message{text-align:center;font-family:var(--calcite-font-family);background-color:#fff;color:#000;padding:var(--calcite-spacing-sm);font-size:var(--calcite-font-size);overflow-wrap:break-word}`;
class h extends r {
  constructor() {
    super(...arguments), this.classificationModel = c(this), this.messages = f(), this._classificationDataTask = n(this), this.portalClassificationSchema = this.classificationModel.portalClassificationSchema, this.status = this.classificationModel.status, this.missingExistingClassification = !1, this.variant = "full", this.autoDestroyDisabled = !1, this.validationErrors = "";
  }
  static {
    this.properties = { portalClassificationSchema: 16, status: 16, itemClassification: 16, portalItem: 0, missingExistingClassification: 5, variant: 1, autoDestroyDisabled: 5, validationErrors: 1 };
  }
  static {
    this.styles = m;
  }
  async destroy() {
    await this.manager.destroy();
  }
  willUpdate(a) {
    if (a.has("portalItem")) {
      const { portalItem: i } = this;
      i?.load().then(() => {
        i?.id && i?.loaded && i?.classification && !this.missingExistingClassification ? this.classificationModel.validateClassification(
          /** don't utilize the cache if preview */
          this.variant === "preview" ? null : i.id,
          i.classification,
          i?.portal.restUrl
        ).catch(console.error).then((s) => {
          this.validationErrors = s ?? "";
        }) : this.validationErrors = "";
      }).catch(console.error);
    }
  }
  getStyle() {
    const a = {};
    if (this.variant === "preview" && (a.fontSize = "var(--calcite-font-size-sm)", a.padding = "var(--calcite-spacing-xs)"), this.itemClassification && this.portalClassificationSchema) {
      const i = this.itemClassification.classification, s = this.portalClassificationSchema.classificationMetadata.classificationValueProperties.find(({ value: o }) => o === i);
      s?.backgroundColor && (a.backgroundColor = s.backgroundColor), s?.textColor && (a.color = s.textColor);
    }
    return a;
  }
  render() {
    return this._classificationDataTask.render({
      /** Do not render anything while fetching */
      pending: void 0,
      /** Do not render anything if there is an error, TODO: confirm error handling case */
      error: void 0,
      complete: () => (
        /** Only render the banner if the portal has a classification schema and it's accessed not anonymously */
        this.portalClassificationSchema && this.portalItem?.portal.authMode !== "anonymous" ? this.validationErrors && this.variant === "full" ? t`<div class="invalid-container"><calcite-icon class="invalid-icon" scale=s icon=exclamation-mark-triangle .textLabel=${this.messages.invalidclassificationerror}></calcite-icon><span class="invalid-message">${this.messages.invalidclassificationerror}</span></div>` : t`<div class="banner" style=${l(this.getStyle())}>${this.itemClassification?.banner ?? this.messages.notyetclassified ?? "Not yet classified"}</div>` : ""
      )
    });
  }
}
e("arcgis-portal-classification-banner", h);
export {
  h as ArcgisPortalClassificationBanner
};
