// @arcgis/charts-components/dist/components/arcgis-charts-config-scatter-plot-popover/customElement.js
import { c as C } from "../../chunks/runtime.js";
import { css as u, html as s } from "lit";
import { keyed as c } from "lit/directives/keyed.js";
import "../../chunks/series-types.js";
import { R as p } from "../../chunks/rest-js-object-literals.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { bd as v, ci as y } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "@arcgis/toolkit/dom";
import { e as b } from "../../chunks/popover-ui-utils.js";
import { S as n } from "../../chunks/interfaces3.js";
import { U as h, a as S, i as d } from "../../chunks/chart-ui-utils2.js";
import { u as $ } from "../../chunks/useT9n.js";
import { createRef as m, ref as g } from "lit/directives/ref.js";
import { LitElement as P, createEvent as a, safeClassMap as l } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const r = {
  jsAppFlyout: "js-app-flyout",
  scatterplotPopover: "scatter-plot-popover",
  fillColor: "fill-color",
  header: "header",
  pickListItemLabel: "pick-list-item-label",
  pickListPanel: "pick-list-panel"
}, F = u`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{width:100%}.pick-list-panel{max-height:60vh}`;
class k extends P {
  constructor() {
    super(...arguments), this.picklist = m(), this._messages = $(), this.popoverElement = m(), this.layerFieldsInfo = [], this.arcgisChartsConfigPopoverClose = a(), this.arcgisChartsConfigScatterplotFillColorChange = a(), this.arcgisChartsConfigScatterplotLineStyleChange = a(), this.arcgisChartsConfigScatterplotSizeVariableFieldChange = a(), this.arcgisChartsConfigScatterplotAdditionalTooltipFieldChange = a(), this.arcgisChartsConfigScatterplotXYFieldChange = a();
  }
  static {
    this.properties = { messageOverrides: 0, headingTitle: 3, contentKind: 3, selectedContent: 3, referenceElement: 2, placement: 3, open: 7, layerFieldsInfo: 0 };
  }
  static {
    this.styles = F;
  }
  async reposition() {
    await this.popoverElement.value?.reposition();
  }
  async updated() {
    setTimeout(async () => {
      await this.picklist.value?.setFocus();
    }, h.PopoverTimer), b(this.popoverElement.value, this.open), this.hasUpdated || setTimeout(() => {
      this.popoverElement.value?.reposition();
    }, h.PopoverTimer);
  }
  seriesPopoverClose() {
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  closePopover() {
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  updateSymbolProperties(e) {
    this.selectedContent = e.target.lineStyle, this.arcgisChartsConfigScatterplotLineStyleChange.emit(e.detail);
  }
  updateColorProperties(e) {
    this.selectedContent = e.detail, this.arcgisChartsConfigScatterplotFillColorChange.emit(e.detail);
  }
  onFieldSelectionChange(e) {
    const t = e.target.selectedItems;
    this.contentKind === n.sizeVariable ? this.arcgisChartsConfigScatterplotSizeVariableFieldChange.emit(t.map((i) => i.value)[0]) : this.contentKind === n.additionalTooltipFields ? this.arcgisChartsConfigScatterplotAdditionalTooltipFieldChange.emit(t.map((i) => i.value)[0]) : this.arcgisChartsConfigScatterplotXYFieldChange.emit(t.map((i) => i.value)[0]), this.open = !1;
  }
  render() {
    const e = this.selectedContent, t = this.selectedContent;
    return s`<div class=${l({ [r.jsAppFlyout]: !0 })}><calcite-popover class=${l(r.scatterplotPopover)} .referenceElement=${this.referenceElement ?? document.body} .placement=${this.placement} .offsetDistance=${S(this.placement)} @calcitePopoverClose=${this.closePopover} .open=${this.open} label auto-close ${g(this.popoverElement)}>${Array.isArray(e) ? s`<arcgis-charts-config-color-style .header=${this._messages.fillColor} class=${l(r.fillColor)} enable-close .open=${this.open} .selectedColor=${e} @arcgisChartsConfigColorStyleClose=${this.seriesPopoverClose} @arcgisChartsConfigColorStyleChange=${this.updateColorProperties}></arcgis-charts-config-color-style>` : null}${t?.type === p.SLS ? s`<arcgis-charts-config-line-style .header=${this._messages.lineStyle} .open=${this.open} enable-close .lineStyle=${t} @arcgisChartsConfigLineStyleClose=${this.seriesPopoverClose} @arcgisChartsConfigLineStyleChange=${this.updateSymbolProperties}></arcgis-charts-config-line-style>` : null}${!Array.isArray(e) && t?.type !== p.SLS ? this.renderVariablePickList() : null}</calcite-popover></div>`;
  }
  renderVariablePickList() {
    const e = this.layerFieldsInfo?.length ?? 0, t = [];
    (this.contentKind === n.sizeVariable || this.contentKind === n.additionalTooltipFields) && t.push(c(" ", s`<calcite-list-item label=" " value=" " .selected=${d("", this.selectedContent)}></calcite-list-item>`));
    for (let i = 0; i < e; i += 1) {
      const o = this.layerFieldsInfo?.[i] ?? void 0;
      if (o !== void 0 && (this.contentKind === n.additionalTooltipFields || v(o))) {
        const f = y(this.layerFieldsInfo, o.name);
        t.push(c(o.name, s`<calcite-list-item .label=${f} .value=${o.name} .selected=${d(o.name, this.selectedContent)}></calcite-list-item>`));
      }
    }
    return s`<calcite-panel class=${l(r.pickListPanel)} closable .closed=${!this.open} @calcitePanelClose=${this.closePopover}><div slot=header-content class=${l(r.header)}>${this.headingTitle ?? ""}</div><calcite-list label class=${l(r.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onFieldSelectionChange} ${g(this.picklist)}>${t}</calcite-list></calcite-panel>`;
  }
}
C("arcgis-charts-config-scatter-plot-popover", k);
export {
  k as ArcgisChartsConfigScatterplotPopover
};
