// @arcgis/charts-components/dist/components/arcgis-charts-config-guides-type-popover/customElement.js
import { c as t } from "../../chunks/runtime.js";
import { html as e } from "lit";
import { LitElement as s, createEvent as i } from "@arcgis/lumina";
import { u as o } from "../../chunks/useT9n.js";
import { createRef as a, ref as r } from "lit/directives/ref.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "../../chunks/series-types.js";
import "d3-array";
import "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "@arcgis/toolkit/dom";
import { e as l } from "../../chunks/popover-ui-utils.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
class p extends s {
  constructor() {
    super(...arguments), this._messages = o(), this.popoverElement = a(), this.onAddXGuide = () => {
      this.open = !1, this.arcgisChartsAddXAxisGuide.emit();
    }, this.onAddYGuide = () => {
      this.open = !1, this.arcgisChartsAddYAxisGuide.emit();
    }, this.enableXAxisGuide = !1, this.enableYAxisGuide = !1, this.placement = "bottom", this.open = !1, this.arcgisChartsAddXAxisGuide = i(), this.arcgisChartsAddYAxisGuide = i(), this.arcgisChartsConfigPopoverClose = i();
  }
  static {
    this.properties = { enableXAxisGuide: 7, enableYAxisGuide: 7, messageOverrides: 0, referenceElement: 2, placement: 3, open: 7 };
  }
  async reposition() {
    await this.popoverElement.value?.reposition();
  }
  updated() {
    l(this.popoverElement.value, this.open);
  }
  guidePopoverClose() {
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  render() {
    return e`<calcite-popover .open=${this.open} .referenceElement=${this.referenceElement ?? document.body} flip-disabled pointer-disabled offset-distance=-18 @calcitePopoverClose=${this.guidePopoverClose} .placement=${this.placement} auto-close label ${r(this.popoverElement)}><calcite-panel .heading=${this._messages.guideType} closable @calcitePanelClose=${this.guidePopoverClose}>${this.enableXAxisGuide ? e`<calcite-action .label=${this._messages.addXAxisGuide} .text=${this._messages.addXAxisGuide ?? ""} @click=${this.onAddXGuide} icon=x-axis-guide text-enabled></calcite-action>` : null}${this.enableYAxisGuide ? e`<calcite-action .label=${this._messages.addYAxisGuide} .text=${this._messages.addYAxisGuide ?? ""} @click=${this.onAddYGuide} icon=y-axis-guide text-enabled></calcite-action>` : null}</calcite-panel></calcite-popover>`;
  }
}
t("arcgis-charts-config-guides-type-popover", p);
export {
  p as ArcgisChartsConfigGuidesTypePopover
};
