// @arcgis/charts-components/dist/components/arcgis-charts-config-color-style/customElement.js
import { c as o } from "../../chunks/runtime.js";
import { css as a, html as s } from "lit";
import { u as i } from "../../chunks/useT9n.js";
import { LitElement as l, createEvent as t, safeClassMap as r } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const n = {
  colorStyleContainer: "color-style-container"
}, c = a`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{overflow-y:auto;width:min-content}:host calcite-panel .color-style-container{display:flex;padding:0 1rem;margin-top:1rem}:host calcite-panel .color-style-container calcite-label{flex:1 1 0px;display:flex;width:0px}.color-styling-panel>.panel-body{padding:var(--arcgis-charts-panel-padding);margin-top:var(--arcgis-charts-inter-section-margin)}`;
class d extends l {
  constructor() {
    super(...arguments), this._messages = i(), this.enableClose = !1, this.arcgisChartsConfigColorStyleChange = t(), this.arcgisChartsConfigColorStyleClose = t();
  }
  static {
    this.properties = { messageOverrides: 0, header: 1, selectedColor: 2, open: 7, enableClose: 7 };
  }
  static {
    this.styles = c;
  }
  closePanel(e) {
    e.target.closed && this.enableClose && (this.open = !1, this.arcgisChartsConfigColorStyleClose.emit());
  }
  colorChange(e) {
    this.selectedColor = e.detail, this.arcgisChartsConfigColorStyleChange.emit(this.selectedColor);
  }
  render() {
    return s`<calcite-panel .heading=${this.header ?? this._messages.header} .closed=${this.open === !1} .closable=${this.enableClose} @calcitePanelClose=${this.closePanel}><div class=${r(n.colorStyleContainer)}><arcgis-charts-config-colors @colorSelect=${this.colorChange} .rgba=${this.selectedColor}></arcgis-charts-config-colors></div></calcite-panel>`;
  }
}
o("arcgis-charts-config-color-style", d);
export {
  d as ArcgisChartsConfigColorStyle
};
