// @arcgis/charts-components/dist/components/arcgis-charts-config-invalid-input-message/customElement.js
import { c as s } from "../../chunks/runtime.js";
import { css as i, html as t } from "lit";
import { c as a } from "../../chunks/chart-ui-utils2.js";
import { LitElement as e, safeClassMap as l } from "@arcgis/lumina";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const n = {
  invalidMessageColor: "invalid-message-color"
}, o = i`:host{display:block}.invalid-message-color{color:red}`;
class r extends e {
  constructor() {
    super(...arguments), this.isInvalid = !1;
  }
  static {
    this.properties = { invalidInputMessage: 3, isInvalid: 5 };
  }
  static {
    this.styles = o;
  }
  render() {
    return this.isInvalid ? t`<calcite-input-message .status=${a.INVALID} icon=exclamationMarkTriangle .hidden=${!this.isInvalid} class=${l(n.invalidMessageColor)}>${this.invalidInputMessage}</calcite-input-message>` : null;
  }
}
s("arcgis-charts-config-invalid-input-message", r);
export {
  r as ArcgisChartsConfigInvalidInputMessage
};
