// @arcgis/map-components/dist/components/arcgis-feature-element-info/customElement.js
import { c as a } from "../../chunks/runtime.js";
import { css as c, html as i } from "lit";
import { H as o } from "../../chunks/heading.js";
import { LitElement as l, safeClassMap as r } from "@arcgis/lumina";
import { renderingSanitizer as s } from "@arcgis/core/applications/Components/sanitizerUtils.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const g = c`.arcgis-feature-element-info__heading{margin-top:0;margin-bottom:var(--calcite-spacing-xxs);color:var(--calcite-color-text-2);font-size:var(--calcite-font-size-xl);font-weight:var(--calcite-font-weight-regular);line-height:var(--calcite-font-line-height-relative-tight)}.arcgis-feature-element-info__description{margin-bottom:var(--calcite-spacing-sm);font-size:var(--calcite-font-size-sm);font-weight:var(--calcite-font-weight-regular);line-height:var(--calcite-font-line-height-relative-normal)}`, t = "arcgis-feature-element-info", n = {
  base: t,
  heading: `${t}__heading`,
  description: `${t}__description`
};
class d extends l {
  constructor() {
    super(...arguments), this.headingLevel = 2;
  }
  static {
    this.properties = { description: 1, headingLevel: 9, heading: 1 };
  }
  static {
    this.styles = g;
  }
  render() {
    return i`<div class=${r(n.base)}>${this._renderHeading()}${this._renderDescription()}</div>`;
  }
  _renderHeading() {
    const { heading: e } = this;
    return e ? o({ class: n.heading, level: this.headingLevel, excludeGlobalCss: !0, children: i`<div .innerHTML=${s.sanitize(e) ?? ""}></div>` }) : null;
  }
  _renderDescription() {
    const { description: e } = this;
    return e ? i`<div class=${r(n.description)} .innerHTML=${s.sanitize(e) ?? ""}></div>` : null;
  }
}
a("arcgis-feature-element-info", d);
export {
  d as ArcgisFeatureElementInfo
};
