// @arcgis/portal-components/dist/chunks/requests.js
import n from "@arcgis/core/request.js";
import { once as c } from "@arcgis/core/core/reactiveUtils.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
async function f(s, t) {
  const i = `${s}/portals/self/classification/classificationSchema`, e = {};
  try {
    const a = (await n(i, { query: { f: "json" }, signal: t })).data;
    if (typeof a == "object" && a !== null && "classificationSchema" in a)
      e.result = a.classificationSchema;
    else
      throw new Error("Invalid response format received from the classification schema endpoint");
  } catch (r) {
    e.error = r;
  }
  return e;
}
async function d(s, t, i) {
  const e = `${s}/portals/self/classification/validateClassification`, r = { classificationValue: JSON.stringify(t), f: "json" }, a = {};
  try {
    const o = await n(e, { query: r, method: "post", signal: i.signal });
    a.result = o.data;
  } catch (o) {
    console.error(o), a.error = o;
  }
  return a;
}
async function y(s) {
  if (!s)
    return;
  s.loaded || await s.load();
  const t = s.portal.user ?? await c(() => s?.portal.user);
  if (!t)
    return;
  const i = `${s.portal.restUrl}/community/users/${t.username}/resources/classification_favorites`, e = {};
  try {
    const r = await n(i, { query: { f: "json" } });
    e.result = r.data;
  } catch (r) {
    console.error(r), e.error = r;
  }
  return e;
}
async function v(s, t) {
  if (!s)
    return;
  s.loaded || await s.load();
  const i = s.portal.user ?? await c(() => s?.portal.user);
  if (!i)
    return;
  const e = `${s.portal.restUrl}/community/users/${i.username}/addResource`, r = {}, a = {
    f: "json",
    key: "classification_favorites",
    text: JSON.stringify({ favorites: t }),
    access: "UserPrivateAllApps"
  };
  try {
    const o = await n(e, { query: a, method: "post" });
    r.result = o.data.success;
  } catch (o) {
    console.error("error: ", o), r.error = o;
  }
  return r;
}
export {
  f as a,
  y as r,
  v as u,
  d as v
};
