// @arcgis/charts-components/dist/components/arcgis-charts-config-line-chart-data/customElement.js
import { c as $ } from "../../chunks/runtime.js";
import { keyed as F } from "lit/directives/keyed.js";
import { a as c, i as b, m as d } from "../../chunks/series-types.js";
import { a as p } from "../../chunks/rest-js-object-literals.js";
import { LitElement as x, createEvent as g, renderElement as A, safeClassMap as s } from "@arcgis/lumina";
import { css as I, html as r } from "lit";
import { createRef as m, ref as h } from "lit/directives/ref.js";
import { u as f } from "../../chunks/useT9n.js";
import { L as a } from "../../chunks/interfaces3.js";
import { d as k, s as B } from "../../chunks/store.js";
import "@arcgis/toolkit/intl";
import "lodash-es";
import "d3-array";
import { bc as v, aX as T, aY as S, ci as u, cp as L } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import "@arcgis/toolkit/dom";
import { r as w } from "../../chunks/popover-ui-utils.js";
import { x as E } from "../../chunks/data.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const P = I`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}:host{display:flex;flex-direction:column;width:var(--arcgis-charts-config-calcite-block-default-width);margin:var(--arcgis-charts-config-calcite-block-default-margin);margin-top:var(--arcgis-charts-top-section-margin)}.label,.disabled-label{display:flex;margin-left:.5rem;margin-right:.5rem;margin-bottom:.5rem;justify-content:space-between}.time-binning-header{display:flex;font-weight:700;margin-left:.5rem;margin-right:.5rem;justify-content:space-between}.text{display:flex;margin-left:.5rem;width:calc(100% - 1.5rem);height:2rem;margin-bottom:.5rem;justify-content:space-between}.field-list{max-height:50vh;overflow-y:auto}.disabled-label{pointer-events:none}.fab{display:flex;margin-left:.5rem;margin-right:.5rem;margin-bottom:1.5rem;justify-content:center}.link-container{display:flex;justify-content:flex-end}`, n = {
  label: "label",
  fieldList: "field-list",
  fab: "fab",
  timeBinningHeader: "time-binning-header",
  linkContainer: "link-container",
  dimText: "dim-text"
};
class _ extends x {
  constructor() {
    super(...arguments), this._messages = f(), this.commonStrings = f({ name: "common", blocking: !0 }), this.lineChartPopover = null, this.aggregationElement = m(), this.xAxisFieldElement = m(), this.numericFieldsElement = m(), this.splitByFieldElement = m(), this.intervalAlignmentElement = m(), this.intervalElement = m(), this.onLineChartDataChange = async (e) => {
      try {
        if (e.detail !== null && e.detail?.value !== void 0) {
          const { contentKind: t } = e.detail, i = [...e.detail.value];
          switch (k(B), t) {
            case a.aggregation:
              this.model.aggregationType = i[0] !== c.NoAggregation ? i[0] : c.NoAggregation;
              break;
            case a.xAxisField: {
              const l = i[0]?.trim() || "";
              this.model.xAxisField = l, this.arcgisChartsConfigResetTimeInterval.emit();
              break;
            }
            case a.numericFields:
              this.model.numericFields = i, this.model.numericFields.length > 0 && this.model.aggregationType === c.Count && this.model.numericFields[0] !== v(this.layerFieldsInfo) && (this.model.aggregationType = c.Sum);
              break;
            case a.splitByField: {
              const l = i[0]?.trim() || "";
              this.model.splitByField = l, this.model.colorMatch = !0;
              break;
            }
            case a.interval:
              this.model.temporalBinningSize = i[0]?.intervalSize, this.model.temporalBinningUnit = i[0]?.intervalUnits, this.arcgisChartsConfigCustomizedTimeInterval.emit();
              break;
            case a.intervalAlignment:
              this.model.timeAggregationType = i[0] ?? "";
              break;
            default:
              break;
          }
        }
      } catch (t) {
        throw this.arcgisChartsConfigDataChangeError.emit({ error: t, model: this.model }), t;
      }
      this.removeLineChartDataPopover(), this.requestUpdate();
    }, this.onModelConfigChange = () => {
      this.requestUpdate();
    }, this.popoverClose = (e) => {
      this.lineChartPopover === e?.target && this.removeLineChartDataPopover();
    }, this.layerFieldsInfo = [], this.arcgisChartsConfigCustomizedTimeInterval = g(), this.arcgisChartsConfigDataChangeError = g(), this.arcgisChartsConfigLineChartDataNumericFieldValidate = g(), this.arcgisChartsConfigResetTimeInterval = g();
  }
  static {
    this.properties = { aggregationLabel: 16, timeAggregationLabel: 16, disableSplitBy: 16, messageOverrides: 0, model: 0, resetTimeInterval: 0, layerFieldsInfo: 0 };
  }
  static {
    this.styles = P;
  }
  load() {
    this.modelChange(this.model), this.onModelConfigChange();
  }
  willUpdate(e) {
    e.has("model") && this.modelChange(this.model, e.get("model")), this.aggregationLabel = T(this.model.aggregationType, this.commonStrings);
    const t = this.model.timeAggregationType;
    t !== void 0 && (this.timeAggregationLabel = S(t, this.commonStrings)), this.disableSplitBy = this.model.numericFields.length > 1;
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.removeLineChartDataPopover(), this.model.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  modelChange(e, t) {
    t?.removeEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange), e?.addEventListener("arcgisChartsModelConfigChange", this.onModelConfigChange);
  }
  buildNumericFieldItem(e) {
    return F(e, r`<calcite-list-item .label=${u(this.layerFieldsInfo, e)} .value=${e} closable @calciteListItemClose=${this.onRemoveNumericField}></calcite-list-item>`);
  }
  resetTimeIntervalOriginal(e) {
    this.popoverClose(e), this.model.resetTimeInterval !== void 0 && (this.model.resetTimeInterval(), this.arcgisChartsConfigResetTimeInterval.emit());
  }
  onAggregationSelect(e) {
    this.contentKind = a.aggregation, this.setupPopover(e);
  }
  onXAxisFieldSelect(e) {
    this.contentKind = a.xAxisField, this.setupPopover(e);
  }
  onSplitBySelect(e) {
    this.splitByFieldElement.value && !this.splitByFieldElement.value.disabled && (this.contentKind = a.splitByField, this.setupPopover(e));
  }
  onAddNumericFields(e) {
    this.contentKind = a.numericFields, this.setupPopover(e);
  }
  onRemoveNumericField(e) {
    const t = e.target.value;
    this.model.numericFields = this.model.numericFields.filter((i) => i !== t);
  }
  onIntervalAlignmentSelect(e) {
    this.intervalAlignmentElement.value && !this.intervalAlignmentElement.value.disabled && (this.contentKind = a.intervalAlignment, this.setupPopover(e));
  }
  onIntervalSelect(e) {
    this.intervalElement.value && !this.intervalElement.value.disabled && (this.contentKind = a.interval, this.setupPopover(e));
  }
  getCorrectedSplitByValue() {
    return this.model.splitByField === "" || this.model.numericFields.length > 1 ? "" : u(this.layerFieldsInfo, this.model.splitByField);
  }
  getCorrectedXAxisValue() {
    return E({
      fieldName: this.model.xAxisField,
      fieldList: this.layerFieldsInfo
    }) ? "" : u(this.layerFieldsInfo, this.model.xAxisField);
  }
  onShowDataLabelsChange(e) {
    this.model.dataLabelsVisibility = e.target.checked;
  }
  onTrimIncompleteIntervalChange(e) {
    this.model.trimIncompleteTimeInterval = e.target.checked;
  }
  onNullPolicyChange(e) {
    this.model.nullPolicy = e.target.selectedItem.value;
  }
  shouldShowTimeBinning() {
    const e = this.layerFieldsInfo.find((l) => l.name === this.model.xAxisField), t = e?.type === p.Date || e?.type === p.DateOnly || e?.type === p.TimestampOffset, i = this.model.aggregationType !== c.NoAggregation;
    return t && i;
  }
  setupPopover(e) {
    e.stopImmediatePropagation(), this.removeLineChartDataPopover();
    const t = {
      size: this.model.temporalBinningSize ?? 1,
      unit: this.model.temporalBinningUnit ?? b.Second
    };
    let i, l, o;
    switch (this.contentKind) {
      case a.aggregation:
        i = this.aggregationElement.value, l = this._messages.aggregationHeader, o = this.model.aggregationType;
        break;
      case a.xAxisField:
        i = this.xAxisFieldElement.value, l = this._messages.xAxisFieldHeader, o = this.model.xAxisField;
        break;
      case a.numericFields:
        i = this.numericFieldsElement.value, l = this._messages.numericFieldsHeader, o = this.model.numericFields;
        break;
      case a.splitByField:
        i = this.splitByFieldElement.value, l = this._messages.splitByFieldHeader, o = this.model.splitByField;
        break;
      case a.interval:
        i = this.intervalElement.value, l = this._messages.editInterval, o = t;
        break;
      case a.intervalAlignment:
        i = this.intervalAlignmentElement.value, l = this._messages.intervalAlignment, o = this.model.timeAggregationType ?? "";
        break;
    }
    this.lineChartPopover = A(r`<arcgis-charts-config-line-chart-popover .headingTitle=${l} .referenceElement=${i} open .contentKind=${this.contentKind} .layerFieldsInfo=${this.layerFieldsInfo} .selectedContent=${o} @arcgisChartsConfigLineChartDataPopoverChange=${this.onLineChartDataChange} @arcgisChartsConfigLineChartPopoverClose=${this.popoverClose}></arcgis-charts-config-line-chart-popover>`), !document.body.contains(this.lineChartPopover) && this.lineChartPopover !== null && document.body.appendChild(this.lineChartPopover);
  }
  removeLineChartDataPopover() {
    this.lineChartPopover !== null && (this.lineChartPopover.open = !1, this.lineChartPopover = w(this.lineChartPopover));
  }
  renderAggregation() {
    return r`<arcgis-charts-config-field-select .label=${this.model.aggregationType === c.NoAggregation ? this._messages.noAggregation : this.aggregationLabel} @click=${this.onAggregationSelect} ${h(this.aggregationElement)}></arcgis-charts-config-field-select>`;
  }
  renderXAxisField() {
    const e = this.getCorrectedXAxisValue();
    return r`<arcgis-charts-config-field-select .label=${e} @click=${this.onXAxisFieldSelect} ${h(this.xAxisFieldElement)}></arcgis-charts-config-field-select>`;
  }
  renderSplitBy() {
    const e = this.getCorrectedSplitByValue();
    return r`<arcgis-charts-config-field-select .label=${e} .disabled=${this.disableSplitBy} @click=${this.onSplitBySelect} ${h(this.splitByFieldElement)}></arcgis-charts-config-field-select>`;
  }
  renderNumericFields(e) {
    const t = v(this.layerFieldsInfo), i = [];
    for (const l of e)
      l !== t && i.push(this.buildNumericFieldItem(l));
    return r`<calcite-list label class=${s(n.fieldList)} ${h(this.numericFieldsElement)}>${i}</calcite-list>`;
  }
  renderInterval() {
    const e = (this.model.temporalBinningSize ?? 0) <= 0, t = this.model.temporalBinningSize ?? "", i = L(this.model.temporalBinningUnit ?? b.Second, this._messages) ?? "";
    return [
      r`<calcite-list label selection-appearance=border><calcite-list-item .value=${e ? "" : `${t} ${i}`} .label=${e ? "" : `${t} ${i}`} @click=${e ? () => {
      } : this.onIntervalSelect} ${h(this.intervalElement)}><calcite-icon slot=content-end icon=pencil></calcite-icon></calcite-list-item></calcite-list>`,
      r`<div class=${s(n.linkContainer)}><calcite-button appearance=transparent scale=s icon-end=reset @click=${this.resetTimeIntervalOriginal}>${this._messages.reset}</calcite-button></div>`
    ];
  }
  renderIntervalAlignment() {
    return r`<arcgis-charts-config-field-select .label=${this.timeAggregationLabel} @click=${this.onIntervalAlignmentSelect} ${h(this.intervalAlignmentElement)}></arcgis-charts-config-field-select>`;
  }
  renderEmptyBins() {
    return r`<calcite-radio-button-group name=empty-bins layout=vertical @calciteRadioButtonGroupChange=${this.onNullPolicyChange}><calcite-label layout=inline><calcite-radio-button .value=${d.Zero} .checked=${this.model.nullPolicy === d.Zero}></calcite-radio-button>${this._messages.treatAsZero}</calcite-label><calcite-label layout=inline><calcite-radio-button .value=${d.Interpolate} .checked=${this.model.nullPolicy === d.Interpolate}></calcite-radio-button>${this._messages.connectLine}</calcite-label><calcite-label layout=inline><calcite-radio-button .value=${d.Null} .checked=${this.model.nullPolicy === d.Null}></calcite-radio-button>${this._messages.breakLine}</calcite-label></calcite-radio-button-group>`;
  }
  render() {
    const e = this.renderAggregation(), t = this.renderXAxisField(), i = this.renderSplitBy(), l = this.renderNumericFields([...this.model.numericFields]), o = this.renderInterval(), y = this.renderIntervalAlignment(), C = this.renderEmptyBins();
    return r`<div><calcite-label class=${s(n.label)}>${this._messages.xAxisField}${t}</calcite-label><calcite-label class=${s(n.label)}>${this._messages.aggregation}${e}</calcite-label><calcite-label class=${s(n.label)}>${this._messages.numericFields}${l}</calcite-label><div class=${s(n.fab)}><calcite-fab appearance=outline-fill kind=neutral scale=m .label=${this._messages.selectNumericFields} text-enabled .text=${this._messages.selectNumericFields} icon=plus @click=${this.onAddNumericFields}></calcite-fab></div><calcite-label class=${s(n.label)}><span class=${s(this.disableSplitBy ? n.dimText : "")}>${this._messages.splitBy}</span>${i}</calcite-label><div .hidden=${!this.shouldShowTimeBinning()}><h3 class=${s(n.timeBinningHeader)}>${this._messages.timeBinningOptions}</h3><calcite-label class=${s(n.label)}>${this._messages.interval}${o}</calcite-label><calcite-label class=${s(n.label)}>${this._messages.intervalAlignment}${y}</calcite-label><calcite-label class=${s(n.label)} layout=inline-space-between>${this._messages.trimIncompleteInterval}<calcite-switch .checked=${this.model.trimIncompleteTimeInterval} @calciteSwitchChange=${this.onTrimIncompleteIntervalChange}></calcite-switch></calcite-label><calcite-label class=${s(n.label)}>${this._messages.emptyBins}${C}</calcite-label></div><calcite-label class=${s(n.label)} layout=inline-space-between>${this._messages.showDataLabels}<calcite-switch .checked=${this.model.dataLabelsVisibility} @calciteSwitchChange=${this.onShowDataLabelsChange}></calcite-switch></calcite-label></div>`;
  }
}
$("arcgis-charts-config-line-chart-data", _);
export {
  _ as ArcgisChartsConfigLineChartData
};
