// @arcgis/charts-components/dist/components/arcgis-charts-config-bar-chart-popover/customElement.js
import { c as T } from "../../chunks/runtime.js";
import { repeat as f } from "lit/directives/repeat.js";
import { keyed as v } from "lit/directives/keyed.js";
import { n as u, a as S, i as p } from "../../chunks/series-types.js";
import { a as h } from "../../chunks/rest-js-object-literals.js";
import { LitElement as F, createEvent as $, renderElement as x, safeStyleMap as P, safeClassMap as c } from "@arcgis/lumina";
import { substitute as A } from "@arcgis/core/intl.js";
import { css as D, html as a } from "lit";
import { createRef as b, ref as d } from "lit/directives/ref.js";
import { isEqual as k } from "lodash-es";
import { u as L } from "../../chunks/useT9n.js";
import { B as n } from "../../chunks/interfaces3.js";
import { S as m } from "../../chunks/common.js";
import "@arcgis/toolkit/intl";
import "d3-array";
import { aX as w, aY as E, ci as z, bf as I, bd as N, cp as B } from "../../chunks/interfaces.js";
import "@arcgis/core/geometry/support/jsonUtils.js";
import "@arcgis/core/rest/support/AttributeBinsQuery.js";
import "@arcgis/core/rest/support/Query.js";
import "@arcgis/core/rest/support/StatisticDefinition.js";
import "@arcgis/core/time/TimeExtent.js";
import "@arcgis/core/core/promiseUtils.js";
import "@arcgis/core/request.js";
import { d as K } from "../../chunks/index3.js";
import { e as _, p as O } from "../../chunks/popover-ui-utils.js";
import { h as C, U as y, i as g, d as U, e as j, s as H } from "../../chunks/chart-ui-utils2.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const M = D`.hide{display:none}.am5-modal{width:100%;height:100%;position:absolute;z-index:100000;top:0;left:0}.am5-modal-curtain{top:0;left:0;width:100%;height:100%;position:absolute;background:#fff!important;z-index:100}.am5-modal-wrapper{top:0;left:0;width:100%;height:100%;position:absolute;display:flex;align-items:center;justify-content:center;white-space:nowrap;background:#ffffff80;z-index:101}.am5-modal-content{display:inline-block;padding:1.2em;vertical-align:middle;text-align:start;white-space:normal;background:#fff;border-radius:4px;box-shadow:#00000073 0 0 36px;color:#000}.am5-layer-1000{z-index:1000!important}.am5-layer-30{z-index:100!important}.arcgis-charts-modal{box-shadow:none!important}.arcgis-charts-modal-header{background-color:#0000000d;font-weight:700;padding:4px;align-content:center}.show{display:block}.notifyPanel{flex:0 1 auto}.disable-interactions{pointer-events:none}.dim-text{color:var(--arcgis-charts-dim-text)}.name{display:flex;flex-direction:column;margin:.5rem;padding:.5rem;background:#fff;min-width:250px}.header{margin:0;font-weight:bolder}.fab{display:flex;justify-content:stretch}.pick-list-item-label{font-size:small}.sort-icon-color{color:#000}.sort-y-axis{transform:rotate(90deg)}.sort-order-pick-list{overflow-y:hidden}.panel{max-height:60vh}.panel .size-section{width:19rem;margin:1rem 0;display:flex;flex-wrap:wrap;justify-content:space-evenly}.panel .size-section calcite-label{width:8rem;margin-bottom:.25rem}.panel .apply-button{width:17rem}`, o = {
  jsAppFlyout: "js-app-flyout",
  name: "bar-chart-data-popover",
  numericFieldPickList: "bar-chart-data-numeric-fields-pick-list",
  header: "header",
  fab: "fab",
  pickListItemLabel: "pick-list-item-label",
  rotateClass: "sort-y-axis",
  sortOrderPickList: "sort-order-pick-list",
  sizeSection: "size-section",
  applyButton: "apply-button",
  panel: "panel"
};
class W extends F {
  constructor() {
    super(...arguments), this._messages = L(), this.commonStrings = L({ name: "common", blocking: !0 }), this.picklist = b(), this.intervalSizeElement = b(), this.intervalTypeElement = b(), this.popoverElement = b(), this.placement = "leading", this.offsetDistance = -200, this.pickListHasChanged = !1, this.aggregationLabels = [], this.layerFieldsInfo = [], this.isCustomSortDisabled = !1, this.isNoAggregationDisabled = !1, this.arcgisChartsConfigBarChartDataPopoverChange = $(), this.arcgisChartsConfigPopoverClose = $();
  }
  static {
    this.properties = { aggregationLabels: 16, timeAggregationLabels: 16, messageOverrides: 0, headingTitle: 3, referenceElement: 2, open: 7, contentKind: 1, layerFieldsInfo: 0, selectedContent: 1, isCustomSortDisabled: 5, isNoAggregationDisabled: 5 };
  }
  static {
    this.styles = M;
  }
  async reposition() {
    await this.popoverElement.value?.reposition();
  }
  willUpdate(e) {
    e.has("selectedContent") && this.selectedContentChange(this.selectedContent, e.get("selectedContent"));
    const t = C.map((s) => w(s, this.commonStrings));
    k(this.aggregationLabels, t) || (this.aggregationLabels = t);
    const l = [u.Start, u.End].map((s) => E(s, this.commonStrings));
    k(this.timeAggregationLabels, l) || (this.timeAggregationLabels = l);
  }
  updated() {
    setTimeout(async () => {
      await this.picklist.value?.setFocus();
    }, y.PopoverTimer), _(this.popoverElement.value, this.open), this.hasUpdated || setTimeout(() => {
      this.popoverElement.value?.reposition();
    }, y.PopoverTimer);
  }
  selectedContentChange(e, t) {
    this.pickListHasChanged = O(e, t);
  }
  buildAggregationListItems() {
    const e = this.isNoAggregationDisabled ? C.filter((t) => t !== S.NoAggregation) : C;
    return f(e, (t) => t, (t, i) => a`<calcite-list-item .label=${this.aggregationLabels[i]} .value=${t} .selected=${g(t, this.selectedContent)}></calcite-list-item>`);
  }
  createTooltip(e) {
    if (e.target?.label === this._messages.customSort && this.isCustomSortDisabled) {
      const t = e.target, i = K(t), s = new Intl.NumberFormat(i).format(y.customSortLimit), r = A(this._messages.customSortLimit ?? "", {
        totalLimit: s
      });
      this.tooltip = x(a`<calcite-tooltip style=${P({ opacity: 1, zIndex: "var(--calcite-floating-ui-z-index)" })} .referenceElement=${t} open placement=top>${r}</calcite-tooltip>`), document.body.appendChild(this.tooltip);
    }
  }
  destroyTooltip() {
    U(this.tooltip);
  }
  buildSortOrder() {
    const e = Object.keys(m);
    let t;
    return f(e, (i) => i, (i) => {
      let l = "";
      switch (i) {
        case m.xAxisAsc: {
          t = "a-z-down";
          break;
        }
        case m.xAxisDesc: {
          t = "a-z-up";
          break;
        }
        case m.yAxisAsc: {
          t = "sort-descending", l = o.rotateClass;
          break;
        }
        case m.yAxisDesc: {
          t = "sort-ascending", l = o.rotateClass;
          break;
        }
        case m.customSort: {
          t = "arrow-up-down";
          break;
        }
      }
      return a`<calcite-list-item .disabled=${this.isCustomSortDisabled && i === m.customSort} .label=${this._messages[i]} .selected=${this.selectedContent === i} .value=${m[i]} @mouseover=${this.createTooltip} @mouseout=${this.destroyTooltip}><calcite-icon class=${`sort-icon-color ${l}`} slot=content-start .icon=${t}></calcite-icon></calcite-list-item>`;
    });
  }
  buildPickListItem(e) {
    return v(e.name, a`<calcite-list-item .label=${z(this.layerFieldsInfo, e.name)} .value=${e.name} .selected=${g(e.name, this.selectedContent)}></calcite-list-item>`);
  }
  buildPickListItems(e) {
    const t = this.layerFieldsInfo?.length ?? 0, i = [];
    switch (e) {
      case n.numericFields: {
        for (let l = 0; l < t; l += 1) {
          const s = this.layerFieldsInfo?.[l];
          s && N(s) && i.push(this.buildPickListItem(s));
        }
        break;
      }
      case n.category: {
        const l = v(" ", a`<calcite-list-item label=" " value=" " .selected=${g("", this.selectedContent)}></calcite-list-item>`);
        i.push(l);
        for (let s = 0; s < t; s += 1) {
          const r = this.layerFieldsInfo?.[s];
          r && (I(r) || r.type === h.String || r.type === h.Date || r.type === h.DateOnly || r.type === h.TimestampOffset) && i.push(this.buildPickListItem(r));
        }
        break;
      }
      case n.splitByField: {
        const l = v(" ", a`<calcite-list-item label=" " value=" " .selected=${g("", this.selectedContent)}></calcite-list-item>`);
        i.push(l);
        for (let s = 0; s < t; s += 1) {
          const r = this.layerFieldsInfo?.[s];
          r && (I(r) || r.type === h.String) && i.push(this.buildPickListItem(r));
        }
        break;
      }
    }
    return i;
  }
  closePopover() {
    if (this.contentKind === n.numericFields && this.pickListHasChanged) {
      const e = {
        pickListElement: this.picklist.value,
        eventEmitter: this.arcgisChartsConfigBarChartDataPopoverChange,
        contentKind: this.contentKind
      };
      j(e);
    }
    this.open = !1, this.arcgisChartsConfigPopoverClose.emit();
  }
  onDataContentTypeChange(e) {
    const t = e.target.selectedItems;
    this.arcgisChartsConfigBarChartDataPopoverChange.emit({
      contentKind: this.contentKind,
      value: t.map((i) => i.value)
    }), this.open = !1;
  }
  intervalChange() {
    const e = Number.parseInt(this.intervalSizeElement.value?.value ?? ""), t = this.intervalTypeElement.value?.selectedOption?.value;
    !Number.isNaN(e) && e > 0 && this.arcgisChartsConfigBarChartDataPopoverChange.emit({
      contentKind: this.contentKind,
      value: [{ intervalSize: e, intervalUnits: t }]
    }), this.open = !1;
  }
  onNumericFieldsChange(e) {
    e.target !== null && (this.selectedContent = e.target.selectedItems.map((t) => t.value));
  }
  onNumericFieldSelectionDone() {
    this.pickListHasChanged && this.closePopover();
  }
  buildIntervalAlignmentPickListItems() {
    const e = [u.Start, u.End];
    return f(e, (t) => t, (t, i) => a`<calcite-list-item .label=${this.timeAggregationLabels?.[i]} .value=${t} .selected=${g(t, this.selectedContent)}></calcite-list-item>`);
  }
  renderAggregationType() {
    const e = this.buildAggregationListItems();
    return a`<calcite-list label class=${c(o.pickListItemLabel)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderCategory() {
    const e = this.buildPickListItems(n.category);
    return a`<calcite-list label class=${c(o.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${d(this.picklist)}>${e}</calcite-list>`;
  }
  renderNumericFields() {
    const e = this.buildPickListItems(n.numericFields);
    return a`<calcite-list label class=${c(o.numericFieldPickList)} selection-mode=multiple filter-enabled @calciteListChange=${this.onNumericFieldsChange} ${d(this.picklist)}>${e}</calcite-list>`;
  }
  renderSortOrder() {
    const e = this.buildSortOrder();
    return a`<calcite-list label class=${c(o.sortOrderPickList)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderSplitByField() {
    const e = this.buildPickListItems(n.splitByField);
    return a`<calcite-list label class=${c(o.pickListItemLabel)} selection-mode=single selection-appearance=border filter-enabled @calciteListChange=${this.onDataContentTypeChange} ${d(this.picklist)}>${e}</calcite-list>`;
  }
  renderIntervalAlignment() {
    const e = this.buildIntervalAlignmentPickListItems();
    return a`<calcite-list label class=${c(o.pickListItemLabel)} selection-mode=single selection-appearance=border @calciteListChange=${this.onDataContentTypeChange}>${e}</calcite-list>`;
  }
  renderInterval() {
    const e = this.selectedContent, t = [
      p.Second,
      p.Minute,
      p.Hour,
      p.Day,
      p.Week,
      p.Month,
      p.Year
    ];
    return a`<div><div class=${c(o.sizeSection)}><calcite-label scale=m>${this._messages.intervalSize}<calcite-input type=number scale=m min=1 step=any .value=${H(e?.size, 1)} ${d(this.intervalSizeElement)}></calcite-input></calcite-label><calcite-label scale=m>${this._messages.intervalType}<calcite-select .label=${this._messages.intervalType ?? ""} ${d(this.intervalTypeElement)}>${t.map((i) => a`<calcite-option .value=${i} .label=${B(i, this._messages)} .selected=${e?.unit === i}></calcite-option>`)}</calcite-select></calcite-label><calcite-button class=${c(o.applyButton)} appearance=outline @click=${this.intervalChange}>${this._messages.apply}</calcite-button></div></div>`;
  }
  renderPopoverInfo() {
    let e;
    switch (this.contentKind) {
      case n.aggregation:
        e = this.renderAggregationType();
        break;
      case n.category:
        e = this.renderCategory();
        break;
      case n.numericFields:
        e = this.renderNumericFields();
        break;
      case n.splitByField:
        e = this.renderSplitByField();
        break;
      case n.sortOrder:
        e = this.renderSortOrder();
        break;
      case n.intervalAlignment:
        e = this.renderIntervalAlignment();
        break;
      case n.interval:
        e = this.renderInterval();
        break;
    }
    return e;
  }
  render() {
    let e;
    return this.contentKind === n.numericFields && (e = a`<div slot=footer><calcite-fab class=${c(o.fab)} appearance=outline-fill kind=neutral icon .label=${this._messages.selectionDone} scale=s text-enabled .text=${this._messages.selectionDone} @click=${this.onNumericFieldSelectionDone}></calcite-fab></div>`), a`<div class=${c(o.jsAppFlyout)}><calcite-popover class=${c(o.name)} .referenceElement=${this.referenceElement ?? document.body} .placement=${this.placement} .offsetDistance=${this.offsetDistance} .open=${this.open} @calcitePopoverClose=${this.closePopover} label auto-close ${d(this.popoverElement)}><calcite-panel class=${c(o.panel)} closable .closed=${!this.open} @calcitePanelClose=${this.closePopover}><div slot=header-content class=${c(o.header)}>${this.headingTitle ?? ""}</div>${this.renderPopoverInfo()}${e}</calcite-panel></calcite-popover></div>`;
  }
}
T("arcgis-charts-config-bar-chart-popover", W);
export {
  W as ArcgisChartsConfigBarChartPopover
};
