// @arcgis/coding-components/dist/components/arcgis-arcade-coding-assistant/customElement.js
import { c as V } from "../../chunks/runtime.js";
import { nothing as w, LitElement as L } from "@arcgis/lumina";
import { ref as H, createRef as B } from "lit/directives/ref.js";
import j from "@arcgis/core/identity/IdentityManager.js";
import { u as W } from "../../chunks/useT9n.js";
import { b as I } from "../../chunks/arcade-defaults.js";
import { css as Q, html as r } from "lit";
import T from "@arcgis/core/request.js";
import { e as G, a as J } from "../../chunks/monaco-importer.js";
import { repeat as U } from "lit/directives/repeat.js";
import "@arcgis/toolkit/function";
import { generateGuid as K } from "@arcgis/toolkit/string";
import { unsafeHTML as $ } from "lit/directives/unsafe-html.js";
import Y from "@arcgis/core/layers/FeatureLayer.js";
import X from "@arcgis/core/config.js";
import Z from "@arcgis/core/Graphic.js";
import { b as ee } from "../../chunks/language-defaults-base.js";
/*! All material copyright Esri, All Rights Reserved, unless otherwise specified.
See https://js.arcgis.com/4.34/esri/copyright.txt for details.
v4.34.3 */
const te = Q`:host(:not([closed])){container-type:size;--calcite-block-border-color: var(--calcite-color-transparent);flex:1 0 auto;border-inline-start:var(--arcgis-coding-components-border);box-sizing:border-box;max-width:var(--arcgis-coding-components-code-editor-side-panel-max-w);width:var(--arcgis-coding-components-code-editor-side-panel-w);.submission-actions-wrapper{display:flex;flex-direction:column;width:100%;gap:.75rem}.submission-actions{display:flex;flex-direction:row;gap:.5rem;justify-content:space-between}.gaps{display:flex;gap:.25rem}.align-inline-end{margin-inline-start:auto}.action .error-display{display:flex;flex-direction:row;gap:.5rem;justify-content:space-between}.context-popover p{padding:0 .5rem;font-size:var(--calcite-font-size-sm)}.feedback-content{padding:1rem;display:flex;flex-direction:column;gap:.75rem;max-height:30vh;overflow-y:auto}.feedback-sheet{--calcite-sheet-height: auto;--calcite-sheet-max-height: 100%;--calcite-sheet-min-height: auto}calcite-chip[slot=header-actions-end]{align-self:center;margin-inline-end:.5rem}calcite-flow-item calcite-shell{background:transparent}calcite-block{width:100%;max-width:100%;align-self:flex-end;background-color:transparent;margin-block:0}calcite-sheet{--calcite-sheet-scrim-background: rgba(200, 200, 200, .5)}.type-suggestion{--calcite-color-foreground-1: var(--calcite-color-foreground-current);--calcite-color-text-3: var(--calcite-color-text-2);--calcite-icon-color: var(--calcite-color-brand)}.type-history{--calcite-icon-color: var(--calcite-color-text-2)}.type-response{--calcite-icon-color: var(--calcite-color-brand);--calcite-font-family: monospace}.block-response{align-self:flex-start;margin-block-end:4rem;border-block-end:0;--calcite-block-text-color: var(--calcite-color-brand)}calcite-list-item calcite-button{margin-inline-end:.25rem}.type-docs{--calcite-color-foreground-1: var(--calcite-color-foreground-3);--calcite-icon-color: var(--calcite-color-text-1)}.type-error{--calcite-icon-color: var(--calcite-color-status-danger);--calcite-font-family: monospace;border:1px solid var(--calcite-color-status-danger)}#thumbs-up{margin-inline-start:auto;--calcite-icon-color: var(--calcite-color-border-input)}#thumbs-down{--calcite-icon-color: var(--calcite-color-border-input)}calcite-flow-item calcite-shell calcite-notice{margin-block:1rem}.no-space-block,.no-space-block[open]{margin:0}calcite-block[open]:last-of-type{border-block-end:0}.text-area-wrapper{display:flex;border-radius:var(--calcite-corner-radius-round);margin:.75rem;overflow:hidden;--calcite-color-border-input: transparent;position:sticky;box-shadow:var(--calcite-shadow-sm);flex:none}calcite-text-area{height:auto;min-height:5.75rem;overflow:hidden;border-radius:var(--calcite-corner-radius-round) var(--calcite-corner-radius-round) 0 0;--calcite-font-weight-normal: 500;font-weight:500;line-height:1.2}calcite-tooltip{--calcite-tooltip-z-index: 1000}.standalone-list{margin:1rem;width:calc(100% - 2rem)}.result-disclaimer{display:block;width:24rem;max-width:90%;margin:1rem auto;text-align:center;line-height:1.125rem;font-size:.875rem;font-style:italic}calcite-flow{height:100%}calcite-popover calcite-list{--calcite-list-background-color: transparent;--calcite-input-border-color: transparent}calcite-popover calcite-list calcite-filter{--calcite-spacing-sm: 0}.contained-popover-or-tooltip{--calcite-popover-max-size-x: 100%;--calcite-tooltip-max-size-x: 100%}.break-spaces{white-space:break-spaces}.feedback-button-container{margin-inline-start:auto;gap:0}.info-heading{font-size:var(--calcite-font-size-1);color:var(--calcite-color-text-1)}.unstyled-h3{margin:unset;font-size:var(--calcite-font-size-md)}.unstyled-shell{position:unset;inset:unset;display:unset;block-size:unset;inline-size:unset}.suggestion-button{font-weight:var(--calcite-font-weight-medium)}.info-block{max-width:450px}#result-block{calcite-card-group{--calcite-card-group-gap: 1.25rem}calcite-card{--calcite-card-corner-radius: var(--calcite-corner-radius-round);--calcite-card-border-color: transparent;width:100%}.positioned-code-content{display:flex;flex-direction:row;gap:.5rem;padding:.5rem}calcite-card [slot=title] calcite-chip{margin-block-start:.25rem}.code-response-wrapper{position:relative;background:var(--calcite-color-foreground-2);border-radius:var(--calcite-corner-radius-round);margin-top:-1.5rem}.prior-prompt{display:flex;flex-direction:row;padding:0;overflow-y:hidden;overflow-x:auto;margin:.25rem 0 .75rem;align-items:center;max-width:100%;span{overflow-x:auto;overflow-y:hidden}}.copy-feedback{display:none}.copy-prompt-button{align-self:flex-start;margin-inline-start:auto}.response-secondary-action-button[data-copied]+.copy-feedback,.copy-prompt-button[data-copied]+.copy-feedback{display:block;span{padding:.75rem 1rem;font-size:var(--calcite-font-size--2);line-height:1.375;color:var(--calcite-color-text--2);font-weight:var(--calcite-font-weight-medium)}}.copy-prompt-button[data-copied]~.copy-tooltip,.response-secondary-action-button[data-copied]~.copy-tooltip{display:none}code{white-space:pre;font-size:12px;margin:0;overflow-x:auto;font-family:Fira Mono,Consolas,Menlo,monospace;display:block;max-height:12rem;position:relative;padding:.5rem}.code-collapse{margin-inline-start:auto}.response-secondary-action-button{margin-left:.5rem}.add-to-editor-button{--calcite-internal-button-border-color: unset}}.collapsible-code{padding:.5rem}.collapsible-code summary{cursor:pointer;font-weight:500;color:var(--calcite-color-brand);outline:none;padding:.25rem 0}.collapsible-code[open] summary{color:var(--calcite-color-text-2)}.collapsible-error{padding:.5rem}.collapsible-error summary{cursor:pointer;font-weight:500;color:var(--calcite-color-status-danger);outline:none;padding:.25rem 0}.collapsible-error[open] summary{color:var(--calcite-color-text-2)}.response-error-container{padding:.5rem;background:color-mix(in srgb,var(--calcite-color-status-warning) 10%,transparent);border-radius:var(--calcite-corner-radius-round);margin:1rem 0;font-size:var(--calcite-font-size--2);--calcite-input-message-icon-color: var(--calcite-color-status-warning)}.response-error-chip{--calcite-internal-chip-background-color: var(--calcite-color-status-warning);--calcite-chip-icon-color: var(--calcite-color-foreground-1);--calcite-chip-border-color: var(--calcite-color-transparent)}.feedback-actions{gap:unset}.contained{max-width:100%;max-height:100%;max-width:100cqi;max-height:75cqb;overflow-y:auto}}:host([closed]){display:none}`;
class y extends Error {
  constructor({
    message: e,
    code: t,
    subCode: o,
    details: c,
    source: n
  }) {
    super(e), this.name = "ArcadeAssistantError", this.code = t, this.subCode = o, this.details = c, this.source = n;
  }
}
function E(i) {
  if (!i)
    return;
  const e = i.indexOf('{"error":');
  if (e !== -1)
    try {
      const t = i.slice(e), o = JSON.parse(t);
      if (o && typeof o == "object" && o.error)
        return o.error;
    } catch {
      return;
    }
}
async function ie(i) {
  const { baseUrl: e, skillId: t, message: o, authToken: c, previousConversationId: n, context: p } = i, d = t ? `${e}/skills/${t}/chat` : `${e}/chat`, s = [], h = {
    message: o,
    context: p
  };
  n && (h.conversation_id = n);
  let a;
  try {
    a = await T(d, {
      method: "post",
      headers: {
        "Content-Type": "application/json",
        token: c
      },
      body: JSON.stringify(h),
      responseType: "json"
    });
  } catch (f) {
    const b = f, k = b.details?.rawResponse || b.message || String(f), g = E(k);
    throw g ? new y(g) : f;
  }
  const l = a.data;
  s.push(l);
  const u = l.conversationId, D = l.inquiryId;
  let A = l.sequenceNumber;
  async function _() {
    let f = 0;
    const b = 25;
    for (; f < b; ) {
      await new Promise((v) => setTimeout(v, 1e3));
      const k = oe(u, A, D);
      let g;
      try {
        g = await T(d, {
          method: "post",
          headers: {
            "Content-Type": "application/json",
            token: c
          },
          body: JSON.stringify(k),
          responseType: "json"
        });
      } catch (v) {
        const C = v, O = C.details?.rawResponse || C.message || String(v), S = E(O);
        throw S ? new y(S) : v;
      }
      const m = g.data;
      if (m.context && m.context.kind === "ArcgisErrorAsContext" && m.context.error)
        throw new y({
          message: m.context.error.message || "ArcGIS Assistant error",
          code: m.context.error.code,
          subCode: m.context.error.subCode,
          details: m.context.error.details,
          source: "polling-response"
        });
      if (s.push(m), A = m.sequenceNumber, f++, !m.hasMore)
        break;
    }
    if (f === b)
      throw new y({
        message: "Request timed out",
        code: "TIMEOUT",
        source: "polling"
      });
  }
  const z = 2 * 60 * 1e3;
  return await Promise.race([
    _(),
    new Promise(
      (f, b) => setTimeout(
        () => b(
          new y({
            message: "Request timed out",
            code: "TIMEOUT",
            source: "hard-timeout"
          })
        ),
        z
      )
    )
  ]), s;
}
function oe(i, e, t) {
  return {
    conversationId: i,
    inquiryId: t,
    ackSequenceNumber: e
  };
}
function ae(i) {
  return i.message.match(/Error in prompt shield/u) ? "I'm having trouble with this request. Try again by rephrasing the question." : i.code === "TIMEOUT" ? "I'm having trouble with this request. Try again by re-asking the same question or rephrasing it." : i.message;
}
const se = (i) => {
  const e = i.value.length, t = 72, o = 180, c = i.offsetWidth, p = Math.max(1, Math.floor(c / 8)), d = Math.ceil(e / p), s = t + (d - 2) * 24;
  i.style.setProperty("--calcite-text-area-min-height", `${Math.min(s, o)}px`);
};
function ce({ messages: i }) {
  return r`<calcite-chip slot=header-actions-end scale=s appearance=outline-fill .label=${i.beta ?? "Beta"}>${i.beta ?? "Beta"}</calcite-chip>`;
}
function re({ popoverContent: i }) {
  return r`<calcite-action icon=information slot=header-actions-end id=legal-disclaimer text></calcite-action><calcite-popover reference-element=legal-disclaimer label=legal-disclaimer placement=bottom auto-close focus-trap-disabled><div class="contained">${i}</div></calcite-popover>`;
}
function F({ errorMessage: i, slot: e }) {
  return r`<div class="error-display"><calcite-input-message id=error-message status=invalid icon=exclamation-mark-triangle scale=m slot=${e ?? w}>${i}</calcite-input-message></div>`;
}
function ne({ active: i, setActive: e, messages: t }) {
  return r`<calcite-alert slot=alerts .open=${i} placement=bottom kind=success icon=check-circle-f auto-close @calciteAlertClose=${() => e(!1)} .label=${t.expressionapplied ?? "Expression applied"}><div slot=message>${t.expressionapplied ?? "Expression applied to editor."}</div></calcite-alert>`;
}
function le({ context: i, messages: e }) {
  return i ? r`<div class="gaps">${i.profileName && r`<calcite-popover auto-close placement=top .label=${(e.profilecontextlabel ?? "Profile context: {profileName}").replace("{profileName}", i.profileName)} reference-element=profile-context-button class="context-popover contained-popover-or-tooltip"><p>${(e.profilecontextlabel ?? "Profile context: {profileName}").replace("{profileName}", i.profileName)}</p></calcite-popover><calcite-button round kind=neutral icon-start=map-information id=profile-context-button></calcite-button>` || ""}${i.layerName && r`<calcite-popover auto-close placement=top .label=${(e.layercontextlabel ?? "Layer context: {layerName}").replace("{layerName}", i.layerName)} reference-element=layer-context-button class="context-popover contained-popover-or-tooltip"><p>${(e.layercontextlabel ?? "Layer context: {layerName}").replace("{layerName}", i.layerName)}</p></calcite-popover><calcite-button round kind=neutral icon-start=layers id=layer-context-button></calcite-button>` || ""}</div>` : null;
}
function de({ isDisabled: i, isReadOnly: e, setQuestion: t, setIsViewingResult: o, textAreaRef: c, question: n, errorMessage: p, setErrorMessage: d, submitQuestion: s, messages: h, mode: a, context: l }) {
  return r`<div class="text-area-wrapper"><calcite-text-area resize=vertical .readOnly=${i || e} class="styled-text-area" .placeholder=${h.askaquestion ?? `Specify the calculation or logic you need in an Arcade expression. (e.g., 'Return "Yes" if value > 10')`} .value=${n} @calciteTextAreaInput=${(u) => {
    se(u.target), t(u.target.value);
  }} ${H(c)}><div class="submission-actions-wrapper" slot=footer-start>${p && F({ errorMessage: p, slot: "footer-start" }) || ""}<div class="submission-actions">${le({ context: l, messages: h })}<div class="align-inline-end gaps"><calcite-button round kind=neutral .hidden=${a !== "refine"} icon-start=arrow-left icon-flip-rtl=both appearance=solid .disabled=${i} @click=${async () => {
    o(!1), t("");
  }}>${h.startover ?? "Start Over"}</calcite-button><calcite-button round icon-end=effects appearance=solid .loading=${i} .disabled=${i} @click=${async () => {
    d(n ? "" : h.enteraprompt ?? "Please enter a prompt"), await s();
  }}>${a === "prompt" ? h.ask ?? "Generate" : h.refine ?? "Refine"}</calcite-button></div></div></div></calcite-text-area></div>`;
}
const pe = "https://www.esri.com/en-us/privacy/privacy-statements/privacy-statement", ue = "https://www.esri.com/en-us/privacy/privacy-statements/privacy-supplement";
function he({ setOpen: i, messages: e, feedback: t, setFeedback: o, onSubmit: c }) {
  return t.active ? r`<calcite-sheet class="feedback-sheet" .label=${e.feedback ?? "Feedback"} slot=sheets display-mode=float position=block-end .open=${t.active} @calciteSheetClose=${() => i(!1)}><calcite-panel .heading=${e.providefeedback ?? "Provide feedback"} .description=${e.experiencehelps ?? "Your experience helps guide future improvements"}><div class="feedback-content"><calcite-label>${e.experiencehelpful ?? "Was the assistant helpful?"}<calcite-segmented-control scale=l .disabled=${t.submitting}><calcite-segmented-control-item icon-start=thumbs-up .checked=${t.value === "good"} @click=${() => o({ ...t, value: "good" })}></calcite-segmented-control-item><calcite-segmented-control-item icon-start=thumbs-down .checked=${t.value === "bad"} @click=${() => o({ ...t, value: "bad" })}></calcite-segmented-control-item></calcite-segmented-control></calcite-label><calcite-label>${e.describeexperience ?? "Describe your experience"}<calcite-text-area resize=vertical .value=${t.text} @calciteTextAreaInput=${(n) => o({ ...t, text: n.target.value })} .disabled=${t.submitting}></calcite-text-area></calcite-label><calcite-label layout=inline scale=s><calcite-checkbox scale=l .checked=${t.termsAccepted} @calciteCheckboxChange=${(n) => {
    o({ ...t, termsAccepted: n.target.checked });
  }} .disabled=${t.submitting}></calcite-checkbox><span>${$((e.agreement ?? "I agree to the {EsriPrivacyStatement} (and {Supplement}).").replace("{EsriPrivacyStatement}", `<calcite-link href=${pe} target="_blank">${e.esriprivacystatement ?? "Esri Privacy Statement"}</calcite-link>`).replace("{Supplement}", `<calcite-link href=${ue} target="_blank">${e.supplement ?? "Supplement"}</calcite-link>`))}</span></calcite-label></div><calcite-button slot=footer width=full @click=${c} .disabled=${!t.termsAccepted || t.submitting} .loading=${t.submitting}>${e.submitfeedback ?? "Submit Feedback"}</calcite-button></calcite-panel></calcite-sheet>` : null;
}
function R({ assistantsEnabled: i, messages: e, slot: t, assistantHelpUrl: o, helpTopicUrl: c }) {
  return r`<calcite-block .label=${e.assistantinformation ?? "Assistant information"} slot=${t ?? w} expanded class="info-block"><calcite-chip .label=${e.new ?? "New"} kind=brand>${e.new ?? "New"}</calcite-chip><h1 class="info-heading">${e.introducing ?? "Introducing Arcade Assistant (beta)"}</h1><p>${e.capabilities ?? "The assistant uses artificial intelligence to generate Arcade expressions."}</p>${i ? r`<p>${e.keepinmind ?? "Keep in mind the following:"}<ul><li>${e.beconsise ?? "Be concise, but also clear in your request"}</li><li>${e.avoidambiguity ?? "Avoid ambiguity and vagueness"}</li><li>${e.moredetails ?? "you do not need to use exact field names in your instructions, but more detail is generally better"}</li></ul>${$((e.seedocumentation ?? "Please see the {documentation} for more information, including limitations and additional resources.").replace("{documentation}", `<calcite-link href=${c} target="_blank">${e.documentation ?? "documentation"}</calcite-link>`))}</p>` : r`<p>${$(e.useofassistant?.replace("{aiassistants}", `<strong>${e.aiassistants ?? "Ai Assistants"}</strong>`))}<calcite-link .href=${o} target=_blank icon-end=launch-2>${e.configureassistants ?? "Configure Assistants"}</calcite-link></p>`}</calcite-block>`;
}
function me({ messages: i, onProceed: e, onExit: t, helpTopicUrl: o }) {
  return r`<calcite-dialog close-disabled placement=top .heading=${i.welcome ?? "Welcome to the new assistant experience"} slot=dialogs open width-scale=m id=welcome-dialog><h3 class="unstyled-h3">${i.beforeyougetstarted ?? "Before you get started be aware:"}</h3><ul><li>${i.maybeincorrect ?? "AI generated code may occasionally be incorrect or biased. Review carefully."}</li><li>${$((i.pleasereadhelp ?? "Read the Arcade Assistant (beta) {helpTopic} to get started and learn details on security, privacy and limitations.").replace("{helpTopic}", `<calcite-link href=${o} target="_blank">${i.helptopic ?? "help topic"}</calcite-link>`))}</li></ul><calcite-button id=welcome-proceed slot=footer-end @click=${e}>${i.proceed ?? "Proceed"}</calcite-button><calcite-button id=welcome-decline slot=footer-start @click=${t} kind=neutral>${i.exit ?? "Exit"}</calcite-button></calcite-dialog>`;
}
function fe({ messages: i, suggestions: e, onSuggestionClick: t }) {
  return r`<calcite-block .heading=${i.promptsuggestions} expanded collapsible id=suggestion-block><calcite-chip-group class="suggestions" .label=${i.promptsuggestions ?? "Prompt Suggestions"}>${U(e, (o) => o, (o) => r`<calcite-button round appearance=outline-fill kind=neutral .label=${i.promptsuggestion} class="suggestion-button" @click=${() => t?.(o)}>${o}</calcite-button>`)}</calcite-chip-group></calcite-block>`;
}
const N = 200;
function P({ hidden: i = !1, expanded: e = !1, messages: t, cards: o, heading: c, collapsible: n = !1, showEffectsIcon: p = !1, onCopyCode: d, standalone: s }) {
  const h = t.addtoeditor ?? "Add to editor";
  return r`<calcite-block .hidden=${i} id=result-block .heading=${c ?? t.besteffort ?? "Here is the assistant's best effort"} .expanded=${e} .collapsible=${n}>${p && r`<calcite-chip .label=${t.assistantresponses ?? "Assistant responses"} slot=content-start icon=effects scale=m class="suggestion-chips"></calcite-chip>` || ""}<calcite-card-group .label=${t.assistantresponses ?? "Assistant responses"}>${U(o, (a) => a.conversationId, (a) => r`<calcite-card><div slot=heading><div class="prior-prompt"><span>${a.priorPrompt.length > N ? `${a.priorPrompt.slice(0, N)}...` : a.priorPrompt}</span><calcite-action icon=duplicate scale=s slot=actions-end class="copy-prompt-button" text .id=${`copy-prompt-${a.conversationId}-${s ? "standalone" : "list"}`} @click=${(l) => {
    const u = l.currentTarget;
    navigator.clipboard.writeText(a.priorPrompt).then(() => {
      u.setAttribute("data-copied", "true"), setTimeout(() => u.removeAttribute("data-copied"), 1500);
    });
  }}></calcite-action><calcite-popover .referenceElement=${`copy-prompt-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start class="copy-feedback" .label=${t.copiedsuccessfully ?? "copied successfully"} open trigger-disabled><span>${t.copiedsuccessfully ?? "copied successfully"}</span></calcite-popover><calcite-tooltip .referenceElement=${`copy-prompt-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start class="copy-tooltip">${t.copyprompttoclipboard ?? "Copy prompt to clipboard"}</calcite-tooltip></div>${a.message && s && r`<div class="response-error-container">${F({ errorMessage: a.message })}</div>` || ""}</div><div><div class="code-response-wrapper"><div class="positioned-code-content"><calcite-chip-group .label=${t.responsecontext ?? "Response Context"} scale=s>${a.message && !s && r`<calcite-tooltip .referenceElement=${`response-chip-${a.conversationId}-error`} placement=top-start>${a.message}</calcite-tooltip><calcite-chip .id=${`response-chip-${a.conversationId}-error`} .label=${t.error ?? "Error"} scale=s appearance=outline icon=exclamation-mark-triangle-f class="response-error-chip"></calcite-chip>` || ""}${a.chips?.map((l) => {
    const u = K();
    return r`${l.tooltip && r`<calcite-tooltip .referenceElement=${`response-chip-${l.conversationId}-${u}`} placement=top-start class="contained-popover-or-tooltip">${l.tooltip}</calcite-tooltip>` || ""}<calcite-chip .label=${l.label} .scale=${l.scale ?? "s"} .appearance=${l.appearance ?? "outline"} .icon=${l.icon} .id=${`response-chip-${l.conversationId}-${u}`}>${l.label}</calcite-chip>`;
  })}</calcite-chip-group></div><details class="collapsible-code" open><summary>Code response</summary><code>${a.code}</code></details></div></div><div slot=footer-start class="feedback-actions"><calcite-button primary-text=${h ?? w} class="add-to-editor-button" @click=${() => a.onAddToEditor("add")} scale=s icon-start=plus round>${t.addtoeditor ?? "Add to editor"}</calcite-button><calcite-button round icon-start=file-code kind=neutral scale=s @click=${() => a.onAddToEditor("replace")} class="response-secondary-action-button" .id=${`replace-${a.conversationId}-${s ? "standalone" : "list"}`}></calcite-button><calcite-tooltip .referenceElement=${`replace-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start><span>${t.replaceineditor ?? "Replace all code in editor"}</span></calcite-tooltip><calcite-button round icon-start=copy-to-clipboard kind=neutral scale=s @click=${(l) => {
    const u = l.currentTarget;
    d(a.code).then(() => {
      u.setAttribute("data-copied", "true"), setTimeout(() => u.removeAttribute("data-copied"), 1500);
    });
  }} class="response-secondary-action-button" .id=${`copy-response-${a.conversationId}-${s ? "standalone" : "list"}`}></calcite-button><calcite-popover .referenceElement=${`copy-response-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start class="copy-feedback" .label=${t.copiedsuccessfully ?? "copied successfully"} open trigger-disabled><span>${t.copiedsuccessfully ?? "copied successfully"}</span></calcite-popover><calcite-tooltip .referenceElement=${`copy-response-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start class="copy-tooltip">${t.copytoclipboard ?? "Copy code to clipboard"}</calcite-tooltip></div><div slot=footer-end class="feedback-actions"><calcite-action .text=${t.thumbsup ?? "Thumbs up"} scale=s icon=thumbs-up @click=${a.onThumbsUp} .id=${`thumbs-up-${a.conversationId}-${s ? "standalone" : "list"}`}></calcite-action><calcite-tooltip .referenceElement=${`thumbs-up-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start><span>${t.helpful ?? "Helpful"}</span></calcite-tooltip><calcite-action .text=${t.thumbsdown ?? "Thumbs down"} scale=s icon=thumbs-down @click=${a.onThumbsDown} .id=${`thumbs-down-${a.conversationId}-${s ? "standalone" : "list"}`}></calcite-action><calcite-tooltip .referenceElement=${`thumbs-down-${a.conversationId}-${s ? "standalone" : "list"}`} placement=top-start><span>${t.unhelpful ?? "Unhelpful"}</span></calcite-tooltip></div></calcite-card>`)}</calcite-card-group></calcite-block>`;
}
const be = "arcade_generation", ge = "ArcadeCodeRequest", x = "visualization", ve = "ArcadeAssistant", ye = [
  "your_comments",
  "user_evaluation",
  "question",
  "answer",
  "useragent",
  "_source",
  "version",
  "context",
  "extras"
];
let q = [], M = !0;
class $e extends L {
  constructor() {
    super(...arguments), this.textAreaRef = B(), this.messages = W(), this._feedbackService = void 0, this.handleFeedbackButton = (e, t) => {
      this.feedback = { active: !0, value: e, item: t, termsAccepted: !1, text: "", submitting: !1 };
    }, this.handleFeedbackSheetClose = () => {
      this.feedback = {
        active: !1
      };
    }, this.handleFeedbackSubmit = async () => {
      if (!this.feedback.active)
        return;
      this.feedback = { ...this.feedback, submitting: !0 };
      const e = await this.ensureFeedbackService();
      if (!e) {
        this.handleFeedbackSheetClose();
        return;
      }
      const t = this.getModel(), o = t ? await this.getProfileAndMetadata(t.uri) : void 0, c = X.applicationName;
      !c && !this.appVersion && console.warn("Arcade Assistant: esriConfig.applicationName is not set. Consider setting it to help identify the application version in feedback.");
      const n = {
        user_evaluation: this.feedback.value,
        your_comments: this.feedback.text || void 0,
        question: this.feedback.item?.question ?? "",
        /**
         * The answer may contain invalid or partial HTML, which causes the feature service's HTML sanitizer to strip or reject it.
         * To preserve the content reliably, we encode it using encodeURIComponent, which escapes all non-ASCII and special characters.
         * This ensures the content passes through the sanitizer safely and can be decoded later. This is a temporary workaround until
         * the sanitizer can be disabled on the service.
         */
        answer: encodeURIComponent(this.feedback.item?.formattedScript ?? ""),
        useragent: navigator.userAgent,
        _source: ve,
        version: this.appVersion || c || "NOT_SET",
        context: o ? JSON.stringify(o) : "",
        extras: o?.profileName ?? "",
        conversation_id: this.feedback.item?.conversationId ?? "",
        privacy_statement: this.feedback.termsAccepted ? "I_agree" : ""
      };
      try {
        await e.applyEdits({ addFeatures: [new Z({ attributes: n })] });
      } catch (p) {
        console.error("Error submitting feedback:", p);
      }
      this.handleFeedbackSheetClose();
    }, this.handleClose = (e) => {
      this.closed = !0, this.closePanel?.(e);
    }, this.setQuestion = (e) => {
      this.question = e;
    }, this.onSuggestionClick = (e) => {
      this.textAreaRef.value && (this.textAreaRef.value.value = e, this.textAreaRef.value.setFocus(), this.setQuestion(e));
    }, this.isViewingResult = !1, this.confirmationActive = !1, this.isSubmitting = !1, this.errorMessage = void 0, this.question = "", this.history = [], this.feedback = {
      active: !1
    }, this.splashActive = !0, this.context = void 0, this.historyCards = [], this.helpBase = "", this.assistantsEnabled = !1, this.closed = !1;
  }
  static {
    this.properties = { isViewingResult: 16, confirmationActive: 16, isSubmitting: 16, errorMessage: 16, question: 16, history: 16, feedback: 16, splashActive: 16, context: 16, historyCards: 16, closePanel: 0, insertText: 0, helpBase: 1, assistantsEnabled: 5, layer: [1, { type: Object }], portalUrl: 1, serviceUrl: 1, closed: 7, editorRef: 0, feedbackServiceUrl: 1, appVersion: 1 };
  }
  static {
    this.styles = te;
  }
  get disclaimerComment() {
    return `// ${this.messages.disclaimercomment}
// ${this.messages.disclaimerpt2 ?? "AI-generated content may be inaccurate. Review before using."}`;
  }
  get assistantHelpUrl() {
    return `${this.helpBase ?? "https://doc.arcgis.com/en/arcgis-online/"}administer/configure-assistants.htm`;
  }
  get assistantOverviewDocUrl() {
    return `${this.helpBase ?? "https://doc.arcgis.com/en/arcgis-online/"}create-maps/understand-arcade-assistant.htm`;
  }
  get usingTheAssitantDocUrl() {
    return `${this.helpBase ?? "https://doc.arcgis.com/en/arcgis-online/"}create-maps/use-arcade-assistant.htm`;
  }
  async destroy() {
  }
  load() {
    const e = async () => {
      const t = this.getModel();
      if (!t) {
        this.context = void 0;
        return;
      }
      const o = await this.getProfileAndMetadata(t.uri);
      this.context = o;
    };
    e().catch((t) => {
      console.error("Error setting initial context in Arcade Assistant:", t);
    }), this.manager.onLifecycle(() => ({ remove: I.onModelContextDidChange(async () => {
      e().catch((o) => {
        console.error("Error setting context in Arcade Assistant:", o);
      });
    }).dispose }));
  }
  willUpdate(e) {
    (e.has("history") || e.has("messages")) && (this.historyCards = this.getHistoryCards());
  }
  connectedCallback() {
    super.connectedCallback(), this.history = q, this.splashActive = M;
  }
  disconnectedCallback() {
    super.disconnectedCallback(), q = this.history, M = this.splashActive, this.copySuccessTimer && (clearTimeout(this.copySuccessTimer), this.copySuccessTimer = void 0);
  }
  getHistoryCards() {
    return this.history.reduce((e, t) => (!t.script || (e.push({
      priorPrompt: t.question,
      code: t.formattedScript || t.error || "",
      chips: [
        {
          label: t.profile ?? x,
          icon: "map-information",
          appearance: "outline",
          scale: "s",
          conversationId: t.conversationId ?? "",
          tooltip: (this.messages.profilecontextlabel ?? "Profile context: {profileName}").replace("{profileName}", t.profile ?? x)
        },
        ...t.layerName ? [
          {
            label: (this.messages.numlayers ?? "{numLayers} {layerOrLayers}").replace("{numLayers}", "1").replace("{layerOrLayers}", this.messages.numlayerssingular ?? "layer"),
            icon: "layer",
            appearance: "outline",
            scale: "s",
            conversationId: t.conversationId ?? "",
            tooltip: (this.messages.layercontextlabel ?? "Layer context: {layerName}").replace("{layerName}", t.layerName ?? "No layer")
          }
        ] : []
      ],
      onAddToEditor: (o) => {
        G(this.editorRef.editorInstance, t.formattedScript ?? "", o, this.disclaimerComment);
      },
      onThumbsUp: () => this.handleFeedbackButton("good", t),
      onThumbsDown: () => this.handleFeedbackButton("bad", t),
      conversationId: t.conversationId ?? "",
      message: t.message
    }), this.isViewingResult), e), []);
  }
  setFeedbackState(e) {
    this.feedback = e;
  }
  async ensureFeedbackService() {
    if (this._feedbackService)
      return this._feedbackService;
    if (!this.feedbackServiceUrl)
      return;
    const e = new Y({ url: this.feedbackServiceUrl });
    await e.load();
    const t = ye.filter((o) => !e.fields.some((c) => c.name === o));
    if (t.length > 0) {
      console.error(`Missing the following fields in feedback service: ${t.join(", ")}. Feedback will not be logged.`);
      return;
    }
    return this._feedbackService = e, e;
  }
  getModel() {
    return this.editorRef.editorInstance?.getModel();
  }
  async submitQuestion() {
    if (this.question) {
      this.isSubmitting = !0;
      try {
        const e = await this.getToken(), t = this.getModel();
        if (!t)
          return;
        const { profileName: o, metadata: c, layerName: n } = await this.getProfileAndMetadata(t.uri), d = (await ie({
          baseUrl: this.serviceUrl,
          skillId: be,
          message: this.question,
          authToken: e,
          context: {
            kind: ge,
            context: {
              profile_name: o,
              metadata: c
            }
          }
        })).find((s) => s.context?.kind === "ArcadeCodeResponse");
        if (!d)
          return;
        d.context?.kind === "ArcadeCodeResponse" && (this.history = [
          {
            script: d?.context?.arcadeCode?.code,
            formattedScript: J(d?.context?.arcadeCode?.code ?? ""),
            error: void 0,
            question: this.question,
            profile: o,
            conversationId: d.conversationId,
            layerName: n ?? void 0,
            ...d.message ? { message: d.message } : {}
          },
          ...this.history
        ], this.isViewingResult = !0, this.errorMessage = void 0);
      } catch (e) {
        e instanceof Error && e.name === "ArcadeAssistantError" ? this.errorMessage = ae(e) || this.messages.erroroccurred || "An error occurred." : this.errorMessage = this.messages.erroroccurred || "An error occurred.", console.error("Error in Arcade Assistant:", e);
      } finally {
        this.isSubmitting = !1;
      }
    }
  }
  async getToken() {
    const { token: e } = await j.getCredential(this.portalUrl);
    return e;
  }
  async getProfileAndMetadata(e) {
    let t = x;
    const o = I.getEditorProfileForModel(e);
    o?.loaded || await o?.loadSource();
    const c = this.editorRef.profile;
    c && "id" in c && (t = c.id);
    const n = o?.definition?.variables?.find((s) => s.type === "feature");
    let p = [], d;
    return n?.definition && "fields" in n.definition && (p = n.definition.fields.map((s) => ({
      name: s.name,
      type: s.type,
      alias: s.alias
    })), ee(n?.definition) && (d = n.definition.title)), { profileName: t, metadata: p, layerName: d };
  }
  onCopyCode(e) {
    const t = e ? `${this.disclaimerComment}
${e}` : "";
    return navigator.clipboard.writeText(t);
  }
  render() {
    return this.closed ? null : this.assistantsEnabled ? r`<calcite-flow><calcite-flow-item closable heading-level=2 .heading=${this.messages.arcadeassistant ?? "Arcade assistant"} @calciteFlowItemClose=${this.handleClose}>${this.splashActive && r`<calcite-scrim></calcite-scrim>` || ""}${ce({ messages: this.messages })}${re({ popoverContent: R({ assistantsEnabled: this.assistantsEnabled, messages: this.messages, assistantHelpUrl: this.assistantHelpUrl, helpTopicUrl: this.usingTheAssitantDocUrl }) })}<calcite-shell class="unstyled-shell">${this.splashActive && me({ messages: this.messages, onProceed: () => {
      this.splashActive = !1;
    }, helpTopicUrl: this.assistantOverviewDocUrl, onExit: (e) => this.closePanel && this.closePanel(e) }) || ""}${ne({ active: this.confirmationActive, setActive: (e) => this.confirmationActive = e, messages: this.messages })}${de({ isDisabled: this.isSubmitting, isReadOnly: !1, setQuestion: this.setQuestion, setIsViewingResult: (e) => this.isViewingResult = e, mode: this.isViewingResult ? "refine" : "prompt", textAreaRef: this.textAreaRef, question: this.question, errorMessage: this.errorMessage, setErrorMessage: (e) => this.errorMessage = e, submitQuestion: this.submitQuestion.bind(this), messages: this.messages, context: this.context })}<calcite-block-group label="interactive blocks">${!this.isViewingResult && !this.isSubmitting && fe({ messages: this.messages, suggestions: [this.messages.prompsuggestion1 ?? "", this.messages.prompsuggestion2 ?? ""], onSuggestionClick: this.onSuggestionClick }) || ""}<calcite-block .hidden=${!this.isSubmitting} .heading=${this.messages.generatingresponse ?? "Generating response..."}></calcite-block>${P({
      expanded: this.isViewingResult,
      messages: this.messages,
      collapsible: !0,
      showEffectsIcon: !0,
      /** Only show the latest history card */
      cards: this.historyCards.length ? [this.historyCards[0]] : [],
      heading: this.messages.besteffort ?? "Here's the assistant's best effort",
      hidden: this.isSubmitting || !this.isViewingResult,
      onCopyCode: this.onCopyCode.bind(this),
      standalone: !0
    })}${P({
      expanded: !this.isViewingResult,
      messages: this.messages,
      collapsible: !0,
      /**
       * When the user is viewing the latest result, we display the first history card
       * (historyCards[0]) separately at the top in "standalone" mode.
       * To avoid showing the same card twice, we remove (slice out) the first card
       * from the list of recent prompts below.
       */
      cards: this.isViewingResult ? this.historyCards.slice(1) : this.historyCards,
      heading: this.messages.recentprompts ?? "Recent prompts",
      hidden: this.isSubmitting || this.history.length === 0,
      onCopyCode: this.onCopyCode.bind(this)
    })}</calcite-block-group>${this.feedback.active ? he({ setOpen: this.handleFeedbackSheetClose, messages: this.messages, onSubmit: this.handleFeedbackSubmit, feedback: this.feedback, setFeedback: this.setFeedbackState.bind(this) }) : null}</calcite-shell></calcite-flow-item></calcite-flow>` : r`<calcite-flow><calcite-flow-item .selected=${!this.isViewingResult} closable heading-level=2 .heading=${this.messages.arcadeassistant ?? "Arcade assistant"} @calciteFlowItemClose=${this.handleClose}>${R({ slot: "content-top", messages: this.messages, assistantsEnabled: this.assistantsEnabled, assistantHelpUrl: this.assistantHelpUrl, helpTopicUrl: this.usingTheAssitantDocUrl })}</calcite-flow-item></calcite-flow>`;
  }
}
V("arcgis-arcade-coding-assistant", $e);
export {
  $e as ArcgisArcadeCodingAssistant
};
